{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioNeuralNet Cancer Example 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZN3c1010Dispatcher17runRecordFunctionERN2at14RecordFunctionESt17reference_wrapperIKNS_14FunctionSchemaEENS_11DispatchKeyE\n",
      "  warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "/home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSs\n",
      "  warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "/home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSs\n",
      "  warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "/home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSs\n",
      "  warnings.warn(\n",
      "/home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.venv/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSs\n",
      "  warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common Patients: 103\n",
      "Genes shaope: (103, 60525)\n",
      "Proteins shape: (103, 11889)\n",
      "Clinical shape: (103, 11)\n",
      "Phenotype shape: (103, 1)\n",
      "            ENSG00000121410.12  ENSG00000268895.6  ENSG00000148584.15  \\\n",
      "Patient_ID                                                              \n",
      "C3L-00004                0.041              0.041               0.004   \n",
      "C3L-00010               -0.007             -0.007               0.007   \n",
      "C3L-00011               -0.064             -0.064               0.007   \n",
      "C3L-00026                0.010              0.010              -0.007   \n",
      "C3L-00079                0.186              0.186               0.266   \n",
      "\n",
      "            ENSG00000175899.15  ENSG00000245105.4  ENSG00000166535.20  \\\n",
      "Patient_ID                                                              \n",
      "C3L-00004                0.015              0.015               0.015   \n",
      "C3L-00010               -0.001             -0.001              -0.001   \n",
      "C3L-00011                0.004              0.004               0.004   \n",
      "C3L-00026                0.008              0.008               0.008   \n",
      "C3L-00079               -0.023             -0.023              -0.023   \n",
      "\n",
      "            ENSG00000256661.1  ENSG00000256904.1  ENSG00000256069.7  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004               0.015              0.015              0.015   \n",
      "C3L-00010              -0.001             -0.001             -0.001   \n",
      "C3L-00011               0.004              0.004              0.004   \n",
      "C3L-00026               0.008              0.008              0.008   \n",
      "C3L-00079              -0.023             -0.023             -0.023   \n",
      "\n",
      "            ENSG00000184389.9  ...  ENSG00000198455.4  ENSG00000070476.15  \\\n",
      "Patient_ID                     ...                                          \n",
      "C3L-00004              -0.004  ...             -0.016               0.001   \n",
      "C3L-00010              -0.005  ...             -0.037               0.005   \n",
      "C3L-00011              -0.037  ...             -0.690               0.007   \n",
      "C3L-00026               0.000  ...             -0.624              -0.007   \n",
      "C3L-00079               0.134  ...              0.167               0.314   \n",
      "\n",
      "            ENSG00000203995.10  ENSG00000232242.2  ENSG00000162378.13  \\\n",
      "Patient_ID                                                              \n",
      "C3L-00004               -0.004             -0.009              -0.004   \n",
      "C3L-00010               -0.005              0.001              -0.005   \n",
      "C3L-00011                0.018             -0.676               0.018   \n",
      "C3L-00026                0.000              0.003               0.000   \n",
      "C3L-00079                0.134             -0.077               0.134   \n",
      "\n",
      "            ENSG00000159840.16  ENSG00000274572.1  ENSG00000074755.15  \\\n",
      "Patient_ID                                                              \n",
      "C3L-00004                0.024             -0.009               0.021   \n",
      "C3L-00010                0.003              0.006               0.007   \n",
      "C3L-00011                0.007              0.007              -0.697   \n",
      "C3L-00026                0.669             -0.008               0.001   \n",
      "C3L-00079                0.134              0.462               0.003   \n",
      "\n",
      "            ENSG00000272920.1  ENSG00000266919.3  \n",
      "Patient_ID                                        \n",
      "C3L-00004               0.021              0.021  \n",
      "C3L-00010               0.007              0.007  \n",
      "C3L-00011              -0.697             -0.015  \n",
      "C3L-00026               0.001              0.001  \n",
      "C3L-00079               0.003              0.001  \n",
      "\n",
      "[5 rows x 60525 columns]\n",
      "            ENSP00000000233.5  ENSP00000000412.3  ENSP00000000442.6  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004            0.248796           0.308289          -0.315825   \n",
      "C3L-00010            0.245087           0.131992          -0.016811   \n",
      "C3L-00011            0.576821           0.443678          -0.544452   \n",
      "C3L-00026            0.293042           0.250350           0.142924   \n",
      "C3L-00079            0.174315           0.026805          -0.663191   \n",
      "\n",
      "            ENSP00000001008.4  ENSP00000002125.4  ENSP00000002165.5  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004            0.051528          -0.323475          -0.720420   \n",
      "C3L-00010           -0.114728          -0.816992           0.104260   \n",
      "C3L-00011            0.120776          -0.075035          -0.738826   \n",
      "C3L-00026           -0.173812           0.589458           0.238597   \n",
      "C3L-00079            0.075943          -0.240099          -0.293380   \n",
      "\n",
      "            ENSP00000002501.6  ENSP00000003084.6  ENSP00000003100.8  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004                 NaN          -0.832085          -0.629139   \n",
      "C3L-00010                 NaN           1.266034          -0.573329   \n",
      "C3L-00011                 NaN          -0.484732          -0.725274   \n",
      "C3L-00026            0.957386          -3.178976           0.372489   \n",
      "C3L-00079                 NaN           0.008966          -0.469829   \n",
      "\n",
      "            ENSP00000003302.4  ...  ENSP00000500245.1  ENSP00000500275.1  \\\n",
      "Patient_ID                     ...                                         \n",
      "C3L-00004           -0.317251  ...           0.359093           1.609902   \n",
      "C3L-00010           -0.172712  ...          -1.592004          -0.542294   \n",
      "C3L-00011            1.079570  ...          -0.523301          -0.389680   \n",
      "C3L-00026            0.404721  ...          -0.249875          -1.492455   \n",
      "C3L-00079                 NaN  ...          -1.423105          -0.921139   \n",
      "\n",
      "            ENSP00000500313.1  ENSP00000500633.1  ENSP00000500710.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004            0.054346          -0.471825          -0.095824   \n",
      "C3L-00010                 NaN          -0.208478          -0.143536   \n",
      "C3L-00011                 NaN           0.550085          -1.617542   \n",
      "C3L-00026                 NaN           0.005406          -0.281776   \n",
      "C3L-00079                 NaN          -0.042932          -0.921862   \n",
      "\n",
      "            ENSP00000501064.1  ENSP00000501126.1  ENSP00000501261.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004            0.240579           0.689392                NaN   \n",
      "C3L-00010            0.049889           0.499405           1.422138   \n",
      "C3L-00011            0.485815          -0.327129                NaN   \n",
      "C3L-00026            0.057764          -0.073197                NaN   \n",
      "C3L-00079            0.385364                NaN                NaN   \n",
      "\n",
      "            ENSP00000501300.1  ENSP00000501491.1  \n",
      "Patient_ID                                        \n",
      "C3L-00004           -0.072279          -0.655139  \n",
      "C3L-00010           -0.058976          -0.189293  \n",
      "C3L-00011            0.344999          -0.688779  \n",
      "C3L-00026           -0.159555          -0.023920  \n",
      "C3L-00079            0.387207          -1.057636  \n",
      "\n",
      "[5 rows x 11889 columns]\n",
      "             age  sex  tumor_site  tumor_laterality  tumor_focality  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004   72.0    1           0                 0               1   \n",
      "C3L-00010   30.0    1           0                 0               1   \n",
      "C3L-00011   63.0    0           0                 1               1   \n",
      "C3L-00026   65.0    0           0                 1               1   \n",
      "C3L-00079   49.0    1           0                 1               1   \n",
      "\n",
      "            tumor_size_cm  histologic_grade    bmi  alcohol_consumption  \\\n",
      "Patient_ID                                                                \n",
      "C3L-00004            12.0                 2  22.80                    4   \n",
      "C3L-00010             6.5                 2  34.15                    1   \n",
      "C3L-00011            12.0                 3  27.47                    0   \n",
      "C3L-00026             2.0                 2  28.23                    4   \n",
      "C3L-00079            14.0                 2  37.88                    4   \n",
      "\n",
      "            tobacco_smoking_history  medical_condition  \n",
      "Patient_ID                                              \n",
      "C3L-00004                         0                 68  \n",
      "C3L-00010                         2                 13  \n",
      "C3L-00011                         3                 60  \n",
      "C3L-00026                         3                 26  \n",
      "C3L-00079                         3                 16  \n",
      "            tumor_stage_pathological\n",
      "Patient_ID                          \n",
      "C3L-00004                          1\n",
      "C3L-00010                          0\n",
      "C3L-00011                          1\n",
      "C3L-00026                          0\n",
      "C3L-00079                          1\n"
     ]
    }
   ],
   "source": [
    "from bioneuralnet.external_tools import get_cancer_data, preprocess_clinical, filter_common_patients\n",
    "import cptac\n",
    "\n",
    "ccrcc = cptac.Ccrcc()\n",
    "\n",
    "genes_raw, proteins_raw, clinical_raw = get_cancer_data(ccrcc)\n",
    "clinical_raw.to_csv(\"clinical_raw.csv\")\n",
    "genes, proteins, clinical_filtered = filter_common_patients(genes_raw, proteins_raw, clinical_raw)\n",
    "\n",
    "clinical, phenotype = preprocess_clinical(clinical_filtered)\n",
    "\n",
    "print(f\"Genes shaope: {genes.shape}\")\n",
    "print(f\"Proteins shape: {proteins.shape}\")\n",
    "print(f\"Clinical shape: {clinical.shape}\")\n",
    "print(f\"Phenotype shape: {phenotype.shape}\")\n",
    "\n",
    "print(genes.head())\n",
    "print(proteins.head())\n",
    "print(clinical.head())\n",
    "print(phenotype.head())\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before filtering:\n",
      "\n",
      "Gnes has 0 missing values\n",
      "Proteins has 261664 missing values\n",
      "\n",
      "After Droping Nans:\n",
      "(103, 60525)\n",
      "(103, 7016)\n",
      "After filtering:\n",
      "Genomics shape: (103, 539)\n",
      "Proteomics shape: (103, 1746)\n",
      "            ENSG00000286130.1  ENSG00000278847.1  ENSG00000275310.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004              -0.911             -0.911             -0.911   \n",
      "C3L-00010              -1.019             -1.019             -1.019   \n",
      "C3L-00011              -0.145             -0.145             -0.145   \n",
      "C3L-00026               0.023              0.023              0.023   \n",
      "C3L-00079              -1.137             -1.137             -1.137   \n",
      "\n",
      "            ENSG00000286120.1  ENSG00000286173.1  ENSG00000233619.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004              -0.911             -0.911             -0.911   \n",
      "C3L-00010              -1.019             -1.019             -1.019   \n",
      "C3L-00011              -0.145             -0.145             -0.147   \n",
      "C3L-00026               0.023              0.023              0.023   \n",
      "C3L-00079              -1.137             -1.137             -1.137   \n",
      "\n",
      "            ENSG00000284380.1  ENSG00000235094.3  ENSG00000275352.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004              -0.911             -0.911             -0.911   \n",
      "C3L-00010              -1.019             -1.019             -1.019   \n",
      "C3L-00011              -0.147             -0.145             -0.145   \n",
      "C3L-00026               0.023              0.023              0.023   \n",
      "C3L-00079              -1.137             -1.137             -1.137   \n",
      "\n",
      "            ENSG00000278197.1  ...  ENSG00000236379.2  ENSG00000230819.1  \\\n",
      "Patient_ID                     ...                                         \n",
      "C3L-00004              -0.911  ...             -0.911             -0.911   \n",
      "C3L-00010              -1.019  ...             -1.019             -1.019   \n",
      "C3L-00011              -0.147  ...             -0.147             -0.147   \n",
      "C3L-00026               0.023  ...              0.023              0.023   \n",
      "C3L-00079              -1.137  ...             -1.137             -1.137   \n",
      "\n",
      "            ENSG00000224634.1  ENSG00000241200.1  ENSG00000244246.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004              -0.911             -0.911             -0.911   \n",
      "C3L-00010              -1.019             -1.019             -1.019   \n",
      "C3L-00011              -0.145             -0.145             -0.145   \n",
      "C3L-00026               0.023              0.023              0.023   \n",
      "C3L-00079              -1.137             -1.137             -1.137   \n",
      "\n",
      "            ENSG00000239893.2  ENSG00000217896.2  ENSG00000228764.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004              -0.911             -0.911             -0.911   \n",
      "C3L-00010              -1.019             -1.019             -1.019   \n",
      "C3L-00011              -0.145             -0.147             -0.147   \n",
      "C3L-00026               0.023              0.023              0.023   \n",
      "C3L-00079              -1.137             -1.137             -1.137   \n",
      "\n",
      "            ENSG00000232522.1  ENSG00000215603.4  \n",
      "Patient_ID                                        \n",
      "C3L-00004              -0.911             -0.911  \n",
      "C3L-00010              -1.019             -1.019  \n",
      "C3L-00011              -0.147             -0.145  \n",
      "C3L-00026               0.023              0.023  \n",
      "C3L-00079              -1.137             -1.137  \n",
      "\n",
      "[5 rows x 539 columns]\n",
      "            ENSP00000003100.8  ENSP00000004982.3  ENSP00000005178.5  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004           -0.629139          -0.963552           1.071050   \n",
      "C3L-00010           -0.573329          -2.629349          -0.601561   \n",
      "C3L-00011           -0.725274          -2.258016          -0.666070   \n",
      "C3L-00026            0.372489          -1.087033           1.488454   \n",
      "C3L-00079           -0.469829          -0.882213           1.057320   \n",
      "\n",
      "            ENSP00000005260.8  ENSP00000005587.2  ENSP00000006777.6  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004           -0.709452           0.935768          -0.605336   \n",
      "C3L-00010           -0.303856           0.229396          -0.792211   \n",
      "C3L-00011            1.859668          -0.529298          -0.128768   \n",
      "C3L-00026           -0.509028           0.074067           0.517890   \n",
      "C3L-00079            0.901664           0.776101          -0.157385   \n",
      "\n",
      "            ENSP00000009180.4  ENSP00000009530.7  ENSP00000010404.2  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004           -0.656928           0.289200          -0.964154   \n",
      "C3L-00010           -0.530668          -0.073076          -0.486055   \n",
      "C3L-00011           -1.710060           0.193294           0.594775   \n",
      "C3L-00026           -0.414820           0.194276          -0.018702   \n",
      "C3L-00079           -0.562946           0.811032          -1.001052   \n",
      "\n",
      "            ENSP00000011653.4  ...  ENSP00000498075.1  ENSP00000498408.1  \\\n",
      "Patient_ID                     ...                                         \n",
      "C3L-00004            0.563387  ...           0.433326          -0.797754   \n",
      "C3L-00010            0.414442  ...           0.580986          -0.119409   \n",
      "C3L-00011            0.110896  ...           1.647099           1.856893   \n",
      "C3L-00026           -0.310567  ...           0.459815           0.166078   \n",
      "C3L-00079            0.610429  ...          -0.395737          -0.535006   \n",
      "\n",
      "            ENSP00000498441.1  ENSP00000498786.1  ENSP00000498932.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004           -0.506954          -0.363214          -0.215305   \n",
      "C3L-00010            0.485804          -0.052700          -0.386301   \n",
      "C3L-00011            1.350262           0.026752          -0.558920   \n",
      "C3L-00026           -0.822164           0.712062          -0.543199   \n",
      "C3L-00079            0.248644          -0.265680           0.033349   \n",
      "\n",
      "            ENSP00000499129.1  ENSP00000499271.1  ENSP00000499832.1  \\\n",
      "Patient_ID                                                            \n",
      "C3L-00004           -0.408601          -1.092168          -0.045349   \n",
      "C3L-00010           -0.272673          -0.564413          -1.100544   \n",
      "C3L-00011            1.512444          -2.248855          -1.077198   \n",
      "C3L-00026            0.163810          -0.603413          -0.455064   \n",
      "C3L-00079            0.039393          -1.616319          -1.987001   \n",
      "\n",
      "            ENSP00000500245.1  ENSP00000501491.1  \n",
      "Patient_ID                                        \n",
      "C3L-00004            0.359093          -0.655139  \n",
      "C3L-00010           -1.592004          -0.189293  \n",
      "C3L-00011           -0.523301          -0.688779  \n",
      "C3L-00026           -0.249875          -0.023920  \n",
      "C3L-00079           -1.423105          -1.057636  \n",
      "\n",
      "[5 rows x 1746 columns]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.feature_selection import VarianceThreshold\n",
    "import pandas as pd\n",
    "\n",
    "print(\"Before filtering:\\n\")\n",
    "print(f\"Gnes has {genes.isnull().sum().sum()} missing values\")\n",
    "print(f\"Proteins has {proteins.isnull().sum().sum()} missing values\")\n",
    "\n",
    "# remove columns with Nan values\n",
    "genes = genes.dropna(axis=1, how='any')\n",
    "proteins = proteins.dropna(axis=1, how='any')\n",
    "\n",
    "print(\"\\nAfter Droping Nans:\")\n",
    "print(genes.shape)\n",
    "print(proteins.shape)\n",
    "\n",
    "variance_filter = VarianceThreshold(threshold=0.2)\n",
    "genomics_array = variance_filter.fit_transform(genes)\n",
    "\n",
    "genomics = pd.DataFrame(\n",
    "    genomics_array,\n",
    "    index=genes.index,\n",
    "    columns=genes.columns[variance_filter.get_support()]\n",
    ")\n",
    "\n",
    "variance_filter_prot = VarianceThreshold(threshold=0.2)\n",
    "proteomics_array = variance_filter_prot.fit_transform(proteins)\n",
    "\n",
    "proteomics = pd.DataFrame(\n",
    "    proteomics_array,\n",
    "    index=proteins.index,\n",
    "    columns=proteins.columns[variance_filter_prot.get_support()]\n",
    ")\n",
    "\n",
    "print(\"After filtering:\")\n",
    "print(\"Genomics shape:\", genomics.shape)\n",
    "print(\"Proteomics shape:\", proteomics.shape)\n",
    "print(genomics.head())\n",
    "print(proteomics.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-04 20:32:32,646 - bioneuralnet.external_tools.smccnet - INFO - Initialized SmCCNet with parameters:\n",
      "2025-04-04 20:32:32,646 - bioneuralnet.external_tools.smccnet - INFO - K-Fold: 5\n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - Summarization: NetSHy\n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - Evaluation method: \n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - ncomp_pls: 0\n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - subSampNum: 1000\n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - BetweenShrinkage: 5.0\n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - Seed: 723\n",
      "2025-04-04 20:32:32,647 - bioneuralnet.external_tools.smccnet - INFO - Output directory set to: cancer_output_2\n",
      "2025-04-04 20:32:32,648 - bioneuralnet.external_tools.smccnet - INFO - Starting SmCCNet workflow.\n",
      "2025-04-04 20:32:32,648 - bioneuralnet.external_tools.smccnet - INFO - Validating and serializing input data for SmCCNet...\n",
      "2025-04-04 20:32:32,648 - bioneuralnet.external_tools.smccnet - INFO - All DataFrames already share the same named index. No rename needed.\n",
      "2025-04-04 20:32:32,668 - bioneuralnet.external_tools.smccnet - INFO - Serialized omics_1 with 103 samples.\n",
      "2025-04-04 20:32:32,760 - bioneuralnet.external_tools.smccnet - INFO - Serialized omics_2 with 103 samples.\n",
      "2025-04-04 20:32:32,760 - bioneuralnet.external_tools.smccnet - INFO - Executing SmCCNet R script...\n",
      "2025-04-04 20:34:58,036 - bioneuralnet.external_tools.smccnet - INFO - SMCCNET R script output:\n",
      "Allowing multi-threading with up to 32 threads.\n",
      "\n",
      "**********************************\n",
      "* Welcome to Automated SmCCNet! *\n",
      "**********************************\n",
      "\n",
      "This project uses multiomics CCA \n",
      "\n",
      "--------------------------------------------------\n",
      ">> Now determining the scaling factor for multi-omics analysis...\n",
      "--------------------------------------------------\n",
      "\n",
      "\n",
      "The scaling factor selection is:  \n",
      "\n",
      "genomics-proteomics: 0.129377330523842 \n",
      "genomics-Phenotype: 1 \n",
      "proteomics-Phenotype: 1 \n",
      "\n",
      "--------------------------------------------------\n",
      ">> Determining the best penalty selection through cross-validation...\n",
      "--------------------------------------------------\n",
      "\n",
      "\n",
      "The best penalty term for genomics after 5-fold cross-validation is: 0.5 \n",
      "The best penalty term for proteomics after 5-fold cross-validation is: 0.5 \n",
      "with testing canonical correlation = 0.751, and prediction error = 0.549 \n",
      "Now running multi-omics CCA with best penalty term on the complete dataset. \n",
      "\n",
      "--------------------------------------------------\n",
      ">> Now starting network clustering...\n",
      "--------------------------------------------------\n",
      "\n",
      "Clustering completed...\n",
      "\n",
      "--------------------------------------------------\n",
      ">> Now starting network pruning and summarization score extraction...\n",
      "--------------------------------------------------\n",
      "\n",
      "There are 3 network modules before pruning\n",
      " Now extracting subnetwork for network module  1 \n",
      "Network module 1 Result:The final network size is: 70 with maximum PC correlation w.r.t. phenotype to be: 0.36\n",
      " Now extracting subnetwork for network module  2 \n",
      "Network module 2 Result:The final network size is: 46 with maximum PC correlation w.r.t. phenotype to be: -0.224\n",
      " Now extracting subnetwork for network module  3 \n",
      "Network module 3 Result:The final network size is: 21 with maximum PC correlation w.r.t. phenotype to be: 0.075\n",
      "\n",
      "************************************\n",
      "*   Execution Finished!            *\n",
      "*   Automated SmCCNet Completed!   *\n",
      "************************************\n",
      "\n",
      "There are 4 files stored in the local directory: \n",
      "1. CVResult.csv: cross-validation result. \n",
      "2. CVFold.Rdata: omics and phenotype data splited into different folds. \n",
      "3. globalNetwork.Rdata: global network adjacency matrix, omics data, phenotype. \n",
      "4. size_a_net_b.Rdata: subnetworks result file after clustering and network pruning. \n",
      "SUBNETWORK RESULT FILE CONTAINS:\n",
      "1. correlation_sub: correlation matrix for the subnetwork.\n",
      "2. M: adjacency matrix for the subnetwork.\n",
      "3. omics_corelation_data: individual molecular feature correlation with phenotype.\n",
      "4. pc_correlation: first 3 PCs correlation with phenotype.\n",
      "5. pc_loading: principal component loadings.\n",
      "6. pca_x1_score: principal component score and phenotype data.\n",
      "7. mod_size: number of molecular features in the subnetwork.\n",
      "8. sub_type: type of feature for each molecular features.\n",
      "\n",
      "NOTE:\n",
      "1. The results are saved in the user-specified directory.\n",
      "   If not specified, use 'getwd()' to find the current working directory.\n",
      "\n",
      "2. The results are in 'size_a_net_b.Rdata', where:\n",
      "   'a' = network size, 'b' = module number.\n",
      "\n",
      "3. To prevent overwriting result files across projects, rename the result file after each run.\n",
      "\n",
      "4. Subnetwork visualization can be done through shinyApp or cytoscape. Please refer to the network visualization section in the multi-omics or single-omics vignette for more detail. For the Shiny app, visit: https://smccnet.shinyapps.io/smccnetnetwork/\n",
      "\n",
      "************************************\n",
      "\n",
      "2025-04-04 20:34:58,037 - bioneuralnet.external_tools.smccnet - WARNING - SMCCNET R script warnings/errors:\n",
      "Loading required package: dynamicTreeCut\n",
      "Loading required package: fastcluster\n",
      "\n",
      "Attaching package: ‘fastcluster’\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    hclust\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘WGCNA’\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "DEBUG: We have 103 samples in Y. Range: [0, 1]\n",
      "DEBUG: Omics data shapes:\n",
      "  Omics 1: 103 x 539\n",
      "  Omics 2: 103 x 1746\n",
      "Warning message:\n",
      "In split.default(seq_len(nrow(omics_list[[1]])), sample(seq_len(nrow(omics_list[[1]])),  :\n",
      "  data length is not a multiple of split variable\n",
      "Debug fold #1 => size: 21\n",
      "Debug fold #2 => size: 20\n",
      "Debug fold #3 => size: 21\n",
      "Debug fold #4 => size: 20\n",
      "Debug fold #5 => size: 21\n",
      "Multi-omics CCA scenario\n",
      "\n",
      " Progress: ───────────────────────────────────────────────────              100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%\n",
      " Progress: ──────────────────────────────────────────────────────────────── 100%Warning message:\n",
      "In split.default(1:nrow(X[[1]]), sample(1:nrow(X[[1]]), Kfold)) :\n",
      "  data length is not a multiple of split variable\n",
      "Current working directory: /home/vicente/Github/BioNeuralNet/cancer_output_2\n",
      "Found files: /home/vicente/Github/BioNeuralNet/cancer_output_2/size_21_net_3.Rdata, /home/vicente/Github/BioNeuralNet/cancer_output_2/size_46_net_2.Rdata, /home/vicente/Github/BioNeuralNet/cancer_output_2/size_70_net_1.Rdata\n",
      "Processing file: /home/vicente/Github/BioNeuralNet/cancer_output_2/size_21_net_3.Rdata\n",
      "\n",
      "Subnetwork matrix from /home/vicente/Github/BioNeuralNet/cancer_output_2/size_21_net_3.Rdata written to size_21_net_3.csv\n",
      "\n",
      "\n",
      "Processing file: /home/vicente/Github/BioNeuralNet/cancer_output_2/size_46_net_2.Rdata\n",
      "\n",
      "Subnetwork matrix from /home/vicente/Github/BioNeuralNet/cancer_output_2/size_46_net_2.Rdata written to size_46_net_2.csv\n",
      "\n",
      "\n",
      "Processing file: /home/vicente/Github/BioNeuralNet/cancer_output_2/size_70_net_1.Rdata\n",
      "\n",
      "Subnetwork matrix from /home/vicente/Github/BioNeuralNet/cancer_output_2/size_70_net_1.Rdata written to size_70_net_1.csv\n",
      "\n",
      "\n",
      "\n",
      "2025-04-04 20:34:58,037 - bioneuralnet.external_tools.smccnet - INFO - Reading Global Network from: cancer_output_2/GlobalNetwork.csv\n",
      "2025-04-04 20:34:58,253 - bioneuralnet.external_tools.smccnet - INFO - Global Network shape: (2285, 2285)\n",
      "2025-04-04 20:34:58,257 - bioneuralnet.external_tools.smccnet - INFO - Found 3 clusters in cancer_output_2.\n",
      "2025-04-04 20:34:58,257 - bioneuralnet.external_tools.smccnet - INFO - GlobalNetwork stored at index 0 and clusters stored as a list of dataframes at index 1.\n",
      "2025-04-04 20:34:58,257 - bioneuralnet.external_tools.smccnet - INFO - SmCCNet workflow completed successfully.\n"
     ]
    }
   ],
   "source": [
    "from bioneuralnet.external_tools import SmCCNet\n",
    "\n",
    "smccnet = SmCCNet(\n",
    "    phenotype_df=phenotype,\n",
    "    omics_dfs=[genomics,proteomics],\n",
    "    data_types=[\"genomics\",\"proteomics\"],\n",
    "    output_dir=\"cancer_output_2\",\n",
    ")\n",
    "global_network, smccnet_clusters = smccnet.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nonzero entries: 1079462\n",
      "Total entries: 5221225\n",
      "Density (nonzero fraction): 0.2067\n",
      "Sparsity (zero fraction): 0.7933\n",
      "Number of nonzero entries: 2070\n",
      "Total entries: 2116\n",
      "Density (nonzero fraction): 0.9783\n",
      "Sparsity (zero fraction): 0.0217\n",
      "Number of nonzero entries: 4830\n",
      "Total entries: 4900\n",
      "Density (nonzero fraction): 0.9857\n",
      "Sparsity (zero fraction): 0.0143\n",
      "Number of nonzero entries: 420\n",
      "Total entries: 441\n",
      "Density (nonzero fraction): 0.9524\n",
      "Sparsity (zero fraction): 0.0476\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "matrix = global_network.values\n",
    "num_nonzero = np.count_nonzero(matrix)\n",
    "total_entries = matrix.size\n",
    "density = num_nonzero / total_entries\n",
    "sparsity = 1 - density\n",
    "\n",
    "print(f\"Number of nonzero entries: {num_nonzero}\")\n",
    "print(f\"Total entries: {total_entries}\")\n",
    "print(f\"Density (nonzero fraction): {density:.4f}\")\n",
    "print(f\"Sparsity (zero fraction): {sparsity:.4f}\")\n",
    "\n",
    "\n",
    "matrix = smccnet_clusters[0].values\n",
    "num_nonzero = np.count_nonzero(matrix)\n",
    "total_entries = matrix.size\n",
    "density = num_nonzero / total_entries\n",
    "sparsity = 1 - density\n",
    "\n",
    "print(f\"Number of nonzero entries: {num_nonzero}\")\n",
    "print(f\"Total entries: {total_entries}\")\n",
    "print(f\"Density (nonzero fraction): {density:.4f}\")\n",
    "print(f\"Sparsity (zero fraction): {sparsity:.4f}\")\n",
    "\n",
    "\n",
    "matrix = smccnet_clusters[1].values\n",
    "num_nonzero = np.count_nonzero(matrix)\n",
    "total_entries = matrix.size\n",
    "density = num_nonzero / total_entries\n",
    "sparsity = 1 - density\n",
    "\n",
    "print(f\"Number of nonzero entries: {num_nonzero}\")\n",
    "print(f\"Total entries: {total_entries}\")\n",
    "print(f\"Density (nonzero fraction): {density:.4f}\")\n",
    "print(f\"Sparsity (zero fraction): {sparsity:.4f}\")\n",
    "\n",
    "\n",
    "matrix = smccnet_clusters[2].values\n",
    "num_nonzero = np.count_nonzero(matrix)\n",
    "total_entries = matrix.size\n",
    "density = num_nonzero / total_entries\n",
    "sparsity = 1 - density\n",
    "\n",
    "print(f\"Number of nonzero entries: {num_nonzero}\")\n",
    "print(f\"Total entries: {total_entries}\")\n",
    "print(f\"Density (nonzero fraction): {density:.4f}\")\n",
    "print(f\"Sparsity (zero fraction): {sparsity:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BioNeuralNet version: 1.0\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet\n",
    "print(\"BioNeuralNet version:\", bioneuralnet.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
