{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d182cc95",
   "metadata": {},
   "source": [
    "# TCGA-BRCA Demo\n",
    "\n",
    "## Dataset Source\n",
    "\n",
    "- **Omics Data**: [FireHose BRCA](http://firebrowse.org/?cohort=BRCA)\n",
    "- **Clinical and PAM50 Data**: [TCGAbiolinks](http://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html)\n",
    "\n",
    "## Dataset Overview\n",
    "\n",
    "**Original Data**:\n",
    "\n",
    "- **Methylation**: 20,107 × 885\n",
    "- **mRNA**: 18,321 × 1,212\n",
    "- **miRNA**: 503 × 1,189\n",
    "- **PAM50**: 1,087 × 1\n",
    "- **Clinical**: 1,098 × 101\n",
    "\n",
    "- **Note: Omics matrices are features × samples; clinical matrices are samples × fields.**\n",
    "\n",
    "**PAM50 Subtype Counts**:\n",
    "\n",
    "- **LumA**: 419\n",
    "- **LumB**: 140\n",
    "- **Basal**: 130\n",
    "- **Her2**: 46\n",
    "- **Normal**: 34\n",
    "\n",
    "## Patients in Every Dataset\n",
    "\n",
    "- Total patients present in methylation, mRNA, miRNA, PAM50, and clinical: **769**\n",
    "\n",
    "## Final Shapes (Per-Patient)\n",
    "\n",
    "After aggregating multiple aliquots by mean, all modalities align on 769 patients:\n",
    "\n",
    "- **Methylation**: 769 × 20,107\n",
    "- **mRNA**: 769 × 20,531\n",
    "- **miRNA**: 769 × 503\n",
    "- **PAM50**: 769 × 1\n",
    "- **Clinical**: 769 × 119\n",
    "\n",
    "## Data Summary Table\n",
    "\n",
    "| Stage                          | Clinical    | Methylation  | miRNA       | mRNA           | PAM50 (Subtype Counts)                                         | Notes                                   |\n",
    "| ------------------------------ | ----------- | ------------ | ----------- | -------------- | -------------------------------------------------------------- | --------------------------------------- |\n",
    "| **Original Raw Data**          | 1,098 × 101 | 20,107 × 885 | 503 × 1,189 | 18,321 × 1,212 | LumA: 509<br>LumB: 209<br>Basal: 192<br>Her2: 82<br>Normal: 40 | Raw FireHose & TCGAbiolinks files       |\n",
    "| **Patient-Level Intersection** | 769 × 101   | 769 × 20,107 | 769 × 1,046 | 769 × 20,531   | LumA: 419<br>LumB: 140<br>Basal: 130<br>Her2: 46<br>Normal: 34 | Patients with complete data in all sets |\n",
    "\n",
    "## Reference Links\n",
    "\n",
    "- [FireHose BRCA](http://firebrowse.org/?cohort=BRCA)\n",
    "- [TCGAbiolinks](http://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html)\n",
    "- [Direct Download BRCA](http://firebrowse.org/?cohort=BRCA&download_dialog=true)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9698b74",
   "metadata": {},
   "source": [
    "## Lets take a look at the data from FireHose directly after download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9c0bda23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mirna shape: (503, 1189), rna shape: (18321, 1212), meth shape: (20107, 885), clinical shape: (18, 1097)\n",
      "              TCGA-3C-AAAU-01  TCGA-3C-AALI-01  TCGA-3C-AALJ-01  \\\n",
      "gene                                                              \n",
      "hsa-let-7a-1        13.129765        12.918069        13.012033   \n",
      "hsa-let-7a-2        14.117933        13.922300        14.010002   \n",
      "hsa-let-7a-3        13.147714        12.913194        13.028483   \n",
      "hsa-let-7b          14.595135        14.512657        13.419612   \n",
      "hsa-let-7c           8.414890         9.646536         9.312455   \n",
      "\n",
      "              TCGA-3C-AALK-01  TCGA-4H-AAAK-01  TCGA-5L-AAT0-01  \\\n",
      "gene                                                              \n",
      "hsa-let-7a-1        13.144697        13.411684        13.316301   \n",
      "hsa-let-7a-2        14.141721        14.413518        14.310917   \n",
      "hsa-let-7a-3        13.151281        13.420481        13.327144   \n",
      "hsa-let-7b          14.667196        14.438548        14.576493   \n",
      "hsa-let-7c          11.511431        11.693927        11.138419   \n",
      "\n",
      "              TCGA-5L-AAT1-01  TCGA-5T-A9QA-01  TCGA-A1-A0SB-01  \\\n",
      "gene                                                              \n",
      "hsa-let-7a-1        13.445230        13.727850        13.601504   \n",
      "hsa-let-7a-2        14.448556        14.714551        14.608693   \n",
      "hsa-let-7a-3        13.446806        13.736891        13.613105   \n",
      "hsa-let-7b          14.611137        15.098805        16.505758   \n",
      "hsa-let-7c          11.284446         9.197514        13.392164   \n",
      "\n",
      "              TCGA-A1-A0SD-01  ...  TCGA-BH-A0WA-01  TCGA-E2-A105-01  \\\n",
      "gene                           ...                                     \n",
      "hsa-let-7a-1        13.598739  ...        12.225132        13.938134   \n",
      "hsa-let-7a-2        14.606942  ...        13.235065        14.930021   \n",
      "hsa-let-7a-3        13.606224  ...        12.261971        13.972011   \n",
      "hsa-let-7b          15.638855  ...        14.684912        15.230457   \n",
      "hsa-let-7c          11.419823  ...        10.565698        10.483745   \n",
      "\n",
      "              TCGA-E2-A106-01  TCGA-E2-A107-01  TCGA-E2-A108-01  \\\n",
      "gene                                                              \n",
      "hsa-let-7a-1        13.609853        13.508290        13.406359   \n",
      "hsa-let-7a-2        14.603389        14.525026        14.402735   \n",
      "hsa-let-7a-3        13.643274        13.549981        13.438737   \n",
      "hsa-let-7b          15.357655        15.112011        15.040315   \n",
      "hsa-let-7c          11.159056        12.473340        12.405828   \n",
      "\n",
      "              TCGA-E2-A109-01  TCGA-E2-A10B-01  TCGA-E2-A10C-01  \\\n",
      "gene                                                              \n",
      "hsa-let-7a-1        13.730647        13.198426        12.793350   \n",
      "hsa-let-7a-2        14.719166        14.200523        13.796623   \n",
      "hsa-let-7a-3        13.732070        13.212367        12.793350   \n",
      "hsa-let-7b          15.806771        15.645910        14.724106   \n",
      "hsa-let-7c          10.613712        11.395452         9.087202   \n",
      "\n",
      "              TCGA-E2-A10E-01  TCGA-E2-A10F-01  \n",
      "gene                                            \n",
      "hsa-let-7a-1        14.060268        12.990403  \n",
      "hsa-let-7a-2        15.047592        14.006035  \n",
      "hsa-let-7a-3        14.074978        13.018659  \n",
      "hsa-let-7b          16.370741        15.439239  \n",
      "hsa-let-7c          10.885520        11.385638  \n",
      "\n",
      "[5 rows x 1189 columns]\n",
      "             TCGA-3C-AAAU-01  TCGA-3C-AALI-01  TCGA-3C-AALJ-01  \\\n",
      "gene                                                             \n",
      "?|100133144         4.032489         3.211931         3.538886   \n",
      "?|100134869         3.692829         4.119273         3.206237   \n",
      "?|10357             5.704604         6.124231         7.269570   \n",
      "?|10431             8.672694         9.139279        10.410275   \n",
      "?|155060           10.213110         9.011343         9.209506   \n",
      "\n",
      "             TCGA-3C-AALK-01  TCGA-4H-AAAK-01  TCGA-5L-AAT0-01  \\\n",
      "gene                                                             \n",
      "?|100133144         3.595671         2.775430         1.995991   \n",
      "?|100134869         3.469873         3.850979         3.766489   \n",
      "?|10357             7.168565         6.395968         6.836141   \n",
      "?|10431             9.757450         9.581922         9.657753   \n",
      "?|155060            9.110487         8.027083         8.110023   \n",
      "\n",
      "             TCGA-5L-AAT1-01  TCGA-5T-A9QA-01  TCGA-A1-A0SB-01  \\\n",
      "gene                                                             \n",
      "?|100133144              NaN         0.550310         3.939189   \n",
      "?|100134869         3.405298         3.169252         3.847346   \n",
      "?|10357             6.857961         6.749035         6.862786   \n",
      "?|10431            10.114256        10.472185         9.360367   \n",
      "?|155060            7.704865         6.254741         8.128052   \n",
      "\n",
      "             TCGA-A1-A0SD-01  ...  TCGA-UL-AAZ6-01  TCGA-UU-A93S-01  \\\n",
      "gene                          ...                                     \n",
      "?|100133144         3.250628  ...        -1.324816         2.108558   \n",
      "?|100134869         3.501324  ...         3.845189         3.443978   \n",
      "?|10357             5.913201  ...         7.083470         7.088829   \n",
      "?|10431             9.933569  ...        10.616682        11.495054   \n",
      "?|155060            6.387132  ...         8.052478         7.516236   \n",
      "\n",
      "             TCGA-V7-A7HQ-01  TCGA-W8-A86G-01  TCGA-WT-AB41-01  \\\n",
      "gene                                                             \n",
      "?|100133144              NaN         2.475707              NaN   \n",
      "?|100134869         1.622556         3.845099         2.657434   \n",
      "?|10357             4.906766         7.003547         5.744909   \n",
      "?|10431            10.749770         9.446410        10.282241   \n",
      "?|155060            9.280761         9.631306         8.137225   \n",
      "\n",
      "             TCGA-WT-AB44-01  TCGA-XX-A899-01  TCGA-XX-A89A-01  \\\n",
      "gene                                                             \n",
      "?|100133144              NaN         3.846574         4.480524   \n",
      "?|100134869         1.703987         4.422294         4.769476   \n",
      "?|10357             5.401368         7.106177         6.003213   \n",
      "?|10431            10.874534         9.350400         9.497295   \n",
      "?|155060            9.460539         8.738651         8.556414   \n",
      "\n",
      "             TCGA-Z7-A8R5-01  TCGA-Z7-A8R6-01  \n",
      "gene                                           \n",
      "?|100133144         1.178747         2.783771  \n",
      "?|100134869         2.866572         4.631075  \n",
      "?|10357             6.410173         7.388457  \n",
      "?|10431            10.155173         9.970921  \n",
      "?|155060            7.977670         7.894918  \n",
      "\n",
      "[5 rows x 1212 columns]\n",
      "                      TCGA-3C-AAAU-01 TCGA-3C-AALI-01 TCGA-3C-AALJ-01  \\\n",
      "Hybridization REF                                                       \n",
      "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
      "A1BG                   0.483716119676  0.637191226131  0.656092398242   \n",
      "A1CF                   0.295827203492  0.458972998571  0.489725289638   \n",
      "A2BP1                  0.187699869591  0.240515847704  0.279087851226   \n",
      "A2LD1                   0.62958551322  0.666272288675  0.755630499986   \n",
      "\n",
      "                      TCGA-3C-AALK-01 TCGA-4H-AAAK-01 TCGA-5L-AAT0-01  \\\n",
      "Hybridization REF                                                       \n",
      "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
      "A1BG                   0.615194471357  0.612080370511  0.469600740678   \n",
      "A1CF                   0.625765223243  0.507736509665  0.514770866326   \n",
      "A2BP1                  0.488888510474  0.463845494635  0.504450855353   \n",
      "A2LD1                   0.74575121287  0.698515739124  0.706812706661   \n",
      "\n",
      "                      TCGA-5L-AAT1-01 TCGA-5T-A9QA-01 TCGA-A1-A0SB-01  \\\n",
      "Hybridization REF                                                       \n",
      "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
      "A1BG                   0.582188239422   0.66617073097  0.659965611959   \n",
      "A1CF                   0.549850958729  0.381038654448  0.826312156393   \n",
      "A2BP1                  0.480885816745  0.622832399216  0.474678831563   \n",
      "A2LD1                  0.759017355996  0.694010939885  0.847837522256   \n",
      "\n",
      "                      TCGA-A1-A0SE-01  ... TCGA-UL-AAZ6-01 TCGA-UU-A93S-01  \\\n",
      "Hybridization REF                      ...                                   \n",
      "Composite Element REF      Beta_Value  ...      Beta_Value      Beta_Value   \n",
      "A1BG                   0.641701155202  ...  0.631413241724   0.64952294395   \n",
      "A1CF                   0.606699429409  ...  0.383469192855  0.183354853938   \n",
      "A2BP1                  0.339829506578  ...  0.130529915536  0.319855310743   \n",
      "A2LD1                  0.786662091353  ...  0.587475995313  0.667969642321   \n",
      "\n",
      "                      TCGA-V7-A7HQ-01 TCGA-W8-A86G-01 TCGA-WT-AB41-01  \\\n",
      "Hybridization REF                                                       \n",
      "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
      "A1BG                   0.596585169597  0.615558357651  0.580837880262   \n",
      "A1CF                   0.403909161312  0.716980255014  0.613131295074   \n",
      "A2BP1                  0.335517456053  0.512185396638  0.563519806811   \n",
      "A2LD1                  0.689140211036  0.791381283524  0.680499323148   \n",
      "\n",
      "                      TCGA-WT-AB44-01 TCGA-XX-A899-01 TCGA-XX-A89A-01  \\\n",
      "Hybridization REF                                                       \n",
      "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
      "A1BG                   0.615814023324  0.589897794957  0.572606636128   \n",
      "A1CF                   0.665043713213  0.705153725375  0.494848686021   \n",
      "A2BP1                  0.507364324635  0.520542747167  0.412562068574   \n",
      "A2LD1                  0.660476360054  0.745725420412   0.74390049875   \n",
      "\n",
      "                      TCGA-Z7-A8R5-01 TCGA-Z7-A8R6-01  \n",
      "Hybridization REF                                      \n",
      "Composite Element REF      Beta_Value      Beta_Value  \n",
      "A1BG                   0.617859586161  0.568150149265  \n",
      "A1CF                   0.691835387189  0.224696596211  \n",
      "A2BP1                  0.522169978143   0.33955834608  \n",
      "A2LD1                  0.791229999577  0.637764188841  \n",
      "\n",
      "[5 rows x 885 columns]\n",
      "                      tcga-5l-aat0 tcga-5l-aat1 tcga-a1-a0sp tcga-a2-a04v  \\\n",
      "Hybridization REF                                                           \n",
      "Composite Element REF        value        value        value        value   \n",
      "years_to_birth                  42           63           40           39   \n",
      "vital_status                     0            0            0            1   \n",
      "days_to_death                  NaN          NaN          NaN         1920   \n",
      "days_to_last_followup         1477         1471          584          NaN   \n",
      "\n",
      "                      tcga-a2-a04y tcga-a2-a0cq tcga-a2-a1g4 tcga-a2-a25a  \\\n",
      "Hybridization REF                                                           \n",
      "Composite Element REF        value        value        value        value   \n",
      "years_to_birth                  53           62           71           44   \n",
      "vital_status                     0            0            0            0   \n",
      "days_to_death                  NaN          NaN          NaN          NaN   \n",
      "days_to_last_followup         1099         2695          595         3276   \n",
      "\n",
      "                      tcga-a7-a0cd tcga-a7-a13g  ... tcga-s3-aa11  \\\n",
      "Hybridization REF                                ...                \n",
      "Composite Element REF        value        value  ...        value   \n",
      "years_to_birth                  66           79  ...           67   \n",
      "vital_status                     0            0  ...            0   \n",
      "days_to_death                  NaN          NaN  ...          NaN   \n",
      "days_to_last_followup         1165          718  ...          421   \n",
      "\n",
      "                      tcga-s3-aa14 tcga-s3-aa15 tcga-ul-aaz6 tcga-uu-a93s  \\\n",
      "Hybridization REF                                                           \n",
      "Composite Element REF        value        value        value        value   \n",
      "years_to_birth                  47           51           73           63   \n",
      "vital_status                     0            0            0            1   \n",
      "days_to_death                  NaN          NaN          NaN          116   \n",
      "days_to_last_followup          529          525          518          NaN   \n",
      "\n",
      "                      tcga-v7-a7hq tcga-wt-ab44 tcga-xx-a899 tcga-xx-a89a  \\\n",
      "Hybridization REF                                                           \n",
      "Composite Element REF        value        value        value        value   \n",
      "years_to_birth                  75          NaN           46           68   \n",
      "vital_status                     0            0            0            0   \n",
      "days_to_death                  NaN          NaN          NaN          NaN   \n",
      "days_to_last_followup         2033          883          467          488   \n",
      "\n",
      "                      tcga-z7-a8r6  \n",
      "Hybridization REF                   \n",
      "Composite Element REF        value  \n",
      "years_to_birth                  46  \n",
      "vital_status                     0  \n",
      "days_to_death                  NaN  \n",
      "days_to_last_followup         3256  \n",
      "\n",
      "[5 rows x 1097 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "root = Path(\"/home/vicente/Github/BioNeuralNet/TCGA_BRCA_DATA\")\n",
    "\n",
    "mirna_raw = pd.read_csv(root/\"BRCA.miRseq_RPKM_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)                            \n",
    "rna_raw = pd.read_csv(root / \"BRCA.uncv2.mRNAseq_RSEM_normalized_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)\n",
    "meth_raw = pd.read_csv(root/\"BRCA.meth.by_mean.data.txt\", sep='\\t',index_col=0,low_memory=False)\n",
    "clinical_raw = pd.read_csv(root / \"BRCA.clin.merged.picked.txt\",sep=\"\\t\", index_col=0, low_memory=False)\n",
    "\n",
    "print(f\"mirna shape: {mirna_raw.shape}, rna shape: {rna_raw.shape}, meth shape: {meth_raw.shape}, clinical shape: {clinical_raw.shape}\")\n",
    "print(mirna_raw.head())\n",
    "print(rna_raw.head())\n",
    "print(meth_raw.head())\n",
    "print(clinical_raw.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aacae339",
   "metadata": {},
   "source": [
    "## TCGA-BioLink\n",
    "\n",
    "This section demonstrates how to use the `TCGAbiolinks` R package to access and download clinical and molecular subtype data. It begins by ensuring `TCGAbiolinks` is installed, then loads the package. It retrieves PAM50 molecular subtype labels using `TCGAquery_subtype()` and writes them to a CSV file. Additionally, it downloads clinical data using `GDCquery_clinic()` and formats it with `GDCprepare_clinic()`, saving the result as another CSV file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a445601f",
   "metadata": {},
   "source": [
    "```R\n",
    "  # Install TCGAbiolinks\n",
    "  if (!requireNamespace(\"TCGAbiolinks\", quietly = TRUE)) {\n",
    "    if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "      install.packages(\"BiocManager\")\n",
    "    BiocManager::install(\"TCGAbiolinks\")\n",
    "  }\n",
    "\n",
    "  # Load the library\n",
    "  library(TCGAbiolinks)\n",
    "\n",
    "  # Download PAM50 subtype labels\n",
    "  pam50_df <- TCGAquery_subtype(tumor = \"BRCA\")[ , c(\"patient\", \"BRCA_Subtype_PAM50\")]\n",
    "  write.csv(pam50_df, file = \"BRCA_PAM50_labels.csv\", row.names = FALSE, quote = FALSE)\n",
    "\n",
    "  # Download clinical data\n",
    "  clin_raw <- GDCquery_clinic(project = \"TCGA-BRCA\", type = \"clinical\")\n",
    "  clin_df <- GDCprepare_clinic(clin_raw, clinical.info = \"patient\")\n",
    "  write.csv(clin_df, file = \"BRCA_clinical_data.csv\", row.names = FALSE, quote = FALSE)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "128f63dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial shapes\n",
      "meth: (20107, 885)\n",
      "rna: (18321, 1212)\n",
      "mirna: (503, 1189)\n",
      "pam50: (1087, 1)\n",
      "clinical TCGABioLinks: (1098, 101)\n",
      "clinical FireHose: (1097, 18)\n",
      "\n",
      "After tranpose\n",
      "meth: (885, 20107)\n",
      "rna: (1212, 18321)\n",
      "mirna: (1189, 503)\n",
      "Patients in both clinical datasets: 1097\n",
      "Combined Clinical shape (1097, 119)\n",
      "Patients in every dataset: 769\n",
      "\n",
      "Final shapes:\n",
      "meth: (863, 20107)\n",
      "rna: (865, 18321)\n",
      "mirna: (855, 503)\n",
      "pam50: (769, 1)\n",
      "clinical: (769, 119)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# from Firehose\n",
    "mirna = pd.read_csv(root/\"BRCA.miRseq_RPKM_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)\n",
    "meth = pd.read_csv(root/\"BRCA.meth.by_mean.data.txt\", sep='\\t',index_col=0,low_memory=False)                             \n",
    "rna = pd.read_csv(root / \"BRCA.uncv2.mRNAseq_RSEM_normalized_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)\n",
    "clinical_firehose = pd.read_csv(root / \"BRCA.clin.merged.picked.txt\",sep=\"\\t\", index_col=0, low_memory=False).T\n",
    "\n",
    "# from TCGABiolinks\n",
    "pam50 = pd.read_csv(root /\"BRCA_PAM50_labels.csv\",index_col=0)\n",
    "clinical_biolinks = pd.read_csv(root /\"BRCA_clinical_data.csv\",index_col=1)\n",
    "\n",
    "print(\"Initial shapes\")\n",
    "print(f\"meth: {meth.shape}\")\n",
    "print(f\"rna: {rna.shape}\")\n",
    "print(f\"mirna: {mirna.shape}\")\n",
    "print(f\"pam50: {pam50.shape}\")\n",
    "print(f\"clinical TCGABioLinks: {clinical_biolinks.shape}\")\n",
    "print(f\"clinical FireHose: {clinical_firehose.shape}\")\n",
    "\n",
    "meth = meth.T\n",
    "rna = rna.T\n",
    "mirna = mirna.T\n",
    "\n",
    "print(\"\\nAfter tranpose\")\n",
    "print(f\"meth: {meth.shape}\")\n",
    "print(f\"rna: {rna.shape}\")\n",
    "print(f\"mirna: {mirna.shape}\")\n",
    "\n",
    "def trim(idx):\n",
    "    return idx.to_series().str.extract(r'(^TCGA-\\w\\w-\\w\\w\\w\\w)')[0]\n",
    "\n",
    "meth.index = trim(meth.index)\n",
    "rna.index = trim(rna.index)\n",
    "mirna.index = trim(mirna.index)\n",
    "pam50.index = pam50.index.str.upper()\n",
    "clinical_biolinks.index = clinical_biolinks.index.str.upper()\n",
    "clinical_firehose.index = clinical_firehose.index.str.upper()\n",
    "\n",
    "idx1 = clinical_biolinks.index\n",
    "idx2 = clinical_firehose.index\n",
    "\n",
    "# intersection and unique counts\n",
    "common = idx1.intersection(idx2)\n",
    "only_in_1 = idx1.difference(idx2)\n",
    "only_in_2 = idx2.difference(idx1)\n",
    "\n",
    "print(f\"Patients in both clinical datasets: {len(common)}\")\n",
    "common = clinical_biolinks.index.intersection(clinical_firehose.index)\n",
    "clinical_biolinks = clinical_biolinks.loc[common]\n",
    "clinical_firehose = clinical_firehose.loc[common]\n",
    "\n",
    "clinical = pd.concat([clinical_biolinks, clinical_firehose], axis=1)\n",
    "\n",
    "print(f\"Combined Clinical shape {clinical.shape}\")\n",
    "\n",
    "common = sorted(set(meth.index) & set(rna.index) & set(mirna.index) & set(pam50.index) & set(clinical.index))\n",
    "print(f\"Patients in every dataset: {len(common)}\")\n",
    "\n",
    "meth = meth.loc[common]\n",
    "rna = rna.loc[common]\n",
    "mirna = mirna.loc[common]\n",
    "pam50 = pam50.loc[common]\n",
    "clinical = clinical.loc[common]\n",
    "\n",
    "print(\"\\nFinal shapes:\")\n",
    "print(f\"meth: {meth.shape}\")\n",
    "print(f\"rna: {rna.shape}\")\n",
    "print(f\"mirna: {mirna.shape}\")\n",
    "print(f\"pam50: {pam50.shape}\")\n",
    "print(f\"clinical: {clinical.shape}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ba4b2c",
   "metadata": {},
   "source": [
    "## Handling Multiple Aliquots per Sample\n",
    "\n",
    "This section addresses cases where some patients have multiple aliquots per sample in the `meth`, `rna`, and `mirna` datasets. It first identifies and counts patients with duplicate entries. Then, it coerces all data to numeric types and aggregates the duplicates by computing the mean across aliquots for each patient, ensuring only one row per patient. After aggregation, the datasets are aligned by keeping only the patients that are common across all five datasets (`meth`, `rna`, `mirna`, `pam50`, and `clinical`). The result is s set of matched samples ready for integrated analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b841497a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "meth:\n",
      "patients with >1 aliquot: 91\n",
      "total duplicate rows: 94\n",
      "\n",
      "rna:\n",
      "patients with >1 aliquot: 93\n",
      "total duplicate rows: 96\n",
      "\n",
      "mirna:\n",
      "patients with >1 aliquot: 84\n",
      "total duplicate rows: 86\n",
      "\n",
      "Post-aggregation shapes:\n",
      "meth: (769, 20107)\n",
      "rna: (769, 18321)\n",
      "mirna: (769, 503)\n",
      "Patients in every dataset: 769\n",
      "\n",
      "Final shapes\n",
      "meth: (769, 20107)\n",
      "rna: (769, 18321)\n",
      "mirna: (769, 503)\n",
      "pam50: (769, 1)\n",
      "clinical:(769, 119)\n"
     ]
    }
   ],
   "source": [
    "for name, df in [(\"meth\", meth), (\"rna\", rna), (\"mirna\", mirna)]:\n",
    "    counts = df.index.value_counts()\n",
    "    n_multiple = (counts > 1).sum()\n",
    "    total_duplicates = counts[counts > 1].sum() - n_multiple\n",
    "    \n",
    "    print(f\"{name}:\")\n",
    "    print(f\"patients with >1 aliquot: {n_multiple}\")\n",
    "    print(f\"total duplicate rows: {total_duplicates}\\n\")\n",
    "\n",
    "meth = meth.apply(pd.to_numeric, errors=\"coerce\")\n",
    "rna = rna .apply(pd.to_numeric, errors=\"coerce\")\n",
    "mirna = mirna.apply(pd.to_numeric, errors=\"coerce\")\n",
    "\n",
    "meth = meth.groupby(level=0).mean()\n",
    "rna = rna.groupby(level=0).mean()\n",
    "mirna = mirna.groupby(level=0).mean()\n",
    "\n",
    "# Now each has one row per patient\n",
    "print(\"Post-aggregation shapes:\")\n",
    "print(f\"meth: {meth.shape}\")\n",
    "print(f\"rna: {rna.shape}\")\n",
    "print(f\"mirna: {mirna.shape}\")\n",
    "\n",
    "common = sorted( set(meth.index) & set(rna.index) & set(mirna.index)& set(pam50.index) & set(clinical.index) )\n",
    "print(f\"Patients in every dataset: {len(common)}\")\n",
    "\n",
    "meth = meth.loc[common]\n",
    "rna = rna.loc[common]\n",
    "mirna = mirna.loc[common]\n",
    "pam50 = pam50.loc[common]\n",
    "clinical = clinical.loc[common]\n",
    "\n",
    "print(\"\\nFinal shapes\")\n",
    "print(f\"meth: {meth.shape}\")\n",
    "print(f\"rna: {rna.shape}\")\n",
    "print(f\"mirna: {mirna.shape}\")\n",
    "print(f\"pam50: {pam50.shape}\")\n",
    "print(f\"clinical:{clinical.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d8dac23",
   "metadata": {},
   "source": [
    "## Review the first few rows of each file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4f35bd67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hybridization REF  Composite Element REF      A1BG      A1CF     A2BP1  \\\n",
      "0                                                                        \n",
      "TCGA-3C-AAAU                         NaN  0.483716  0.295827  0.187700   \n",
      "TCGA-3C-AALI                         NaN  0.637191  0.458973  0.240516   \n",
      "TCGA-3C-AALJ                         NaN  0.656092  0.489725  0.279088   \n",
      "TCGA-3C-AALK                         NaN  0.615194  0.625765  0.488889   \n",
      "TCGA-4H-AAAK                         NaN  0.612080  0.507737  0.463845   \n",
      "\n",
      "Hybridization REF     A2LD1       A2M     A2ML1    A4GALT     A4GNT      AAA1  \\\n",
      "0                                                                               \n",
      "TCGA-3C-AAAU       0.629586  0.559654  0.835412  0.484800  0.690217  0.807805   \n",
      "TCGA-3C-AALI       0.666272  0.607505  0.842391  0.550047  0.749890  0.395290   \n",
      "TCGA-3C-AALJ       0.755630  0.662360  0.829020  0.476107  0.653756  0.795102   \n",
      "TCGA-3C-AALK       0.745751  0.727982  0.835365  0.556016  0.652005  0.816423   \n",
      "TCGA-4H-AAAK       0.698516  0.692364  0.802388  0.504870  0.531183  0.851114   \n",
      "\n",
      "Hybridization REF  ...    ZWILCH     ZWINT      ZXDC    ZYG11A    ZYG11B  \\\n",
      "0                  ...                                                     \n",
      "TCGA-3C-AAAU       ...  0.112978  0.053939  0.287665  0.328087  0.502935   \n",
      "TCGA-3C-AALI       ...  0.111834  0.046160  0.265322  0.405851  0.434024   \n",
      "TCGA-3C-AALJ       ...  0.113218  0.042657  0.272103  0.391326  0.449525   \n",
      "TCGA-3C-AALK       ...  0.145133  0.047022  0.301284  0.410348  0.446571   \n",
      "TCGA-4H-AAAK       ...  0.118928  0.045057  0.300647  0.379998  0.487929   \n",
      "\n",
      "Hybridization REF       ZYX     ZZEF1      ZZZ3  psiTPTE22      tAKR  \n",
      "0                                                                     \n",
      "TCGA-3C-AAAU       0.220683  0.482044  0.107396   0.247304  0.506404  \n",
      "TCGA-3C-AALI       0.236362  0.458847  0.119652   0.163022  0.623865  \n",
      "TCGA-3C-AALJ       0.210976  0.482641  0.102385   0.252328  0.504451  \n",
      "TCGA-3C-AALK       0.220185  0.485944  0.112941   0.471956  0.682468  \n",
      "TCGA-4H-AAAK       0.233324  0.490736  0.115646   0.314877  0.744877  \n",
      "\n",
      "[5 rows x 20107 columns]\n",
      "gene          ?|100133144  ?|100134869   ?|10357    ?|10431   ?|155060  \\\n",
      "0                                                                        \n",
      "TCGA-3C-AAAU     4.032489     3.692829  5.704604   8.672694  10.213110   \n",
      "TCGA-3C-AALI     3.211931     4.119273  6.124231   9.139279   9.011343   \n",
      "TCGA-3C-AALJ     3.538886     3.206237  7.269570  10.410275   9.209506   \n",
      "TCGA-3C-AALK     3.595671     3.469873  7.168565   9.757450   9.110487   \n",
      "TCGA-4H-AAAK     2.775430     3.850979  6.395968   9.581922   8.027083   \n",
      "\n",
      "gene           ?|26823  ?|340602  ?|388795  ?|390284  ?|391343  ...  \\\n",
      "0                                                               ...   \n",
      "TCGA-3C-AAAU       NaN  0.785174 -1.536587  2.048201       NaN  ...   \n",
      "TCGA-3C-AALI  0.121015  7.170928  2.291014  0.706022  3.027968  ...   \n",
      "TCGA-3C-AALJ       NaN       NaN  1.443554  1.443554       NaN  ...   \n",
      "TCGA-3C-AALK -1.273343       NaN  1.048724  2.186215       NaN  ...   \n",
      "TCGA-4H-AAAK -1.232769 -1.232769  1.574683  1.574683       NaN  ...   \n",
      "\n",
      "gene          ZWINT|11130  ZXDA|7789  ZXDB|158586  ZXDC|79364  ZYG11A|440590  \\\n",
      "0                                                                              \n",
      "TCGA-3C-AAAU     9.864120   7.017830     9.976968   10.695662       8.013988   \n",
      "TCGA-3C-AALI     9.914682   5.902438     8.809329   10.391374       7.632831   \n",
      "TCGA-3C-AALJ    11.305650   5.143969     9.060691    9.586488       8.374267   \n",
      "TCGA-3C-AALK     9.384994   5.782065     8.773906    9.754688       7.454703   \n",
      "TCGA-4H-AAAK     9.397606   5.612830     8.728789   10.035881       3.811738   \n",
      "\n",
      "gene          ZYG11B|79699   ZYX|7791  ZZEF1|23140  ZZZ3|26009  \\\n",
      "0                                                                \n",
      "TCGA-3C-AAAU     10.238851  11.776124    10.887932   10.205129   \n",
      "TCGA-3C-AALI      9.237422  12.426428    10.364848    8.667973   \n",
      "TCGA-3C-AALJ      9.055784  12.414355     9.880935    8.992994   \n",
      "TCGA-3C-AALK      9.246419  12.474556     9.609426    9.453001   \n",
      "TCGA-4H-AAAK      9.599438  11.980747     9.700292    9.784147   \n",
      "\n",
      "gene          psiTPTE22|387590  \n",
      "0                               \n",
      "TCGA-3C-AAAU          0.785174  \n",
      "TCGA-3C-AALI          9.855788  \n",
      "TCGA-3C-AALJ          5.143969  \n",
      "TCGA-3C-AALK          6.057699  \n",
      "TCGA-4H-AAAK          7.548699  \n",
      "\n",
      "[5 rows x 18321 columns]\n",
      "gene          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  \\\n",
      "0                                                                    \n",
      "TCGA-3C-AAAU     13.129765     14.117933     13.147714   14.595135   \n",
      "TCGA-3C-AALI     12.918069     13.922300     12.913194   14.512657   \n",
      "TCGA-3C-AALJ     13.012033     14.010002     13.028483   13.419612   \n",
      "TCGA-3C-AALK     13.144697     14.141721     13.151281   14.667196   \n",
      "TCGA-4H-AAAK     13.411684     14.413518     13.420481   14.438548   \n",
      "\n",
      "gene          hsa-let-7c  hsa-let-7d  hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  \\\n",
      "0                                                                              \n",
      "TCGA-3C-AAAU    8.414890    8.665921   10.521777      3.879392     11.824817   \n",
      "TCGA-3C-AALI    9.646536    9.003653    9.131760      4.386952     12.678841   \n",
      "TCGA-3C-AALJ    9.312455    9.276943   11.395711      5.314692     13.530255   \n",
      "TCGA-3C-AALK   11.511431    8.384763   10.368981      4.159182     12.652559   \n",
      "TCGA-4H-AAAK   11.693927    8.453747   10.741371      4.494537     13.009499   \n",
      "\n",
      "gene          hsa-let-7g  ...  hsa-mir-937  hsa-mir-939  hsa-mir-940  \\\n",
      "0                         ...                                          \n",
      "TCGA-3C-AAAU    8.597744  ...     0.906699    -0.093302     2.672234   \n",
      "TCGA-3C-AALI    8.455144  ...     1.579597    -0.083367     0.139024   \n",
      "TCGA-3C-AALJ    9.230563  ...     3.270298    -2.189134     0.395828   \n",
      "TCGA-3C-AALK    8.471503  ...     0.923965    -0.660997    -0.076034   \n",
      "TCGA-4H-AAAK    8.381220  ...     0.182950    -0.624403    -1.624403   \n",
      "\n",
      "gene          hsa-mir-942  hsa-mir-944  hsa-mir-95  hsa-mir-96  hsa-mir-98  \\\n",
      "0                                                                            \n",
      "TCGA-3C-AAAU     2.467414     1.044202    2.044202    6.906699    5.754696   \n",
      "TCGA-3C-AALI     3.032109    -0.668331    0.331670    5.912870    6.427066   \n",
      "TCGA-3C-AALJ     1.855261    -0.381778    0.717757    6.603657    6.878301   \n",
      "TCGA-3C-AALK     1.798435     1.798435    0.798435    6.181354    5.377922   \n",
      "TCGA-4H-AAAK     1.076036     0.182950   -0.302475    4.318110    5.103516   \n",
      "\n",
      "gene          hsa-mir-99a  hsa-mir-99b  \n",
      "0                                       \n",
      "TCGA-3C-AAAU     7.024602    15.506461  \n",
      "TCGA-3C-AALI     7.885299    13.626182  \n",
      "TCGA-3C-AALJ     7.580704    15.013822  \n",
      "TCGA-3C-AALK    10.031619    14.554783  \n",
      "TCGA-4H-AAAK    10.078201    14.650338  \n",
      "\n",
      "[5 rows x 503 columns]\n",
      "                project synchronous_malignancy ajcc_pathologic_stage  \\\n",
      "TCGA-3C-AAAU  TCGA-BRCA                     No               Stage X   \n",
      "TCGA-3C-AALI  TCGA-BRCA                     No             Stage IIB   \n",
      "TCGA-3C-AALJ  TCGA-BRCA                     No             Stage IIB   \n",
      "TCGA-3C-AALK  TCGA-BRCA                     No              Stage IA   \n",
      "TCGA-4H-AAAK  TCGA-BRCA                     No            Stage IIIA   \n",
      "\n",
      "              days_to_diagnosis laterality  created_datetime  \\\n",
      "TCGA-3C-AAAU                0.0       Left               NaN   \n",
      "TCGA-3C-AALI                0.0      Right               NaN   \n",
      "TCGA-3C-AALJ                0.0      Right               NaN   \n",
      "TCGA-3C-AALK                0.0      Right               NaN   \n",
      "TCGA-4H-AAAK                0.0       Left               NaN   \n",
      "\n",
      "             last_known_disease_status tissue_or_organ_of_origin  \\\n",
      "TCGA-3C-AAAU                       NaN               Breast, NOS   \n",
      "TCGA-3C-AALI                       NaN               Breast, NOS   \n",
      "TCGA-3C-AALJ                       NaN               Breast, NOS   \n",
      "TCGA-3C-AALK                       NaN               Breast, NOS   \n",
      "TCGA-4H-AAAK                       NaN               Breast, NOS   \n",
      "\n",
      "              days_to_last_follow_up  age_at_diagnosis  ... pathology_N_stage  \\\n",
      "TCGA-3C-AAAU                     NaN           20211.0  ...                nx   \n",
      "TCGA-3C-AALI                     NaN           18538.0  ...               n1a   \n",
      "TCGA-3C-AALJ                     NaN           22848.0  ...               n1a   \n",
      "TCGA-3C-AALK                     NaN           19074.0  ...           n0 (i+)   \n",
      "TCGA-4H-AAAK                     NaN           18371.0  ...               n2a   \n",
      "\n",
      "             pathology_M_stage  gender date_of_initial_pathologic_diagnosis  \\\n",
      "TCGA-3C-AAAU                mx  female                                 2004   \n",
      "TCGA-3C-AALI                m0  female                                 2003   \n",
      "TCGA-3C-AALJ                m0  female                                 2011   \n",
      "TCGA-3C-AALK                m0  female                                 2011   \n",
      "TCGA-4H-AAAK                m0  female                                 2013   \n",
      "\n",
      "              days_to_last_known_alive radiation_therapy  \\\n",
      "TCGA-3C-AAAU                       NaN                no   \n",
      "TCGA-3C-AALI                       NaN               yes   \n",
      "TCGA-3C-AALJ                       NaN                no   \n",
      "TCGA-3C-AALK                       NaN                no   \n",
      "TCGA-4H-AAAK                       NaN                no   \n",
      "\n",
      "                           histological_type number_of_lymph_nodes  \\\n",
      "TCGA-3C-AAAU  infiltrating lobular carcinoma                     4   \n",
      "TCGA-3C-AALI   infiltrating ductal carcinoma                     1   \n",
      "TCGA-3C-AALJ   infiltrating ductal carcinoma                     1   \n",
      "TCGA-3C-AALK   infiltrating ductal carcinoma                     0   \n",
      "TCGA-4H-AAAK  infiltrating lobular carcinoma                     4   \n",
      "\n",
      "                                   race               ethnicity  \n",
      "TCGA-3C-AAAU                      white  not hispanic or latino  \n",
      "TCGA-3C-AALI  black or african american  not hispanic or latino  \n",
      "TCGA-3C-AALJ  black or african american  not hispanic or latino  \n",
      "TCGA-3C-AALK  black or african american  not hispanic or latino  \n",
      "TCGA-4H-AAAK                      white  not hispanic or latino  \n",
      "\n",
      "[5 rows x 119 columns]\n",
      "BRCA_Subtype_PAM50\n",
      "LumA                  419\n",
      "LumB                  140\n",
      "Basal                 130\n",
      "Her2                   46\n",
      "Normal                 34\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(meth.head())\n",
    "print(rna.head())\n",
    "print(mirna.head())\n",
    "print(clinical.head())\n",
    "print(pam50.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17f7d599",
   "metadata": {},
   "source": [
    "## Preprocessing\n",
    "\n",
    "After reviewing the data above, we applied the following steps to the data before further analysis.\n",
    "\n",
    "1. Methylation (B -> M-value)\n",
    "   - Clip B-values to \\[E, 1-E] and apply logit transform: M = log_2(B / (1-B)).\n",
    "   - Drop the original `Composite Element REF` column.\n",
    "\n",
    "2. mRNA & miRNA:\n",
    "   - Already in log_2 scale (RSEM normalized and RPKM).\n",
    "\n",
    "3. Quality Control:\n",
    "   - Count samples with all-zero rows in each modality.\n",
    "   - Compute NaN counts post-transformation, then replace all NaNs with 0.\n",
    "\n",
    "4. Column Name Cleaning:\n",
    "   - Replace all `-` and `|` characters with `_`.\n",
    "   - Replace `?` with `unknown`.\n",
    "\n",
    "5. Label Encoding:\n",
    "   - Map `PAM50` subtypes to integers: \n",
    "      - Normal = 0\n",
    "      - Basal = 1 \n",
    "      - Her2 = 2\n",
    "      - LumA = 3\n",
    "      - LumB = 4\n",
    "\n",
    "6. Alignment & Aggregation:\n",
    "   - Trim barcodes to patient level.\n",
    "   - Aggregate duplicate aliquots by mean per patient.\n",
    "   - Drop the `project` column from clinical.\n",
    "   - Subset all tables to the common patient set (no missing or all-zero samples).\n",
    "   - Set up a commong index across all files.\n",
    "\n",
    "7. Final Output Shapes:\n",
    "   - Methylation M-value: 769 × 20,107\n",
    "   - mRNA (log_2): 769 × 20,531\n",
    "   - miRNA (log_2): 769 × 503\n",
    "   - PAM50 labels: 769 × 1\n",
    "   - Clinical covariates: 769 × 101"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5bb6450e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All zeros: meth: 0, rna: 0, mirna: 0\n",
      "nan_meth: 0, nan_rna: 0, nan_mirna: 0, nan_clinical: 0, nan_pam50: 0\n",
      "NaN counts after filling:\n",
      "0 0 0 46476 0\n",
      "new shapes: meth: (769, 20106), rna: (769, 18321), mirna: (769, 503), pam50: (769, 1), clinical: (769, 118)\n",
      "Hybridization REF      A1BG      A1CF     A2BP1     A2LD1       A2M     A2ML1  \\\n",
      "patient                                                                         \n",
      "TCGA-3C-AAAU      -0.094004 -1.251175 -2.113585  0.765262  0.345896  2.343631   \n",
      "TCGA-3C-AALI       0.812517 -0.237291 -1.658888  0.997440  0.630221  2.418135   \n",
      "TCGA-3C-AALJ       0.931878 -0.059301 -1.369104  1.628617  0.972130  2.277584   \n",
      "TCGA-3C-AALK       0.676913  0.741678 -0.064133  1.552454  1.420200  2.343133   \n",
      "TCGA-4H-AAAK       0.657963  0.044649 -0.209004  1.212210  1.170304  2.021628   \n",
      "\n",
      "Hybridization REF    A4GALT     A4GNT      AAA1      AAAS  ...    ZWILCH  \\\n",
      "patient                                                    ...             \n",
      "TCGA-3C-AAAU      -0.087741  1.155791  2.071436 -2.650851  ... -2.972923   \n",
      "TCGA-3C-AALI       0.289780  1.584114 -0.613329 -4.072465  ... -2.989465   \n",
      "TCGA-3C-AALJ      -0.137988  0.916964  1.956230 -3.781647  ... -2.969472   \n",
      "TCGA-3C-AALK       0.324621  0.905816  2.152928 -3.894574  ... -2.558319   \n",
      "TCGA-4H-AAAK       0.028103  0.180184  2.515149 -3.885526  ... -2.889175   \n",
      "\n",
      "Hybridization REF     ZWINT      ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1  \\\n",
      "patient                                                                         \n",
      "TCGA-3C-AAAU      -4.132523 -1.308165 -1.034199  0.016935 -1.820233 -0.103662   \n",
      "TCGA-3C-AALI      -4.369032 -1.469365 -0.549876 -0.382967 -1.691887 -0.238022   \n",
      "TCGA-3C-AALJ      -4.488190 -1.419578 -0.637297 -0.292273 -1.902991 -0.100215   \n",
      "TCGA-3C-AALK      -4.341028 -1.213585 -0.523013 -0.309506 -1.824419 -0.081137   \n",
      "TCGA-4H-AAAK      -4.405580 -1.217950 -0.706284 -0.069670 -1.716283 -0.053464   \n",
      "\n",
      "Hybridization REF      ZZZ3  psiTPTE22      tAKR  \n",
      "patient                                           \n",
      "TCGA-3C-AAAU      -3.055084  -1.605783  0.036955  \n",
      "TCGA-3C-AALI      -2.879231  -2.360128  0.729981  \n",
      "TCGA-3C-AALJ      -3.132087  -1.567104  0.025686  \n",
      "TCGA-3C-AALK      -2.973455  -0.162004  1.103860  \n",
      "TCGA-4H-AAAK      -2.934908  -1.121575  1.545812  \n",
      "\n",
      "[5 rows x 20106 columns]\n",
      "gene          unknown_100133144  unknown_100134869  unknown_10357  \\\n",
      "patient                                                             \n",
      "TCGA-3C-AAAU           4.032489           3.692829       5.704604   \n",
      "TCGA-3C-AALI           3.211931           4.119273       6.124231   \n",
      "TCGA-3C-AALJ           3.538886           3.206237       7.269570   \n",
      "TCGA-3C-AALK           3.595671           3.469873       7.168565   \n",
      "TCGA-4H-AAAK           2.775430           3.850979       6.395968   \n",
      "\n",
      "gene          unknown_10431  unknown_155060  unknown_26823  unknown_340602  \\\n",
      "patient                                                                      \n",
      "TCGA-3C-AAAU       8.672694       10.213110       0.000000        0.785174   \n",
      "TCGA-3C-AALI       9.139279        9.011343       0.121015        7.170928   \n",
      "TCGA-3C-AALJ      10.410275        9.209506       0.000000        0.000000   \n",
      "TCGA-3C-AALK       9.757450        9.110487      -1.273343        0.000000   \n",
      "TCGA-4H-AAAK       9.581922        8.027083      -1.232769       -1.232769   \n",
      "\n",
      "gene          unknown_388795  unknown_390284  unknown_391343  ...  \\\n",
      "patient                                                       ...   \n",
      "TCGA-3C-AAAU       -1.536587        2.048201        0.000000  ...   \n",
      "TCGA-3C-AALI        2.291014        0.706022        3.027968  ...   \n",
      "TCGA-3C-AALJ        1.443554        1.443554        0.000000  ...   \n",
      "TCGA-3C-AALK        1.048724        2.186215        0.000000  ...   \n",
      "TCGA-4H-AAAK        1.574683        1.574683        0.000000  ...   \n",
      "\n",
      "gene          ZWINT_11130  ZXDA_7789  ZXDB_158586  ZXDC_79364  ZYG11A_440590  \\\n",
      "patient                                                                        \n",
      "TCGA-3C-AAAU     9.864120   7.017830     9.976968   10.695662       8.013988   \n",
      "TCGA-3C-AALI     9.914682   5.902438     8.809329   10.391374       7.632831   \n",
      "TCGA-3C-AALJ    11.305650   5.143969     9.060691    9.586488       8.374267   \n",
      "TCGA-3C-AALK     9.384994   5.782065     8.773906    9.754688       7.454703   \n",
      "TCGA-4H-AAAK     9.397606   5.612830     8.728789   10.035881       3.811738   \n",
      "\n",
      "gene          ZYG11B_79699   ZYX_7791  ZZEF1_23140  ZZZ3_26009  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU     10.238851  11.776124    10.887932   10.205129   \n",
      "TCGA-3C-AALI      9.237422  12.426428    10.364848    8.667973   \n",
      "TCGA-3C-AALJ      9.055784  12.414355     9.880935    8.992994   \n",
      "TCGA-3C-AALK      9.246419  12.474556     9.609426    9.453001   \n",
      "TCGA-4H-AAAK      9.599438  11.980747     9.700292    9.784147   \n",
      "\n",
      "gene          psiTPTE22_387590  \n",
      "patient                         \n",
      "TCGA-3C-AAAU          0.785174  \n",
      "TCGA-3C-AALI          9.855788  \n",
      "TCGA-3C-AALJ          5.143969  \n",
      "TCGA-3C-AALK          6.057699  \n",
      "TCGA-4H-AAAK          7.548699  \n",
      "\n",
      "[5 rows x 18321 columns]\n",
      "gene          hsa_let_7a_1  hsa_let_7a_2  hsa_let_7a_3  hsa_let_7b  \\\n",
      "patient                                                              \n",
      "TCGA-3C-AAAU     13.129765     14.117933     13.147714   14.595135   \n",
      "TCGA-3C-AALI     12.918069     13.922300     12.913194   14.512657   \n",
      "TCGA-3C-AALJ     13.012033     14.010002     13.028483   13.419612   \n",
      "TCGA-3C-AALK     13.144697     14.141721     13.151281   14.667196   \n",
      "TCGA-4H-AAAK     13.411684     14.413518     13.420481   14.438548   \n",
      "\n",
      "gene          hsa_let_7c  hsa_let_7d  hsa_let_7e  hsa_let_7f_1  hsa_let_7f_2  \\\n",
      "patient                                                                        \n",
      "TCGA-3C-AAAU    8.414890    8.665921   10.521777      3.879392     11.824817   \n",
      "TCGA-3C-AALI    9.646536    9.003653    9.131760      4.386952     12.678841   \n",
      "TCGA-3C-AALJ    9.312455    9.276943   11.395711      5.314692     13.530255   \n",
      "TCGA-3C-AALK   11.511431    8.384763   10.368981      4.159182     12.652559   \n",
      "TCGA-4H-AAAK   11.693927    8.453747   10.741371      4.494537     13.009499   \n",
      "\n",
      "gene          hsa_let_7g  ...  hsa_mir_937  hsa_mir_939  hsa_mir_940  \\\n",
      "patient                   ...                                          \n",
      "TCGA-3C-AAAU    8.597744  ...     0.906699    -0.093302     2.672234   \n",
      "TCGA-3C-AALI    8.455144  ...     1.579597    -0.083367     0.139024   \n",
      "TCGA-3C-AALJ    9.230563  ...     3.270298    -2.189134     0.395828   \n",
      "TCGA-3C-AALK    8.471503  ...     0.923965    -0.660997    -0.076034   \n",
      "TCGA-4H-AAAK    8.381220  ...     0.182950    -0.624403    -1.624403   \n",
      "\n",
      "gene          hsa_mir_942  hsa_mir_944  hsa_mir_95  hsa_mir_96  hsa_mir_98  \\\n",
      "patient                                                                      \n",
      "TCGA-3C-AAAU     2.467414     1.044202    2.044202    6.906699    5.754696   \n",
      "TCGA-3C-AALI     3.032109    -0.668331    0.331670    5.912870    6.427066   \n",
      "TCGA-3C-AALJ     1.855261    -0.381778    0.717757    6.603657    6.878301   \n",
      "TCGA-3C-AALK     1.798435     1.798435    0.798435    6.181354    5.377922   \n",
      "TCGA-4H-AAAK     1.076036     0.182950   -0.302475    4.318110    5.103516   \n",
      "\n",
      "gene          hsa_mir_99a  hsa_mir_99b  \n",
      "patient                                 \n",
      "TCGA-3C-AAAU     7.024602    15.506461  \n",
      "TCGA-3C-AALI     7.885299    13.626182  \n",
      "TCGA-3C-AALJ     7.580704    15.013822  \n",
      "TCGA-3C-AALK    10.031619    14.554783  \n",
      "TCGA-4H-AAAK    10.078201    14.650338  \n",
      "\n",
      "[5 rows x 503 columns]\n",
      "             synchronous_malignancy ajcc_pathologic_stage  days_to_diagnosis  \\\n",
      "patient                                                                        \n",
      "TCGA-3C-AAAU                     No               Stage X                0.0   \n",
      "TCGA-3C-AALI                     No             Stage IIB                0.0   \n",
      "TCGA-3C-AALJ                     No             Stage IIB                0.0   \n",
      "TCGA-3C-AALK                     No              Stage IA                0.0   \n",
      "TCGA-4H-AAAK                     No            Stage IIIA                0.0   \n",
      "\n",
      "             laterality  created_datetime last_known_disease_status  \\\n",
      "patient                                                               \n",
      "TCGA-3C-AAAU       Left               NaN                       NaN   \n",
      "TCGA-3C-AALI      Right               NaN                       NaN   \n",
      "TCGA-3C-AALJ      Right               NaN                       NaN   \n",
      "TCGA-3C-AALK      Right               NaN                       NaN   \n",
      "TCGA-4H-AAAK       Left               NaN                       NaN   \n",
      "\n",
      "             tissue_or_organ_of_origin  days_to_last_follow_up  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU               Breast, NOS                     NaN   \n",
      "TCGA-3C-AALI               Breast, NOS                     NaN   \n",
      "TCGA-3C-AALJ               Breast, NOS                     NaN   \n",
      "TCGA-3C-AALK               Breast, NOS                     NaN   \n",
      "TCGA-4H-AAAK               Breast, NOS                     NaN   \n",
      "\n",
      "              age_at_diagnosis                 primary_diagnosis  ...  \\\n",
      "patient                                                           ...   \n",
      "TCGA-3C-AAAU           20211.0            Lobular carcinoma, NOS  ...   \n",
      "TCGA-3C-AALI           18538.0  Infiltrating duct carcinoma, NOS  ...   \n",
      "TCGA-3C-AALJ           22848.0  Infiltrating duct carcinoma, NOS  ...   \n",
      "TCGA-3C-AALK           19074.0  Infiltrating duct carcinoma, NOS  ...   \n",
      "TCGA-4H-AAAK           18371.0            Lobular carcinoma, NOS  ...   \n",
      "\n",
      "             pathology_N_stage pathology_M_stage  gender  \\\n",
      "patient                                                    \n",
      "TCGA-3C-AAAU                nx                mx  female   \n",
      "TCGA-3C-AALI               n1a                m0  female   \n",
      "TCGA-3C-AALJ               n1a                m0  female   \n",
      "TCGA-3C-AALK           n0 (i+)                m0  female   \n",
      "TCGA-4H-AAAK               n2a                m0  female   \n",
      "\n",
      "              date_of_initial_pathologic_diagnosis days_to_last_known_alive  \\\n",
      "patient                                                                       \n",
      "TCGA-3C-AAAU                                  2004                      NaN   \n",
      "TCGA-3C-AALI                                  2003                      NaN   \n",
      "TCGA-3C-AALJ                                  2011                      NaN   \n",
      "TCGA-3C-AALK                                  2011                      NaN   \n",
      "TCGA-4H-AAAK                                  2013                      NaN   \n",
      "\n",
      "             radiation_therapy               histological_type  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU                no  infiltrating lobular carcinoma   \n",
      "TCGA-3C-AALI               yes   infiltrating ductal carcinoma   \n",
      "TCGA-3C-AALJ                no   infiltrating ductal carcinoma   \n",
      "TCGA-3C-AALK                no   infiltrating ductal carcinoma   \n",
      "TCGA-4H-AAAK                no  infiltrating lobular carcinoma   \n",
      "\n",
      "              number_of_lymph_nodes                       race  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU                      4                      white   \n",
      "TCGA-3C-AALI                      1  black or african american   \n",
      "TCGA-3C-AALJ                      1  black or african american   \n",
      "TCGA-3C-AALK                      0  black or african american   \n",
      "TCGA-4H-AAAK                      4                      white   \n",
      "\n",
      "                           ethnicity  \n",
      "patient                               \n",
      "TCGA-3C-AAAU  not hispanic or latino  \n",
      "TCGA-3C-AALI  not hispanic or latino  \n",
      "TCGA-3C-AALJ  not hispanic or latino  \n",
      "TCGA-3C-AALK  not hispanic or latino  \n",
      "TCGA-4H-AAAK  not hispanic or latino  \n",
      "\n",
      "[5 rows x 118 columns]\n",
      "pam50\n",
      "3        419\n",
      "4        140\n",
      "1        130\n",
      "2         46\n",
      "0         34\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def beta_to_m(df, eps=1e-6):\n",
    "    B = np.clip(df.values, eps, 1.0 - eps)\n",
    "    M = np.log2(B / (1 - B))\n",
    "    return pd.DataFrame(M, index=df.index, columns=df.columns)\n",
    "\n",
    "# find rows that are all 0s\n",
    "zeros_meth = (meth  == 0).all(axis=1).sum()\n",
    "zeros_rna = (rna   == 0).all(axis=1).sum()\n",
    "zeros_mirna = (mirna == 0).all(axis=1).sum()\n",
    "print(f\"All zeros: meth: {zeros_meth}, rna: {zeros_rna}, mirna: {zeros_mirna}\")\n",
    "\n",
    "# find rows with all nans\n",
    "nan_meth = meth.isna().all(axis=1).sum()\n",
    "nan_rna = rna.isna().all(axis=1).sum()\n",
    "nan_mirna = mirna.isna().all(axis=1).sum()\n",
    "nan_clinical = clinical.isna().all(axis=1).sum()\n",
    "nan_pam50 = pam50.isna().all(axis=1).sum()\n",
    "print(f\"nan_meth: {nan_meth}, nan_rna: {nan_rna}, nan_mirna: {nan_mirna}, nan_clinical: {nan_clinical}, nan_pam50: {nan_pam50}\")\n",
    "\n",
    "# map PAM50 subtypes to integers\n",
    "mapping = {\"Normal\":0, \"Basal\":1, \"Her2\":2, \"LumA\":3, \"LumB\":4}\n",
    "pam50 = pam50[\"BRCA_Subtype_PAM50\"].map(mapping).to_frame(name=\"pam50\")\n",
    "\n",
    "# drop and transform methylation\n",
    "meth_clean = meth.drop(columns=[\"Composite Element REF\"], errors=\"ignore\")\n",
    "meth_m = beta_to_m(meth_clean)\n",
    "clinical = clinical.drop(columns=[\"project\"], errors=\"ignore\")\n",
    "\n",
    "# clean column names and fill nans\n",
    "for df in [meth_m, rna, mirna]:\n",
    "    df.columns = df.columns.str.replace(r\"\\?\", \"unknown_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(r\"\\|\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(\"-\", \"_\", regex=False)\n",
    "    df.columns = df.columns.str.replace(r\"_+\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.strip(\"_\")\n",
    "    df.fillna(0, inplace=True)\n",
    "\n",
    "# check for nans after filling\n",
    "print(\"NaN counts after filling:\")\n",
    "print(meth_m.isna().sum().sum(),rna.isna().sum().sum(),mirna.isna().sum().sum(),clinical.isna().sum().sum(),pam50.isna().sum().sum())\n",
    "\n",
    "# align index to PAM50\n",
    "X_meth = meth_m.loc[pam50.index]\n",
    "X_rna = rna.loc[pam50.index]\n",
    "X_mirna = mirna.loc[pam50.index]\n",
    "clinical= clinical.loc[pam50.index]\n",
    "\n",
    "print(f\"new shapes: meth: {X_meth.shape}, rna: {X_rna.shape}, mirna: {X_mirna.shape}, pam50: {pam50.shape}, clinical: {clinical.shape}\")\n",
    "print(X_meth.head())\n",
    "print(X_rna.head())\n",
    "print(X_mirna.head())\n",
    "print(clinical.head())\n",
    "print(pam50.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2f0714e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting up a commong index and saving to csv\n",
    "X_meth.index.name = \"patient\"\n",
    "X_rna.index.name = \"patient\"\n",
    "X_mirna.index.name = \"patient\"\n",
    "pam50.index.name = \"patient\"\n",
    "clinical.index.name = \"patient\"\n",
    "\n",
    "X_meth.to_csv(root / \"meth.csv\", index=True)\n",
    "X_rna.to_csv(root / \"rna.csv\", index=True)\n",
    "X_mirna.to_csv(root / \"mirna.csv\", index=True)\n",
    "pam50.to_csv(root / \"pam50.csv\", index=True)\n",
    "clinical.to_csv(root / \"clinical.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ef2982ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  A1BG      A1CF     A2BP1     A2LD1       A2M     A2ML1  \\\n",
      "patient                                                                    \n",
      "TCGA-3C-AAAU -0.094004 -1.251175 -2.113585  0.765262  0.345896  2.343631   \n",
      "TCGA-3C-AALI  0.812517 -0.237291 -1.658888  0.997440  0.630221  2.418135   \n",
      "TCGA-3C-AALJ  0.931878 -0.059301 -1.369104  1.628617  0.972130  2.277584   \n",
      "TCGA-3C-AALK  0.676913  0.741678 -0.064133  1.552454  1.420200  2.343133   \n",
      "TCGA-4H-AAAK  0.657963  0.044649 -0.209004  1.212210  1.170304  2.021628   \n",
      "\n",
      "                A4GALT     A4GNT      AAA1      AAAS  ...    ZWILCH     ZWINT  \\\n",
      "patient                                               ...                       \n",
      "TCGA-3C-AAAU -0.087741  1.155791  2.071436 -2.650851  ... -2.972923 -4.132523   \n",
      "TCGA-3C-AALI  0.289780  1.584114 -0.613329 -4.072465  ... -2.989465 -4.369032   \n",
      "TCGA-3C-AALJ -0.137988  0.916964  1.956230 -3.781647  ... -2.969472 -4.488190   \n",
      "TCGA-3C-AALK  0.324621  0.905816  2.152928 -3.894574  ... -2.558319 -4.341028   \n",
      "TCGA-4H-AAAK  0.028103  0.180184  2.515149 -3.885526  ... -2.889175 -4.405580   \n",
      "\n",
      "                  ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \\\n",
      "patient                                                                    \n",
      "TCGA-3C-AAAU -1.308165 -1.034199  0.016935 -1.820233 -0.103662 -3.055084   \n",
      "TCGA-3C-AALI -1.469365 -0.549876 -0.382967 -1.691887 -0.238022 -2.879231   \n",
      "TCGA-3C-AALJ -1.419578 -0.637297 -0.292273 -1.902991 -0.100215 -3.132087   \n",
      "TCGA-3C-AALK -1.213585 -0.523013 -0.309506 -1.824419 -0.081137 -2.973455   \n",
      "TCGA-4H-AAAK -1.217950 -0.706284 -0.069670 -1.716283 -0.053464 -2.934908   \n",
      "\n",
      "              psiTPTE22      tAKR  \n",
      "patient                            \n",
      "TCGA-3C-AAAU  -1.605783  0.036955  \n",
      "TCGA-3C-AALI  -2.360128  0.729981  \n",
      "TCGA-3C-AALJ  -1.567104  0.025686  \n",
      "TCGA-3C-AALK  -0.162004  1.103860  \n",
      "TCGA-4H-AAAK  -1.121575  1.545812  \n",
      "\n",
      "[5 rows x 20106 columns]\n",
      "              unknown_100133144  unknown_100134869  unknown_10357  \\\n",
      "patient                                                             \n",
      "TCGA-3C-AAAU           4.032489           3.692829       5.704604   \n",
      "TCGA-3C-AALI           3.211931           4.119273       6.124231   \n",
      "TCGA-3C-AALJ           3.538886           3.206237       7.269570   \n",
      "TCGA-3C-AALK           3.595671           3.469873       7.168565   \n",
      "TCGA-4H-AAAK           2.775430           3.850979       6.395968   \n",
      "\n",
      "              unknown_10431  unknown_155060  unknown_26823  unknown_340602  \\\n",
      "patient                                                                      \n",
      "TCGA-3C-AAAU       8.672694       10.213110       0.000000        0.785174   \n",
      "TCGA-3C-AALI       9.139279        9.011343       0.121015        7.170928   \n",
      "TCGA-3C-AALJ      10.410275        9.209506       0.000000        0.000000   \n",
      "TCGA-3C-AALK       9.757450        9.110487      -1.273343        0.000000   \n",
      "TCGA-4H-AAAK       9.581922        8.027083      -1.232769       -1.232769   \n",
      "\n",
      "              unknown_388795  unknown_390284  unknown_391343  ...  \\\n",
      "patient                                                       ...   \n",
      "TCGA-3C-AAAU       -1.536587        2.048201        0.000000  ...   \n",
      "TCGA-3C-AALI        2.291014        0.706022        3.027968  ...   \n",
      "TCGA-3C-AALJ        1.443554        1.443554        0.000000  ...   \n",
      "TCGA-3C-AALK        1.048724        2.186215        0.000000  ...   \n",
      "TCGA-4H-AAAK        1.574683        1.574683        0.000000  ...   \n",
      "\n",
      "              ZWINT_11130  ZXDA_7789  ZXDB_158586  ZXDC_79364  ZYG11A_440590  \\\n",
      "patient                                                                        \n",
      "TCGA-3C-AAAU     9.864120   7.017830     9.976968   10.695662       8.013988   \n",
      "TCGA-3C-AALI     9.914682   5.902438     8.809329   10.391374       7.632831   \n",
      "TCGA-3C-AALJ    11.305650   5.143969     9.060691    9.586488       8.374267   \n",
      "TCGA-3C-AALK     9.384994   5.782065     8.773906    9.754688       7.454703   \n",
      "TCGA-4H-AAAK     9.397606   5.612830     8.728789   10.035881       3.811738   \n",
      "\n",
      "              ZYG11B_79699   ZYX_7791  ZZEF1_23140  ZZZ3_26009  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU     10.238851  11.776124    10.887932   10.205129   \n",
      "TCGA-3C-AALI      9.237422  12.426428    10.364848    8.667973   \n",
      "TCGA-3C-AALJ      9.055784  12.414355     9.880935    8.992994   \n",
      "TCGA-3C-AALK      9.246419  12.474556     9.609426    9.453001   \n",
      "TCGA-4H-AAAK      9.599438  11.980747     9.700292    9.784147   \n",
      "\n",
      "              psiTPTE22_387590  \n",
      "patient                         \n",
      "TCGA-3C-AAAU          0.785174  \n",
      "TCGA-3C-AALI          9.855788  \n",
      "TCGA-3C-AALJ          5.143969  \n",
      "TCGA-3C-AALK          6.057699  \n",
      "TCGA-4H-AAAK          7.548699  \n",
      "\n",
      "[5 rows x 18321 columns]\n",
      "              hsa_let_7a_1  hsa_let_7a_2  hsa_let_7a_3  hsa_let_7b  \\\n",
      "patient                                                              \n",
      "TCGA-3C-AAAU     13.129765     14.117933     13.147714   14.595135   \n",
      "TCGA-3C-AALI     12.918069     13.922300     12.913194   14.512657   \n",
      "TCGA-3C-AALJ     13.012033     14.010002     13.028483   13.419612   \n",
      "TCGA-3C-AALK     13.144697     14.141721     13.151281   14.667196   \n",
      "TCGA-4H-AAAK     13.411684     14.413518     13.420481   14.438548   \n",
      "\n",
      "              hsa_let_7c  hsa_let_7d  hsa_let_7e  hsa_let_7f_1  hsa_let_7f_2  \\\n",
      "patient                                                                        \n",
      "TCGA-3C-AAAU    8.414890    8.665921   10.521777      3.879392     11.824817   \n",
      "TCGA-3C-AALI    9.646536    9.003653    9.131760      4.386952     12.678841   \n",
      "TCGA-3C-AALJ    9.312455    9.276943   11.395711      5.314692     13.530255   \n",
      "TCGA-3C-AALK   11.511431    8.384763   10.368981      4.159182     12.652559   \n",
      "TCGA-4H-AAAK   11.693927    8.453747   10.741371      4.494537     13.009499   \n",
      "\n",
      "              hsa_let_7g  ...  hsa_mir_937  hsa_mir_939  hsa_mir_940  \\\n",
      "patient                   ...                                          \n",
      "TCGA-3C-AAAU    8.597744  ...     0.906699    -0.093302     2.672234   \n",
      "TCGA-3C-AALI    8.455144  ...     1.579597    -0.083367     0.139024   \n",
      "TCGA-3C-AALJ    9.230563  ...     3.270298    -2.189134     0.395828   \n",
      "TCGA-3C-AALK    8.471503  ...     0.923965    -0.660997    -0.076034   \n",
      "TCGA-4H-AAAK    8.381220  ...     0.182950    -0.624403    -1.624403   \n",
      "\n",
      "              hsa_mir_942  hsa_mir_944  hsa_mir_95  hsa_mir_96  hsa_mir_98  \\\n",
      "patient                                                                      \n",
      "TCGA-3C-AAAU     2.467414     1.044202    2.044202    6.906699    5.754696   \n",
      "TCGA-3C-AALI     3.032109    -0.668331    0.331670    5.912870    6.427066   \n",
      "TCGA-3C-AALJ     1.855261    -0.381778    0.717757    6.603657    6.878301   \n",
      "TCGA-3C-AALK     1.798435     1.798435    0.798435    6.181354    5.377922   \n",
      "TCGA-4H-AAAK     1.076036     0.182950   -0.302475    4.318110    5.103516   \n",
      "\n",
      "              hsa_mir_99a  hsa_mir_99b  \n",
      "patient                                 \n",
      "TCGA-3C-AAAU     7.024602    15.506461  \n",
      "TCGA-3C-AALI     7.885299    13.626182  \n",
      "TCGA-3C-AALJ     7.580704    15.013822  \n",
      "TCGA-3C-AALK    10.031619    14.554783  \n",
      "TCGA-4H-AAAK    10.078201    14.650338  \n",
      "\n",
      "[5 rows x 503 columns]\n",
      "             synchronous_malignancy ajcc_pathologic_stage  days_to_diagnosis  \\\n",
      "patient                                                                        \n",
      "TCGA-3C-AAAU                     No               Stage X                0.0   \n",
      "TCGA-3C-AALI                     No             Stage IIB                0.0   \n",
      "TCGA-3C-AALJ                     No             Stage IIB                0.0   \n",
      "TCGA-3C-AALK                     No              Stage IA                0.0   \n",
      "TCGA-4H-AAAK                     No            Stage IIIA                0.0   \n",
      "\n",
      "             laterality  created_datetime  last_known_disease_status  \\\n",
      "patient                                                                \n",
      "TCGA-3C-AAAU       Left               NaN                        NaN   \n",
      "TCGA-3C-AALI      Right               NaN                        NaN   \n",
      "TCGA-3C-AALJ      Right               NaN                        NaN   \n",
      "TCGA-3C-AALK      Right               NaN                        NaN   \n",
      "TCGA-4H-AAAK       Left               NaN                        NaN   \n",
      "\n",
      "             tissue_or_organ_of_origin  days_to_last_follow_up  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU               Breast, NOS                     NaN   \n",
      "TCGA-3C-AALI               Breast, NOS                     NaN   \n",
      "TCGA-3C-AALJ               Breast, NOS                     NaN   \n",
      "TCGA-3C-AALK               Breast, NOS                     NaN   \n",
      "TCGA-4H-AAAK               Breast, NOS                     NaN   \n",
      "\n",
      "              age_at_diagnosis                 primary_diagnosis  ...  \\\n",
      "patient                                                           ...   \n",
      "TCGA-3C-AAAU           20211.0            Lobular carcinoma, NOS  ...   \n",
      "TCGA-3C-AALI           18538.0  Infiltrating duct carcinoma, NOS  ...   \n",
      "TCGA-3C-AALJ           22848.0  Infiltrating duct carcinoma, NOS  ...   \n",
      "TCGA-3C-AALK           19074.0  Infiltrating duct carcinoma, NOS  ...   \n",
      "TCGA-4H-AAAK           18371.0            Lobular carcinoma, NOS  ...   \n",
      "\n",
      "             pathology_N_stage pathology_M_stage gender.1  \\\n",
      "patient                                                     \n",
      "TCGA-3C-AAAU                nx                mx   female   \n",
      "TCGA-3C-AALI               n1a                m0   female   \n",
      "TCGA-3C-AALJ               n1a                m0   female   \n",
      "TCGA-3C-AALK           n0 (i+)                m0   female   \n",
      "TCGA-4H-AAAK               n2a                m0   female   \n",
      "\n",
      "              date_of_initial_pathologic_diagnosis days_to_last_known_alive  \\\n",
      "patient                                                                       \n",
      "TCGA-3C-AAAU                                2004.0                      NaN   \n",
      "TCGA-3C-AALI                                2003.0                      NaN   \n",
      "TCGA-3C-AALJ                                2011.0                      NaN   \n",
      "TCGA-3C-AALK                                2011.0                      NaN   \n",
      "TCGA-4H-AAAK                                2013.0                      NaN   \n",
      "\n",
      "             radiation_therapy               histological_type  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU                no  infiltrating lobular carcinoma   \n",
      "TCGA-3C-AALI               yes   infiltrating ductal carcinoma   \n",
      "TCGA-3C-AALJ                no   infiltrating ductal carcinoma   \n",
      "TCGA-3C-AALK                no   infiltrating ductal carcinoma   \n",
      "TCGA-4H-AAAK                no  infiltrating lobular carcinoma   \n",
      "\n",
      "              number_of_lymph_nodes                     race.1  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU                    4.0                      white   \n",
      "TCGA-3C-AALI                    1.0  black or african american   \n",
      "TCGA-3C-AALJ                    1.0  black or african american   \n",
      "TCGA-3C-AALK                    0.0  black or african american   \n",
      "TCGA-4H-AAAK                    4.0                      white   \n",
      "\n",
      "                         ethnicity.1  \n",
      "patient                               \n",
      "TCGA-3C-AAAU  not hispanic or latino  \n",
      "TCGA-3C-AALI  not hispanic or latino  \n",
      "TCGA-3C-AALJ  not hispanic or latino  \n",
      "TCGA-3C-AALK  not hispanic or latino  \n",
      "TCGA-4H-AAAK  not hispanic or latino  \n",
      "\n",
      "[5 rows x 118 columns]\n",
      "              pam50\n",
      "patient            \n",
      "TCGA-3C-AAAU      3\n",
      "TCGA-3C-AALI      2\n",
      "TCGA-3C-AALJ      4\n",
      "TCGA-3C-AALK      3\n",
      "TCGA-4H-AAAK      3\n"
     ]
    }
   ],
   "source": [
    "# To confirm our data saved and loads properly:\n",
    "meth = pd.read_csv(root / \"meth.csv\", index_col=0)\n",
    "rna = pd.read_csv(root / \"rna.csv\", index_col=0)\n",
    "mirna = pd.read_csv(root / \"mirna.csv\", index_col=0)\n",
    "pam50 = pd.read_csv(root / \"pam50.csv\", index_col=0)\n",
    "clinical = pd.read_csv(root / \"clinical.csv\", index_col=0)\n",
    "    \n",
    "print(meth.head())\n",
    "print(rna.head())\n",
    "print(mirna.head())\n",
    "print(clinical.head())\n",
    "print(pam50.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09265512",
   "metadata": {},
   "source": [
    "## Feature Selection\n",
    "\n",
    "To support downstream analysis, we provide three built-in feature selection methods: +\n",
    "- variance thresholding \n",
    "- ANOVA F-test\n",
    "- Random Forest importance. \n",
    "\n",
    "These are designed to help users quickly identify the most informative features from high-dimensional omics datasets. Each method captures different statistical properties, ranging from general variability to class-based separability and model-derived relevance. In this section, we put all three to the test and examine how much they agree with each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa70dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bioneuralnet.utils.preprocess import select_top_k_variance\n",
    "from bioneuralnet.utils.preprocess import top_anova_f_features\n",
    "from bioneuralnet.utils.preprocess import select_top_randomforest\n",
    "\n",
    "# feature selection\n",
    "meth_highvar = select_top_k_variance(meth, k=6000)\n",
    "rna_highvar = select_top_k_variance(rna, k=6000)\n",
    "\n",
    "meth_af = top_anova_f_features(meth, pam50, max_features=6000)\n",
    "rna_af = top_anova_f_features(rna, pam50, max_features=6000)\n",
    "\n",
    "meth_rf = select_top_randomforest(meth, pam50, top_k=6000)\n",
    "rna_rf = select_top_randomforest(rna, pam50, top_k=6000)\n",
    "\n",
    "meth_var = list(meth_highvar.columns)\n",
    "meth_anova = list(meth_af.columns)\n",
    "meth_rf = list(meth_rf.columns)\n",
    "\n",
    "rna_var = list(rna_highvar.columns)\n",
    "rna_anova = list(rna_af.columns)\n",
    "rna_rf = list(rna_rf.columns)\n",
    "\n",
    "inter1 = []\n",
    "for x in meth_anova:\n",
    "    if x in meth_var:\n",
    "        inter1.append(x)\n",
    "\n",
    "inter2 = []\n",
    "for x in meth_rf:\n",
    "    if x in meth_var:\n",
    "        inter2.append(x)\n",
    "\n",
    "inter3 = []\n",
    "for x in meth_anova:\n",
    "    if x in meth_rf:\n",
    "        inter3.append(x)\n",
    "\n",
    "meth_all_three = []\n",
    "for x in meth_anova:\n",
    "    if x in meth_rf and x in meth_var:\n",
    "        meth_all_three.append(x)\n",
    "\n",
    "inter4 = []\n",
    "for x in rna_anova:\n",
    "    if x in rna_var:\n",
    "        inter4.append(x)\n",
    "\n",
    "inter5 = []\n",
    "for x in rna_rf:\n",
    "    if x in rna_var:\n",
    "        inter5.append(x)\n",
    "\n",
    "inter6 = []\n",
    "for x in rna_anova:\n",
    "    if x in rna_rf:\n",
    "        inter6.append(x)\n",
    "\n",
    "rna_all_three = []\n",
    "for x in rna_anova:\n",
    "    if x in rna_rf and x in rna_var:\n",
    "        rna_all_three.append(x)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cc981cdb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Methylation feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 2091 features\n",
      "Random Forest & variance selection share: 1871 features\n",
      "Anova-F & Random Forest share: 2201 features\n",
      "All three methods agree on: 815 features\n"
     ]
    }
   ],
   "source": [
    "print(\"Methylation feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(meth_all_three)} features\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "da639dd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "RNA feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 2340 features\n",
      "Random Forest & variance selection share: 2218 features\n",
      "Anova-F & Random Forest share: 2546 features\n",
      "All three methods agree on: 1134 features\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nRNA feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(inter4)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(inter5)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(inter6)} features\")\n",
    "print(f\"All three methods agree on: {len(rna_all_three)} features\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1c371389",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = Path(\"/home/vicente/Github/BioNeuralNet/TCGA_BRCA_DATA/ANOVA\")\n",
    "\n",
    "rna_af.to_csv(out_dir / \"rna_anova.csv\")\n",
    "meth_af.to_csv(out_dir / \"meth_anova.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2168980",
   "metadata": {},
   "source": [
    "## Easy Access via DatasetLoader\n",
    "\n",
    "To facilitate working with this data, we have made it available through our `DatasetLoader` component. Due to GitHub and PyPI file size limitations, we selected the top 6,000 features from each omics modality using ANOVA F-test. If you have additional pre-processed or raw datasets you would like to include, feel free to reach out, we are happy to support expanding the platform and adding new datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2d0340f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGCA BRCA dataset shape: {'mirna': (769, 503), 'pam50': (769, 1), 'clinical': (769, 118), 'rna': (769, 6000), 'meth': (769, 6000)}\n"
     ]
    }
   ],
   "source": [
    "from bioneuralnet.datasets import DatasetLoader\n",
    "\n",
    "tgca_brca = DatasetLoader(\"brca\")\n",
    "\n",
    "print(f\"TGCA BRCA dataset shape: {tgca_brca.shape}\")\n",
    "brca_meth = tgca_brca.data[\"meth\"]\n",
    "brca_rna = tgca_brca.data[\"rna\"]\n",
    "brca_mirna = tgca_brca.data[\"mirna\"]\n",
    "brca_clinical = tgca_brca.data[\"clinical\"]\n",
    "brca_pam50 = tgca_brca.data[\"pam50\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "338dc995",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-23 12:34:11,231 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-05-23 12:34:11,232 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 31384 NaNs after median imputation\n",
      "2025-05-23 12:34:11,232 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 39 columns dropped due to zero variance\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA shape: (769, 6000)\n",
      "METH shape: (769, 6000)\n",
      "miRNA shape: (769, 503)\n",
      "Clinical shape: (769, 118)\n",
      "Phenotype shape: (769, 1)\n",
      "Phenotype counts:\n",
      "pam50\n",
      "3        419\n",
      "4        140\n",
      "1        130\n",
      "2         46\n",
      "0         34\n",
      "Name: count, dtype: int64\n",
      "Nan values in pam50 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-23 12:34:11,612 - bioneuralnet.utils.preprocess - INFO - Selected top 15 features by RandomForest importance\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              age_at_diagnosis  days_to_birth  years_to_birth  age_at_index  \\\n",
      "patient                                                                       \n",
      "TCGA-3C-AAAU           20211.0       -20211.0            55.0          55.0   \n",
      "TCGA-3C-AALI           18538.0       -18538.0            50.0          50.0   \n",
      "TCGA-3C-AALJ           22848.0       -22848.0            62.0          62.0   \n",
      "TCGA-3C-AALK           19074.0       -19074.0            52.0          52.0   \n",
      "TCGA-4H-AAAK           18371.0       -18371.0            50.0          50.0   \n",
      "\n",
      "              days_to_last_followup  year_of_diagnosis  number_of_lymph_nodes  \\\n",
      "patient                                                                         \n",
      "TCGA-3C-AAAU                 4047.0              -1.50                    1.5   \n",
      "TCGA-3C-AALI                 4005.0              -1.75                    0.0   \n",
      "TCGA-3C-AALJ                 1474.0               0.25                    0.0   \n",
      "TCGA-3C-AALK                 1448.0               0.25                   -0.5   \n",
      "TCGA-4H-AAAK                  348.0               0.75                    1.5   \n",
      "\n",
      "              date_of_initial_pathologic_diagnosis  \\\n",
      "patient                                              \n",
      "TCGA-3C-AAAU                                 -1.50   \n",
      "TCGA-3C-AALI                                 -1.75   \n",
      "TCGA-3C-AALJ                                  0.25   \n",
      "TCGA-3C-AALK                                  0.25   \n",
      "TCGA-4H-AAAK                                  0.75   \n",
      "\n",
      "              histological_type_infiltrating lobular carcinoma  \\\n",
      "patient                                                          \n",
      "TCGA-3C-AAAU                                              True   \n",
      "TCGA-3C-AALI                                             False   \n",
      "TCGA-3C-AALJ                                             False   \n",
      "TCGA-3C-AALK                                             False   \n",
      "TCGA-4H-AAAK                                              True   \n",
      "\n",
      "              histological_type_infiltrating ductal carcinoma  \\\n",
      "patient                                                         \n",
      "TCGA-3C-AAAU                                            False   \n",
      "TCGA-3C-AALI                                             True   \n",
      "TCGA-3C-AALJ                                             True   \n",
      "TCGA-3C-AALK                                             True   \n",
      "TCGA-4H-AAAK                                            False   \n",
      "\n",
      "              primary_diagnosis_Lobular carcinoma, NOS  morphology_8520/3  \\\n",
      "patient                                                                     \n",
      "TCGA-3C-AAAU                                      True               True   \n",
      "TCGA-3C-AALI                                     False              False   \n",
      "TCGA-3C-AALJ                                     False              False   \n",
      "TCGA-3C-AALK                                     False              False   \n",
      "TCGA-4H-AAAK                                      True               True   \n",
      "\n",
      "              method_of_diagnosis_Core Biopsy  morphology_8500/3  \\\n",
      "patient                                                            \n",
      "TCGA-3C-AAAU                            False              False   \n",
      "TCGA-3C-AALI                             True               True   \n",
      "TCGA-3C-AALJ                             True               True   \n",
      "TCGA-3C-AALK                             True               True   \n",
      "TCGA-4H-AAAK                             True              False   \n",
      "\n",
      "              laterality_Right  \n",
      "patient                         \n",
      "TCGA-3C-AAAU             False  \n",
      "TCGA-3C-AALI              True  \n",
      "TCGA-3C-AALJ              True  \n",
      "TCGA-3C-AALK              True  \n",
      "TCGA-4H-AAAK             False  \n"
     ]
    }
   ],
   "source": [
    "from bioneuralnet.utils.preprocess import preprocess_clinical\n",
    "\n",
    "#shapes\n",
    "print(f\"RNA shape: {brca_rna.shape}\")\n",
    "print(f\"METH shape: {brca_meth.shape}\")\n",
    "print(f\"miRNA shape: {brca_mirna.shape}\")\n",
    "print(f\"Clinical shape: {brca_clinical.shape}\")\n",
    "print(f\"Phenotype shape: {brca_pam50.shape}\")\n",
    "print(f\"Phenotype counts:\\n{brca_pam50.value_counts()}\")\n",
    "\n",
    "#check nans in pam50\n",
    "print(f\"Nan values in pam50 {brca_pam50.isna().sum().sum()}\")\n",
    "brca_pam50 = brca_pam50.dropna()\n",
    "\n",
    "X_rna = brca_rna.loc[brca_pam50.index]\n",
    "X_meth = brca_meth.loc[brca_pam50.index]\n",
    "X_mirna = brca_mirna.loc[brca_pam50.index]\n",
    "clinical = brca_clinical.loc[brca_pam50.index]\n",
    "\n",
    "# for more details on the preprocessing function, see bioneuralnet.utils.preprocess\n",
    "clinical = preprocess_clinical(clinical, brca_pam50, top_k=15, scale=True, ignore_columns=[\"days_to_birth\", \"age_at_diagnosis\", \"days_to_last_followup\", \"age_at_index\", \"years_to_birth\"])\n",
    "print(clinical.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89cb8500",
   "metadata": {},
   "source": [
    "## Preparing Multi-Omics Data for downstream tasks\n",
    "\n",
    "1. Check sample overlap.\n",
    "\n",
    "2. Select top features.\n",
    "\n",
    "    - Although each omics dataset has already been filtered down to the top 6,000 features using the ANOVA F-test, this is still considered high-dimensional for most modeling tasks.\n",
    "\n",
    "    - To make the data more tractable and improve downstream performance, we apply ANOVA F-test again to select the top 1,000 most discriminative features from each dataset.\n",
    "\n",
    "3. Combine datasets.\n",
    "\n",
    "    - Selected features from RNA, methylation, and miRNA are combined into a single dataset.\n",
    "\n",
    "4. Clean missing values.\n",
    "\n",
    "    - Counts and removes any missing (nan) values from the combined dataset.\n",
    "\n",
    "5. Build similarity graph.\n",
    "\n",
    "    - Creates a k-nearest neighbors graph from the transposed feature matrix.\n",
    "\n",
    "    - Other supported methods include correlation-based graphs, soft-thresholding (WGCNA-style), Gaussian kernels, and mutual information networks.\n",
    "\n",
    "Note: For more details on preprocessing functions and graph generation algorithms, see the [Utils documentation](https://bioneuralnet.readthedocs.io/en/latest/utils.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b4646135",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-23 12:34:12,482 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-05-23 12:34:12,482 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-05-23 12:34:12,482 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-05-23 12:34:12,527 - bioneuralnet.utils.preprocess - INFO - Selected 1000 features by ANOVA (task=classification), 6000 significant, 0 padded\n",
      "2025-05-23 12:34:13,385 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-05-23 12:34:13,385 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-05-23 12:34:13,385 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-05-23 12:34:13,431 - bioneuralnet.utils.preprocess - INFO - Selected 1000 features by ANOVA (task=classification), 6000 significant, 0 padded\n",
      "2025-05-23 12:34:13,506 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-05-23 12:34:13,507 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-05-23 12:34:13,507 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-05-23 12:34:13,510 - bioneuralnet.utils.preprocess - INFO - Selected 503 features by ANOVA (task=classification), 465 significant, 38 padded\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nan values in X_train_full: 0\n",
      "Nan value in X_train_full after dropping: 0\n",
      "X_train_full shape: (769, 2503)\n",
      "\n",
      "Network shape: (2503, 2503)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import accuracy_score, f1_score\n",
    "from bioneuralnet.utils.preprocess import top_anova_f_features\n",
    "from bioneuralnet.utils.graph import gen_similarity_graph\n",
    "\n",
    "meth_sel = top_anova_f_features(X_meth, brca_pam50, max_features=1000)\n",
    "rna_sel = top_anova_f_features(X_rna, brca_pam50 ,max_features=1000)\n",
    "mirna_sel = top_anova_f_features(X_mirna, brca_pam50,max_features=503)\n",
    "X_train_full = pd.concat([meth_sel, rna_sel, mirna_sel], axis=1)\n",
    "\n",
    "# we check again for nan values then drop if any\n",
    "print(f\"Nan values in X_train_full: {X_train_full.isna().sum().sum()}\")\n",
    "X_train_full = X_train_full.dropna()\n",
    "print(f\"Nan value in X_train_full after dropping: {X_train_full.isna().sum().sum()}\")\n",
    "\n",
    "print(f\"X_train_full shape: {X_train_full.shape}\")\n",
    "# building the graph using the similarity graph function with k=15\n",
    "A_train = gen_similarity_graph(X_train_full.T, k=15)\n",
    "\n",
    "print(f\"\\nNetwork shape: {A_train.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "423f2808",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-23 12:34:14,025 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/TCGA_BRCA/results/run\n",
      "2025-05-23 12:34:14,025 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-05-23 12:34:14,026 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-05-23 12:34:14,038 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-05-23 12:34:14,039 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-05-23 12:34:14,039 - bioneuralnet.downstream_task.dpmon - INFO - Slicing omics dataset based on network nodes.\n",
      "2025-05-23 12:34:14,042 - bioneuralnet.downstream_task.dpmon - INFO - Building PyTorch Geometric Data object from adjacency matrix.\n",
      "2025-05-23 12:34:14,109 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 2503\n",
      "2025-05-23 12:34:14,109 - bioneuralnet.downstream_task.dpmon - INFO - Using clinical vars for node features: ['age_at_diagnosis', 'days_to_birth', 'years_to_birth', 'age_at_index', 'days_to_last_followup', 'year_of_diagnosis', 'number_of_lymph_nodes', 'date_of_initial_pathologic_diagnosis', 'histological_type_infiltrating lobular carcinoma', 'histological_type_infiltrating ductal carcinoma', 'primary_diagnosis_Lobular carcinoma, NOS', 'morphology_8520/3', 'method_of_diagnosis_Core Biopsy', 'morphology_8500/3', 'laterality_Right']\n",
      "2025-05-23 12:34:17,366 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (769, 2504)\n",
      "2025-05-23 12:34:18,207\tINFO worker.py:1888 -- Started a local Ray instance.\n",
      "2025-05-23 12:34:18,641\tINFO tune.py:253 -- Initializing Ray automatically. For cluster usage or custom Ray initialization, call `ray.init(...)` before `tune.run(...)`.\n",
      "2025-05-23 12:34:18,643\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m(raylet)\u001b[0m Warning: The actor ImplicitFunc is very large (32 MiB). Check that its definition is not implicitly capturing a large array or other object in scope. Tip: use ray.put() to put large objects in the Ray object store.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36m(pid=395677)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\n",
      "\u001b[36m(pid=395677)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "\u001b[36m(pid=395677)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395677)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "\u001b[36m(pid=395677)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395677)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "\u001b[36m(pid=395677)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395677)\u001b[0m   warnings.warn(\n",
      "\u001b[36m(pid=395677)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395677)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n",
      "\u001b[36m(tune_train_n pid=395677)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\n",
      "\u001b[36m(tune_train_n pid=395677)\u001b[0m   _log_deprecation_warning(\n",
      "\u001b[36m(tune_train_n pid=395677)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00000/checkpoint_000000)\n",
      "\u001b[36m(tune_train_n pid=395677)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00000/checkpoint_000001)\n",
      "\u001b[36m(tune_train_n pid=395677)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00000/checkpoint_000002)\n",
      "\u001b[36m(pid=395900)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\u001b[32m [repeated 2x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/user-guides/configure-logging.html#log-deduplication for more options.)\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m   warnings.warn(\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395900)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=395900)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=395900)\u001b[0m   _log_deprecation_warning(\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=395900)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00002/checkpoint_000003)\u001b[32m [repeated 5x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=395978)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\n",
      "\u001b[36m(pid=395978)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "\u001b[36m(pid=395978)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395978)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "\u001b[36m(pid=395978)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395978)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "\u001b[36m(pid=395978)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395978)\u001b[0m   warnings.warn(\n",
      "\u001b[36m(pid=395978)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=395978)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n",
      "\u001b[36m(pid=396057)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\n",
      "\u001b[36m(pid=396057)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "\u001b[36m(pid=396057)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396057)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "\u001b[36m(pid=396057)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396057)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "\u001b[36m(pid=396057)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396057)\u001b[0m   warnings.warn(\n",
      "\u001b[36m(pid=396057)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396057)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n",
      "\u001b[36m(tune_train_n pid=395978)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\n",
      "\u001b[36m(tune_train_n pid=395978)\u001b[0m   _log_deprecation_warning(\n",
      "\u001b[36m(tune_train_n pid=395978)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00003/checkpoint_000000)\u001b[32m [repeated 97x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396057)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\n",
      "\u001b[36m(tune_train_n pid=396057)\u001b[0m   _log_deprecation_warning(\n",
      "\u001b[36m(pid=396135)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\n",
      "\u001b[36m(pid=396135)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "\u001b[36m(pid=396135)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396135)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "\u001b[36m(pid=396135)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396135)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "\u001b[36m(pid=396135)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396135)\u001b[0m   warnings.warn(\n",
      "\u001b[36m(pid=396135)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396135)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n",
      "\u001b[36m(tune_train_n pid=396057)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00004/checkpoint_000016)\u001b[32m [repeated 17x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m   _log_deprecation_warning(\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00005/checkpoint_000020)\u001b[32m [repeated 21x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00005/checkpoint_000040)\u001b[32m [repeated 20x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00005/checkpoint_000060)\u001b[32m [repeated 20x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00005/checkpoint_000080)\u001b[32m [repeated 20x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396222)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\n",
      "\u001b[36m(pid=396222)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "\u001b[36m(pid=396222)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396222)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "\u001b[36m(pid=396222)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396222)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "\u001b[36m(pid=396222)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396222)\u001b[0m   warnings.warn(\n",
      "\u001b[36m(pid=396222)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396222)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n",
      "\u001b[36m(tune_train_n pid=396135)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00005/checkpoint_000099)\u001b[32m [repeated 19x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396222)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\n",
      "\u001b[36m(tune_train_n pid=396222)\u001b[0m   _log_deprecation_warning(\n",
      "\u001b[36m(pid=396379)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m   warnings.warn(\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396379)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396300)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00007/checkpoint_000002)\u001b[32m [repeated 4x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396379)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396379)\u001b[0m   _log_deprecation_warning(\u001b[32m [repeated 2x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396379)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00008/checkpoint_000029)\u001b[32m [repeated 30x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396379)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00008/checkpoint_000069)\u001b[32m [repeated 40x across cluster]\u001b[0m\n",
      "\u001b[36m(pid=396459)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:68: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/libpyg.so: undefined symbol: _ZNK5torch8autograd4Node4nameEv\n",
      "\u001b[36m(pid=396459)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'pyg-lib'. \"\n",
      "\u001b[36m(pid=396459)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:86: UserWarning: An issue occurred while importing 'torch-scatter'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_scatter/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396459)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-scatter'. \"\n",
      "\u001b[36m(pid=396459)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:97: UserWarning: An issue occurred while importing 'torch-cluster'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396459)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-cluster'. \"\n",
      "\u001b[36m(pid=396459)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:113: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_spline_conv/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396459)\u001b[0m   warnings.warn(\n",
      "\u001b[36m(pid=396459)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_geometric/typing.py:124: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/torch_sparse/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSsb\n",
      "\u001b[36m(pid=396459)\u001b[0m   warnings.warn(f\"An issue occurred while importing 'torch-sparse'. \"\n",
      "\u001b[36m(tune_train_n pid=396379)\u001b[0m Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/vicente/ray_results/tune_dp/T89880_00008/checkpoint_000099)\u001b[32m [repeated 30x across cluster]\u001b[0m\n",
      "\u001b[36m(tune_train_n pid=396459)\u001b[0m /home/vicente/Github/BioNeuralNet/.enviroment/lib/python3.12/site-packages/ray/train/_internal/session.py:772: RayDeprecationWarning: `ray.train.report` should be switched to `ray.tune.report` when running in a function passed to Ray Tune. This will be an error in the future. See this issue for more context: https://github.com/ray-project/ray/issues/49454\n",
      "\u001b[36m(tune_train_n pid=396459)\u001b[0m   _log_deprecation_warning(\n",
      "2025-05-23 12:35:30,014\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0044s.\n",
      "2025-05-23 12:35:30,024 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 128, 'gnn_hidden_dim': 128, 'lr': 0.0009583495398019029, 'weight_decay': 0.0819865221991172, 'nn_hidden_dim1': 128, 'nn_hidden_dim2': 128, 'num_epochs': 512}\n",
      "2025-05-23 12:35:30,024 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.9468936920166016\n",
      "2025-05-23 12:35:30,024 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 0.9986996098829649\n",
      "2025-05-23 12:35:30,027 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0            128             128  0.000958      0.081987             128   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0             128         512  \n",
      "2025-05-23 12:35:30,028 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 128, 'gnn_hidden_dim': 128, 'lr': 0.0009583495398019029, 'weight_decay': 0.0819865221991172, 'nn_hidden_dim1': 128, 'nn_hidden_dim2': 128, 'num_epochs': 512}\n",
      "2025-05-23 12:35:30,029 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 128, 'gnn_hidden_dim': 128, 'lr': 0.0009583495398019029, 'weight_decay': 0.0819865221991172, 'nn_hidden_dim1': 128, 'nn_hidden_dim2': 128, 'num_epochs': 512}\n",
      "2025-05-23 12:35:30,029 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-05-23 12:35:30,029 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-05-23 12:35:30,030 - bioneuralnet.downstream_task.dpmon - INFO - Slicing omics dataset based on network nodes.\n",
      "2025-05-23 12:35:30,034 - bioneuralnet.downstream_task.dpmon - INFO - Building PyTorch Geometric Data object from adjacency matrix.\n",
      "2025-05-23 12:35:30,106 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 2503\n",
      "2025-05-23 12:35:30,106 - bioneuralnet.downstream_task.dpmon - INFO - Using clinical vars for node features: ['age_at_diagnosis', 'days_to_birth', 'years_to_birth', 'age_at_index', 'days_to_last_followup', 'year_of_diagnosis', 'number_of_lymph_nodes', 'date_of_initial_pathologic_diagnosis', 'histological_type_infiltrating lobular carcinoma', 'histological_type_infiltrating ductal carcinoma', 'primary_diagnosis_Lobular carcinoma, NOS', 'morphology_8520/3', 'method_of_diagnosis_Core Biopsy', 'morphology_8500/3', 'laterality_Right']\n",
      "2025-05-23 12:35:33,430 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-05-23 12:35:33,595 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [1/512], Loss: 1.6216\n",
      "2025-05-23 12:35:33,684 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [10/512], Loss: 1.2003\n",
      "2025-05-23 12:35:33,781 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [20/512], Loss: 1.0579\n",
      "2025-05-23 12:35:33,877 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [30/512], Loss: 0.9766\n",
      "2025-05-23 12:35:33,973 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [40/512], Loss: 0.9418\n",
      "2025-05-23 12:35:34,069 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [50/512], Loss: 0.9767\n",
      "2025-05-23 12:35:34,165 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [60/512], Loss: 0.9356\n",
      "2025-05-23 12:35:34,261 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [70/512], Loss: 0.9267\n",
      "2025-05-23 12:35:34,357 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [80/512], Loss: 1.0248\n",
      "2025-05-23 12:35:34,453 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [90/512], Loss: 0.9387\n",
      "2025-05-23 12:35:34,549 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [100/512], Loss: 0.9261\n",
      "2025-05-23 12:35:34,646 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [110/512], Loss: 0.9273\n",
      "2025-05-23 12:35:34,742 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [120/512], Loss: 0.9610\n",
      "2025-05-23 12:35:34,838 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [130/512], Loss: 0.9516\n",
      "2025-05-23 12:35:34,934 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [140/512], Loss: 0.9283\n",
      "2025-05-23 12:35:35,030 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [150/512], Loss: 0.9265\n",
      "2025-05-23 12:35:35,127 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [160/512], Loss: 0.9271\n",
      "2025-05-23 12:35:35,223 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [170/512], Loss: 1.0042\n",
      "2025-05-23 12:35:35,320 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [180/512], Loss: 0.9589\n",
      "2025-05-23 12:35:35,416 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [190/512], Loss: 0.9301\n",
      "2025-05-23 12:35:35,512 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [200/512], Loss: 0.9257\n",
      "2025-05-23 12:35:35,610 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [210/512], Loss: 0.9342\n",
      "2025-05-23 12:35:35,709 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [220/512], Loss: 0.9312\n",
      "2025-05-23 12:35:35,805 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [230/512], Loss: 0.9306\n",
      "2025-05-23 12:35:35,901 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [240/512], Loss: 1.0224\n",
      "2025-05-23 12:35:35,997 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [250/512], Loss: 0.9424\n",
      "2025-05-23 12:35:36,092 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [260/512], Loss: 0.9295\n",
      "2025-05-23 12:35:36,188 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [270/512], Loss: 0.9287\n",
      "2025-05-23 12:35:36,284 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [280/512], Loss: 0.9484\n",
      "2025-05-23 12:35:36,380 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [290/512], Loss: 0.9358\n",
      "2025-05-23 12:35:36,476 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [300/512], Loss: 0.9339\n",
      "2025-05-23 12:35:36,572 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [310/512], Loss: 0.9390\n",
      "2025-05-23 12:35:36,669 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [320/512], Loss: 0.9669\n",
      "2025-05-23 12:35:36,766 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [330/512], Loss: 0.9349\n",
      "2025-05-23 12:35:36,863 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [340/512], Loss: 0.9316\n",
      "2025-05-23 12:35:36,960 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [350/512], Loss: 0.9586\n",
      "2025-05-23 12:35:37,056 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [360/512], Loss: 0.9425\n",
      "2025-05-23 12:35:37,153 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [370/512], Loss: 0.9365\n",
      "2025-05-23 12:35:37,249 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [380/512], Loss: 0.9370\n",
      "2025-05-23 12:35:37,346 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [390/512], Loss: 0.9526\n",
      "2025-05-23 12:35:37,442 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [400/512], Loss: 1.0081\n",
      "2025-05-23 12:35:37,539 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [410/512], Loss: 0.9446\n",
      "2025-05-23 12:35:37,636 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [420/512], Loss: 0.9366\n",
      "2025-05-23 12:35:37,732 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [430/512], Loss: 0.9437\n",
      "2025-05-23 12:35:37,829 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [440/512], Loss: 0.9421\n",
      "2025-05-23 12:35:37,926 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [450/512], Loss: 0.9581\n",
      "2025-05-23 12:35:38,022 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [460/512], Loss: 0.9613\n",
      "2025-05-23 12:35:38,119 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [470/512], Loss: 0.9473\n",
      "2025-05-23 12:35:38,216 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [480/512], Loss: 0.9423\n",
      "2025-05-23 12:35:38,313 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [490/512], Loss: 0.9440\n",
      "2025-05-23 12:35:38,409 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [500/512], Loss: 0.9661\n",
      "2025-05-23 12:35:38,505 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [510/512], Loss: 0.9663\n",
      "2025-05-23 12:35:38,532 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9571\n",
      "2025-05-23 12:35:38,535 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/TCGA_BRCA/results/run/dpm_model_iter_1.pth\n",
      "2025-05-23 12:35:38,539 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-05-23 12:35:38,563 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [1/512], Loss: 1.6230\n",
      "2025-05-23 12:35:38,650 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [10/512], Loss: 1.2156\n",
      "2025-05-23 12:35:38,746 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [20/512], Loss: 1.0822\n",
      "2025-05-23 12:35:38,841 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [30/512], Loss: 0.9898\n",
      "2025-05-23 12:35:38,937 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [40/512], Loss: 0.9762\n",
      "2025-05-23 12:35:39,033 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [50/512], Loss: 0.9380\n",
      "2025-05-23 12:35:39,129 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [60/512], Loss: 0.9298\n",
      "2025-05-23 12:35:39,225 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [70/512], Loss: 0.9889\n",
      "2025-05-23 12:35:39,321 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [80/512], Loss: 0.9356\n",
      "2025-05-23 12:35:39,418 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [90/512], Loss: 0.9265\n",
      "2025-05-23 12:35:39,514 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [100/512], Loss: 0.9295\n",
      "2025-05-23 12:35:39,610 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [110/512], Loss: 0.9552\n",
      "2025-05-23 12:35:39,706 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [120/512], Loss: 0.9293\n",
      "2025-05-23 12:35:39,802 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [130/512], Loss: 0.9249\n",
      "2025-05-23 12:35:39,899 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [140/512], Loss: 0.9327\n",
      "2025-05-23 12:35:39,995 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [150/512], Loss: 0.9623\n",
      "2025-05-23 12:35:40,091 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [160/512], Loss: 0.9329\n",
      "2025-05-23 12:35:40,187 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [170/512], Loss: 0.9265\n",
      "2025-05-23 12:35:40,283 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [180/512], Loss: 0.9297\n",
      "2025-05-23 12:35:40,380 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [190/512], Loss: 0.9404\n",
      "2025-05-23 12:35:40,476 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [200/512], Loss: 0.9405\n",
      "2025-05-23 12:35:40,572 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [210/512], Loss: 0.9378\n",
      "2025-05-23 12:35:40,668 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [220/512], Loss: 0.9288\n",
      "2025-05-23 12:35:40,764 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [230/512], Loss: 0.9279\n",
      "2025-05-23 12:35:40,860 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [240/512], Loss: 0.9314\n",
      "2025-05-23 12:35:40,956 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [250/512], Loss: 0.9632\n",
      "2025-05-23 12:35:41,053 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [260/512], Loss: 0.9367\n",
      "2025-05-23 12:35:41,149 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [270/512], Loss: 0.9301\n",
      "2025-05-23 12:35:41,245 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [280/512], Loss: 0.9309\n",
      "2025-05-23 12:35:41,341 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [290/512], Loss: 0.9345\n",
      "2025-05-23 12:35:41,438 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [300/512], Loss: 0.9350\n",
      "2025-05-23 12:35:41,535 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [310/512], Loss: 1.1157\n",
      "2025-05-23 12:35:41,631 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [320/512], Loss: 0.9597\n",
      "2025-05-23 12:35:41,728 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [330/512], Loss: 0.9347\n",
      "2025-05-23 12:35:41,824 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [340/512], Loss: 1.0187\n",
      "2025-05-23 12:35:41,920 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [350/512], Loss: 0.9502\n",
      "2025-05-23 12:35:42,017 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [360/512], Loss: 0.9354\n",
      "2025-05-23 12:35:42,113 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [370/512], Loss: 0.9354\n",
      "2025-05-23 12:35:42,209 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [380/512], Loss: 0.9844\n",
      "2025-05-23 12:35:42,306 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [390/512], Loss: 0.9478\n",
      "2025-05-23 12:35:42,402 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [400/512], Loss: 0.9367\n",
      "2025-05-23 12:35:42,498 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [410/512], Loss: 0.9397\n",
      "2025-05-23 12:35:42,595 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [420/512], Loss: 0.9424\n",
      "2025-05-23 12:35:42,691 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [430/512], Loss: 0.9487\n",
      "2025-05-23 12:35:42,788 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [440/512], Loss: 1.0282\n",
      "2025-05-23 12:35:42,884 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [450/512], Loss: 0.9507\n",
      "2025-05-23 12:35:42,981 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [460/512], Loss: 0.9504\n",
      "2025-05-23 12:35:43,077 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [470/512], Loss: 0.9612\n",
      "2025-05-23 12:35:43,173 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [480/512], Loss: 0.9494\n",
      "2025-05-23 12:35:43,270 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [490/512], Loss: 0.9689\n",
      "2025-05-23 12:35:43,367 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [500/512], Loss: 0.9620\n",
      "2025-05-23 12:35:43,463 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [510/512], Loss: 0.9482\n",
      "2025-05-23 12:35:43,487 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.6710\n",
      "2025-05-23 12:35:43,499 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/TCGA_BRCA/results/run/dpm_model_iter_2.pth\n",
      "2025-05-23 12:35:43,503 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-05-23 12:35:43,521 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [1/512], Loss: 1.6589\n",
      "2025-05-23 12:35:43,608 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [10/512], Loss: 1.2345\n",
      "2025-05-23 12:35:43,705 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [20/512], Loss: 1.0722\n",
      "2025-05-23 12:35:43,802 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [30/512], Loss: 0.9751\n",
      "2025-05-23 12:35:43,899 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [40/512], Loss: 0.9493\n",
      "2025-05-23 12:35:43,995 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [50/512], Loss: 0.9430\n",
      "2025-05-23 12:35:44,092 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [60/512], Loss: 0.9356\n",
      "2025-05-23 12:35:44,188 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [70/512], Loss: 0.9294\n",
      "2025-05-23 12:35:44,285 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [80/512], Loss: 0.9554\n",
      "2025-05-23 12:35:44,382 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [90/512], Loss: 0.9287\n",
      "2025-05-23 12:35:44,478 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [100/512], Loss: 0.9254\n",
      "2025-05-23 12:35:44,575 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [110/512], Loss: 1.0082\n",
      "2025-05-23 12:35:44,672 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [120/512], Loss: 0.9364\n",
      "2025-05-23 12:35:44,768 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [130/512], Loss: 0.9257\n",
      "2025-05-23 12:35:44,865 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [140/512], Loss: 0.9276\n",
      "2025-05-23 12:35:44,961 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [150/512], Loss: 0.9294\n",
      "2025-05-23 12:35:45,058 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [160/512], Loss: 1.1293\n",
      "2025-05-23 12:35:45,155 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [170/512], Loss: 0.9677\n",
      "2025-05-23 12:35:45,251 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [180/512], Loss: 0.9305\n",
      "2025-05-23 12:35:45,348 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [190/512], Loss: 0.9551\n",
      "2025-05-23 12:35:45,445 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [200/512], Loss: 0.9361\n",
      "2025-05-23 12:35:45,541 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [210/512], Loss: 0.9290\n",
      "2025-05-23 12:35:45,638 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [220/512], Loss: 0.9320\n",
      "2025-05-23 12:35:45,735 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [230/512], Loss: 0.9913\n",
      "2025-05-23 12:35:45,831 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [240/512], Loss: 0.9376\n",
      "2025-05-23 12:35:45,932 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [250/512], Loss: 0.9294\n",
      "2025-05-23 12:35:46,029 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [260/512], Loss: 0.9394\n",
      "2025-05-23 12:35:46,125 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [270/512], Loss: 1.0032\n",
      "2025-05-23 12:35:46,222 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [280/512], Loss: 0.9436\n",
      "2025-05-23 12:35:46,322 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [290/512], Loss: 0.9309\n",
      "2025-05-23 12:35:46,419 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [300/512], Loss: 1.0317\n",
      "2025-05-23 12:35:46,515 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [310/512], Loss: 0.9558\n",
      "2025-05-23 12:35:46,612 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [320/512], Loss: 0.9349\n",
      "2025-05-23 12:35:46,709 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [330/512], Loss: 0.9328\n",
      "2025-05-23 12:35:46,805 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [340/512], Loss: 1.0015\n",
      "2025-05-23 12:35:46,901 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [350/512], Loss: 0.9807\n",
      "2025-05-23 12:35:46,998 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [360/512], Loss: 0.9446\n",
      "2025-05-23 12:35:47,094 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [370/512], Loss: 0.9440\n",
      "2025-05-23 12:35:47,190 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [380/512], Loss: 0.9772\n",
      "2025-05-23 12:35:47,286 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [390/512], Loss: 0.9424\n",
      "2025-05-23 12:35:47,382 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [400/512], Loss: 0.9354\n",
      "2025-05-23 12:35:47,478 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [410/512], Loss: 0.9379\n",
      "2025-05-23 12:35:47,574 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [420/512], Loss: 0.9570\n",
      "2025-05-23 12:35:47,670 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [430/512], Loss: 0.9418\n",
      "2025-05-23 12:35:47,766 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [440/512], Loss: 0.9400\n",
      "2025-05-23 12:35:47,862 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [450/512], Loss: 0.9947\n",
      "2025-05-23 12:35:47,958 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [460/512], Loss: 0.9953\n",
      "2025-05-23 12:35:48,054 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [470/512], Loss: 0.9481\n",
      "2025-05-23 12:35:48,150 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [480/512], Loss: 0.9414\n",
      "2025-05-23 12:35:48,246 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [490/512], Loss: 1.0044\n",
      "2025-05-23 12:35:48,342 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [500/512], Loss: 0.9557\n",
      "2025-05-23 12:35:48,438 - bioneuralnet.downstream_task.dpmon - INFO - Epoch [510/512], Loss: 0.9466\n",
      "2025-05-23 12:35:48,461 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.7308\n",
      "2025-05-23 12:35:48,464 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/TCGA_BRCA/results/run/dpm_model_iter_3.pth\n",
      "2025-05-23 12:35:48,468 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 0.9571\n",
      "2025-05-23 12:35:48,469 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy: 0.7863\n",
      "2025-05-23 12:35:48,469 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation of Accuracy: 0.1509\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DPMON results:\n",
      "Accuracy: 0.9570871261378413\n",
      "F1 weighted: 0.9611444079317725\n",
      "F1 macro: 0.9270923032997647\n"
     ]
    }
   ],
   "source": [
    "from bioneuralnet.downstream_task import DPMON\n",
    "\n",
    "save = Path(\"/home/vicente/Github/BioNeuralNet/TCGA_BRCA/results\")\n",
    "brca_pam50 = brca_pam50.rename(columns={\"pam50\": \"phenotype\"})\n",
    "\n",
    "dpmon = DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[meth_sel, rna_sel, mirna_sel],\n",
    "    phenotype_data=brca_pam50,\n",
    "    clinical_data=clinical,\n",
    "    repeat_num=3,\n",
    "    tune=True,\n",
    "    gpu=True, \n",
    "    cuda=0,\n",
    "    output_dir=Path(save/\"run\"),\n",
    ")\n",
    "\n",
    "predictions_df, avg_accuracy = dpmon.run()\n",
    "actual = predictions_df[\"Actual\"]\n",
    "pred = predictions_df[\"Predicted\"]\n",
    "dp_acc = accuracy_score(actual, pred)\n",
    "dp_f1w = f1_score(actual, pred, average='weighted')\n",
    "dp_f1m = f1_score(actual, pred, average='macro')\n",
    "\n",
    "print(f\"\\nDPMON results:\")\n",
    "print(f\"Accuracy: {dp_acc}\")\n",
    "print(f\"F1 weighted: {dp_f1w}\")\n",
    "print(f\"F1 macro: {dp_f1m}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".enviroment",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}