{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dc820e6e",
   "metadata": {},
   "source": [
    "# TCGA-PAAD Analysis Demo\n",
    "\n",
    "- **Cohort**: Focuses on the **TCGA-PAAD** (Pancreatic Adenocarcinoma) dataset, a vital cohort for studying pancreatic cancer.\n",
    "\n",
    "- **Goal**: Perform survival prediction using a multi-omics profile.\n",
    "- **Prediction Target**: Predict **Overall Survival (OS)** based on the patient's integrated molecular data (RNA, Methylation, CNV, and clinical features).\n",
    "\n",
    "**Data Sources:**\n",
    "Omics Data: [https://xenabrowser.net/datapages/](https://xenabrowser.net/datapages/)\n",
    "Clinical Data: Broad Institute FireHose (`http://firebrowse.org/?cohort=PAAD`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ec5a7a8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-2J-AAB1-01</th>\n",
       "      <th>TCGA-2J-AAB4-01</th>\n",
       "      <th>TCGA-2J-AAB6-01</th>\n",
       "      <th>TCGA-2J-AAB8-01</th>\n",
       "      <th>TCGA-2J-AAB9-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACAP3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTRT2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGRN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TCGA-2J-AAB1-01  TCGA-2J-AAB4-01  TCGA-2J-AAB6-01  \\\n",
       "Gene Symbol                                                      \n",
       "ACAP3                      0                0                1   \n",
       "ACTRT2                     0                0                1   \n",
       "AGRN                       0                0                1   \n",
       "\n",
       "             TCGA-2J-AAB8-01  TCGA-2J-AAB9-01  \n",
       "Gene Symbol                                    \n",
       "ACAP3                      0                0  \n",
       "ACTRT2                     0                0  \n",
       "AGRN                       0                0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(24776, 184)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-2L-AAQL-01</th>\n",
       "      <th>TCGA-2J-AABI-01</th>\n",
       "      <th>TCGA-3A-A9J0-01</th>\n",
       "      <th>TCGA-3A-A9I7-01</th>\n",
       "      <th>TCGA-2J-AABO-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <td>10.2430</td>\n",
       "      <td>9.5734</td>\n",
       "      <td>10.3872</td>\n",
       "      <td>10.1431</td>\n",
       "      <td>9.8022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIF3A</th>\n",
       "      <td>6.6983</td>\n",
       "      <td>6.4428</td>\n",
       "      <td>5.5034</td>\n",
       "      <td>4.6491</td>\n",
       "      <td>7.2137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.1814</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           TCGA-2L-AAQL-01  TCGA-2J-AABI-01  TCGA-3A-A9J0-01  TCGA-3A-A9I7-01  \\\n",
       "sample                                                                          \n",
       "ARHGEF10L          10.2430           9.5734          10.3872          10.1431   \n",
       "HIF3A               6.6983           6.4428           5.5034           4.6491   \n",
       "RNF17               0.0000           0.0000           0.0000           1.1814   \n",
       "\n",
       "           TCGA-2J-AABO-01  \n",
       "sample                      \n",
       "ARHGEF10L           9.8022  \n",
       "HIF3A               7.2137  \n",
       "RNF17               0.0000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20530, 183)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-S4-A8RP-01</th>\n",
       "      <th>TCGA-IB-A6UG-01</th>\n",
       "      <th>TCGA-US-A776-01</th>\n",
       "      <th>TCGA-FZ-5926-01</th>\n",
       "      <th>TCGA-HZ-A8P1-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg13332474</th>\n",
       "      <td>0.0320</td>\n",
       "      <td>0.3927</td>\n",
       "      <td>0.0351</td>\n",
       "      <td>0.0442</td>\n",
       "      <td>0.0398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00651829</th>\n",
       "      <td>0.3976</td>\n",
       "      <td>0.1451</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0153</td>\n",
       "      <td>0.5791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg17027195</th>\n",
       "      <td>0.0464</td>\n",
       "      <td>0.5564</td>\n",
       "      <td>0.0417</td>\n",
       "      <td>0.0342</td>\n",
       "      <td>0.6533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            TCGA-S4-A8RP-01  TCGA-IB-A6UG-01  TCGA-US-A776-01  \\\n",
       "sample                                                          \n",
       "cg13332474           0.0320           0.3927           0.0351   \n",
       "cg00651829           0.3976           0.1451           0.0209   \n",
       "cg17027195           0.0464           0.5564           0.0417   \n",
       "\n",
       "            TCGA-FZ-5926-01  TCGA-HZ-A8P1-01  \n",
       "sample                                        \n",
       "cg13332474           0.0442           0.0398  \n",
       "cg00651829           0.0153           0.5791  \n",
       "cg17027195           0.0342           0.6533  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(485577, 195)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcga-2j-aabr</th>\n",
       "      <th>tcga-2j-aabt</th>\n",
       "      <th>tcga-3a-a9i5</th>\n",
       "      <th>tcga-3a-a9ij</th>\n",
       "      <th>tcga-3a-a9il</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hybridization REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>years_to_birth</th>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>57</td>\n",
       "      <td>65</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vital_status</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tcga-2j-aabr tcga-2j-aabt tcga-3a-a9i5 tcga-3a-a9ij  \\\n",
       "Hybridization REF                                                           \n",
       "Composite Element REF        value        value        value        value   \n",
       "years_to_birth                  60           72           57           65   \n",
       "vital_status                     0            0            0            0   \n",
       "\n",
       "                      tcga-3a-a9il  \n",
       "Hybridization REF                   \n",
       "Composite Element REF        value  \n",
       "years_to_birth                  39  \n",
       "vital_status                     0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20, 185)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS.time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS.time</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1-01</th>\n",
       "      <td>TCGA-2J-AAB1</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4-01</th>\n",
       "      <td>TCGA-2J-AAB4</td>\n",
       "      <td>0</td>\n",
       "      <td>729</td>\n",
       "      <td>0.0</td>\n",
       "      <td>729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6-01</th>\n",
       "      <td>TCGA-2J-AAB6</td>\n",
       "      <td>1</td>\n",
       "      <td>293</td>\n",
       "      <td>1.0</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     _PATIENT  OS  OS.time  DSS  DSS.time\n",
       "sample                                                   \n",
       "TCGA-2J-AAB1-01  TCGA-2J-AAB1   1       66  1.0        66\n",
       "TCGA-2J-AAB4-01  TCGA-2J-AAB4   0      729  0.0       729\n",
       "TCGA-2J-AAB6-01  TCGA-2J-AAB6   1      293  1.0       293"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(196, 10)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "probe map methylation:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg13332474</th>\n",
       "      <td>.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>25935146</td>\n",
       "      <td>25935148</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00651829</th>\n",
       "      <td>RSPH14,GNAZ</td>\n",
       "      <td>chr22</td>\n",
       "      <td>23413065</td>\n",
       "      <td>23413067</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg17027195</th>\n",
       "      <td>AUTS2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>69064092</td>\n",
       "      <td>69064094</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene  chrom  chromStart  chromEnd strand\n",
       "#id                                                        \n",
       "cg13332474            .   chr7    25935146  25935148      .\n",
       "cg00651829  RSPH14,GNAZ  chr22    23413065  23413067      .\n",
       "cg17027195        AUTS2   chr7    69064092  69064094      ."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(395985, 5)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "root = Path(\"/home/vicente/Github/BioNeuralNet/PAAN\")\n",
    "\n",
    "cnv_gistic_raw = pd.read_csv(root/\"Gistic2_CopyNumber_Gistic2_all_thresholded_by_genes.txt\", sep=\"\\t\",index_col=0,low_memory=False)                            \n",
    "rna_raw = pd.read_csv(root / \"HiSeqV2.txt\", sep=\"\\t\",index_col=0,low_memory=False)\n",
    "meth_raw = pd.read_csv(root/\"HumanMethylation450.txt\", sep='\\t',index_col=0,low_memory=False)\n",
    "clinical_raw = pd.read_csv(root / \"PAAD.clin.merged.picked.txt\",sep=\"\\t\", index_col=0, low_memory=False)\n",
    "target = pd.read_csv(root / \"survival_PAAD_survival.txt\", sep=\"\\t\", index_col=0, low_memory=False)\n",
    "\n",
    "probe_map_meth = pd.read_csv(\"probeMap_illuminaMethyl450_hg19_GPL16304_TCGAlegacy.txt\", sep=\"\\t\", index_col=0, low_memory=False)\n",
    "\n",
    "# display all shapes and first few rows of each dataset\n",
    "display(cnv_gistic_raw.iloc[:3,:5])\n",
    "display(cnv_gistic_raw.shape)\n",
    "\n",
    "display(rna_raw.iloc[:3,:5])\n",
    "display(rna_raw.shape)\n",
    "\n",
    "display(meth_raw.iloc[:3,:5])\n",
    "display(meth_raw.shape)\n",
    "\n",
    "display(clinical_raw.iloc[:3,:5])\n",
    "display(clinical_raw.shape)\n",
    "\n",
    "display(target.iloc[:3,:5])\n",
    "display(target.shape)\n",
    "\n",
    "print(f\"\\nprobe map methylation:\")\n",
    "display(probe_map_meth.iloc[:3,:5])\n",
    "display(probe_map_meth.shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d7d8a40",
   "metadata": {},
   "source": [
    "## Data Processing Summary\n",
    "\n",
    "1.  **Transpose Data:** All omics data (`cnv_gistic`, `rna`, `meth`) was flipped so rows represent patients and columns represent features.\n",
    "2.  **Process Methylation:** Raw methylation probes were mapped to gene names. Probes without genes were dropped, and probes mapping to multiple genes were split. The data was then averaged by gene to create a single feature per gene.\n",
    "3.  **Standardize Patient IDs:** Patient IDs in all tables (omics, clinical features, and outcomes) were cleaned and standardized to the 12-character TCGA format (e.g., `TCGA-2J-AAB1`) for matching.\n",
    "4.  **Handle Duplicates:** Duplicate patient rows were averaged in the omics data. The first entry was kept for duplicate patients in the clinical and outcomes data.\n",
    "5.  **Find Common Patients:** The script identified the list of patients that exist in *all five* datasets (`meth`, `rna`, `cnv_gistic`, `clinical`, and `outcomes`).\n",
    "6.  **Subset Data:** All data tables were filtered down to *only* this common list of patients, ensuring perfect alignment.\n",
    "7.  **Process CNV Data:** The GISTIC-formatted CNV data was vectorized to create two separate binary matrices: `cnv_amp_processed` (for gains/amplifications) and `cnv_del_processed` (for losses/deletions).\n",
    "8.  **Extract Target:** The `OS` (Overall Survival) column was pulled from the processed `outcomes` data to be used as the prediction target (y-variable)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "45293a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- 1. Load, Transpose, and Standardize IDs ---\n",
    "cnv_gistic = cnv_gistic_raw.T\n",
    "rna = rna_raw.T\n",
    "meth_transposed = meth_raw.T\n",
    "clinical = clinical_raw.T\n",
    "\n",
    "def trim_barcode(idx):\n",
    "    return idx.to_series().str.slice(0, 12)\n",
    "\n",
    "# Standardize all patient IDs FIRST\n",
    "meth_transposed.index = trim_barcode(meth_transposed.index)\n",
    "rna.index = trim_barcode(rna.index)\n",
    "cnv_gistic.index = trim_barcode(cnv_gistic.index)\n",
    "clinical.index = clinical.index.str.upper()\n",
    "clinical.index.name = \"Patient_ID\"\n",
    "\n",
    "outcomes = target.copy() \n",
    "outcomes = outcomes.set_index('_PATIENT')\n",
    "outcomes.index = outcomes.index.str.upper()\n",
    "outcomes.index.name = \"Patient_ID\"\n",
    "\n",
    "# --- 2. Handle Duplicate Patients ---\n",
    "# Average any patients that appear twice (e.g., from different vials)\n",
    "meth_transposed = meth_transposed.groupby(meth_transposed.index).mean()\n",
    "rna = rna.groupby(rna.index).mean()\n",
    "cnv_gistic = cnv_gistic.groupby(cnv_gistic.index).mean()\n",
    "clinical = clinical[~clinical.index.duplicated(keep='first')]\n",
    "outcomes = outcomes[~outcomes.index.duplicated(keep='first')]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fe45c812",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Found: 177 patients across all data types.\n"
     ]
    }
   ],
   "source": [
    "# --- 3. Find Common Patients & Subset ---\n",
    "common_patients = sorted(list(\n",
    "    set(meth_transposed.index) &\n",
    "    set(rna.index) &\n",
    "    set(cnv_gistic.index) &\n",
    "    set(clinical.index) &\n",
    "    set(outcomes.index)\n",
    "))\n",
    "\n",
    "print(f\"\\nFound: {len(common_patients)} patients across all data types.\")\n",
    "\n",
    "# SUBSET THE DATA *BEFORE* HEAVY PROCESSING\n",
    "meth_processed = meth_transposed.loc[common_patients]\n",
    "rna_processed = rna.loc[common_patients]\n",
    "cnv_gistic_processed = cnv_gistic.loc[common_patients]\n",
    "clinical_processed = clinical.loc[common_patients]\n",
    "outcomes_processed = outcomes.loc[common_patients]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "79b658f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "meth_processed_T = meth_processed.T\n",
    "\n",
    "probe_cleaned_gene = probe_map_meth.dropna(subset=['gene'])\n",
    "probe_cleaned_gene = probe_cleaned_gene[probe_cleaned_gene['gene'] != '.']\n",
    "probe_cleaned_gene['gene'] = probe_cleaned_gene['gene'].str.split(',')\n",
    "probe_map_meth_exploded = probe_cleaned_gene.explode('gene')\n",
    "\n",
    "meth_with_genes = meth_processed_T.join(probe_map_meth_exploded['gene'])\n",
    "meth_with_genes_filtered = meth_with_genes.dropna(subset=['gene'])\n",
    "\n",
    "\n",
    "data_columns = meth_with_genes_filtered.columns.drop('gene')\n",
    "meth_with_genes_filtered.loc[:, data_columns] = meth_with_genes_filtered[data_columns].apply(pd.to_numeric)\n",
    "meth_processed = meth_with_genes_filtered.groupby('gene').mean().T\n",
    "\n",
    "rna_processed = rna_processed.apply(pd.to_numeric, errors='coerce')\n",
    "cnv_gistic_processed = cnv_gistic_processed.apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "dfs_to_process = {\n",
    "    \"meth_\": meth_processed,\n",
    "    \"rna_\": rna_processed\n",
    "}\n",
    "\n",
    "\n",
    "for prefix, df in dfs_to_process.items():\n",
    "    df.columns = df.columns.str.replace(r\"\\?\", \"unknown_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(r\"\\|\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(\"-\", \"_\", regex=False)\n",
    "    df.columns = df.columns.str.replace(r\"_+\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.strip(\"_\")\n",
    "\n",
    "    # b. Prefixing (Ensures unique names)\n",
    "    # The .add_prefix() method is used here.\n",
    "    dfs_to_process[prefix] = df.add_prefix(prefix)\n",
    "\n",
    "# --- 3. Reassigning Results (DataFrames are now clean and prefixed) ---\n",
    "meth_processed = dfs_to_process[\"meth_\"]\n",
    "rna_processed = dfs_to_process[\"rna_\"]\n",
    "\n",
    "\n",
    "targets = outcomes_processed['OS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1999da58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cnv_gistic = cnv_gistic_raw.T\n",
    "# rna = rna_raw.T\n",
    "# meth_transposed = meth_raw.T\n",
    "# clinical = clinical_raw.T\n",
    "\n",
    "# probe_cleaned_gene = probe_map_meth.dropna(subset=['gene'])\n",
    "# probe_cleaned_gene = probe_cleaned_gene[probe_cleaned_gene['gene'] != '.']\n",
    "# probe_cleaned_gene['gene'] = probe_cleaned_gene['gene'].str.split(',')\n",
    "# probe_map_meth_exploded = probe_cleaned_gene.explode('gene')\n",
    "\n",
    "# meth_with_genes = meth_transposed.join(probe_map_meth_exploded['gene'])\n",
    "# meth_with_genes_filtered = meth_with_genes.dropna(subset=['gene'])\n",
    "# data_columns = meth_with_genes_filtered.columns.drop('gene')\n",
    "# meth_with_genes_filtered[data_columns] = meth_with_genes_filtered[data_columns].apply(pd.to_numeric)\n",
    "\n",
    "# meth = meth_with_genes_filtered.groupby('gene').mean()\n",
    "# print(f\"DNA Meth shape (all genes): {meth.shape}\")\n",
    "\n",
    "\n",
    "# def trim_barcode(idx):\n",
    "#     return idx.to_series().str.slice(0, 12)\n",
    "\n",
    "# meth.index = trim_barcode(meth.index)\n",
    "# rna.index = trim_barcode(rna.index)\n",
    "# cnv_gistic.index = trim_barcode(cnv_gistic.index)\n",
    "\n",
    "# clinical.index = clinical.index.str.upper()\n",
    "# clinical.index.name = \"Patient_ID\"\n",
    "# clinical = clinical[~clinical.index.duplicated(keep='first')]\n",
    "\n",
    "# outcomes = target.copy() \n",
    "# outcomes = outcomes.set_index('_PATIENT')\n",
    "# outcomes.index = outcomes.index.str.upper()\n",
    "# outcomes.index.name = \"Patient_ID\"\n",
    "# outcomes = outcomes[~outcomes.index.duplicated(keep='first')]\n",
    "\n",
    "# rna = rna.apply(pd.to_numeric, errors='coerce')\n",
    "# cnv_gistic = cnv_gistic.apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "\n",
    "# meth = meth.groupby(meth.index).mean()\n",
    "# rna = rna.groupby(rna.index).mean()\n",
    "# cnv_gistic = cnv_gistic.groupby(cnv_gistic.index).mean()\n",
    "\n",
    "# print(f\"\\nMethylation shape: {meth.shape}\")\n",
    "# print(f\"RNA shape: {rna.shape}\")\n",
    "# print(f\"cnv_gistic shape: {cnv_gistic.shape}\")\n",
    "# print(f\"Clinical shape: {clinical.shape}\")\n",
    "# print(f\"Outcomes shape: {outcomes.shape}\")\n",
    "\n",
    "# for df in [meth, rna, cnv_gistic]:\n",
    "#     df.columns = df.columns.str.replace(r\"\\?\", \"unknown_\", regex=True)\n",
    "#     df.columns = df.columns.str.replace(r\"\\|\", \"_\", regex=True)\n",
    "#     df.columns = df.columns.str.replace(\"-\", \"_\", regex=False)\n",
    "#     df.columns = df.columns.str.replace(r\"_+\", \"_\", regex=True)\n",
    "#     df.columns = df.columns.str.strip(\"_\")\n",
    "    \n",
    "#     df.fillna(df.mean(), inplace=True)\n",
    "\n",
    "# common_patients = sorted(list(\n",
    "#     set(meth.index) &\n",
    "#     set(rna.index) &\n",
    "#     set(cnv_gistic.index) &\n",
    "#     set(clinical.index) &\n",
    "#     set(outcomes.index)\n",
    "# ))\n",
    "\n",
    "# print(f\"\\nFound: {len(common_patients)} patients across all data types.\")\n",
    "\n",
    "# # subset to only common patients\n",
    "# meth_processed = meth.loc[common_patients]\n",
    "# rna_processed= rna.loc[common_patients]\n",
    "# cnv_gistic_processed = cnv_gistic.loc[common_patients]\n",
    "# clinical_processed = clinical.loc[common_patients]\n",
    "# outcomes_processed = outcomes.loc[common_patients]\n",
    "# outcomes = outcomes_processed['OS']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3d9ded19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ACAP3</th>\n",
       "      <th>ACTRT2</th>\n",
       "      <th>AGRN</th>\n",
       "      <th>ANKRD65</th>\n",
       "      <th>ATAD3A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol   ACAP3  ACTRT2  AGRN  ANKRD65  ATAD3A\n",
       "TCGA-2J-AAB1    0.0     0.0   0.0      0.0     0.0\n",
       "TCGA-2J-AAB4    0.0     0.0   0.0      0.0     0.0\n",
       "TCGA-2J-AAB6    1.0     1.0   1.0      1.0     1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 24776)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>rna_ARHGEF10L</th>\n",
       "      <th>rna_HIF3A</th>\n",
       "      <th>rna_RNF17</th>\n",
       "      <th>rna_RNF10</th>\n",
       "      <th>rna_RNF11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>9.7910</td>\n",
       "      <td>8.7310</td>\n",
       "      <td>0.5732</td>\n",
       "      <td>11.9111</td>\n",
       "      <td>10.2312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>10.5186</td>\n",
       "      <td>5.9351</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.7311</td>\n",
       "      <td>10.7823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>10.2412</td>\n",
       "      <td>6.6414</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>12.2415</td>\n",
       "      <td>11.4475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "sample        rna_ARHGEF10L  rna_HIF3A  rna_RNF17  rna_RNF10  rna_RNF11\n",
       "TCGA-2J-AAB1         9.7910     8.7310     0.5732    11.9111    10.2312\n",
       "TCGA-2J-AAB4        10.5186     5.9351     0.0000    11.7311    10.7823\n",
       "TCGA-2J-AAB6        10.2412     6.6414     0.0000    12.2415    11.4475"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 20530)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>meth_5S_rRNA</th>\n",
       "      <th>meth_7SK</th>\n",
       "      <th>meth_A1BG</th>\n",
       "      <th>meth_A1BG_AS1</th>\n",
       "      <th>meth_A1CF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>0.58890</td>\n",
       "      <td>0.025075</td>\n",
       "      <td>0.616812</td>\n",
       "      <td>0.615167</td>\n",
       "      <td>0.595000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>0.60546</td>\n",
       "      <td>0.026175</td>\n",
       "      <td>0.379675</td>\n",
       "      <td>0.356613</td>\n",
       "      <td>0.516583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>0.64782</td>\n",
       "      <td>0.023825</td>\n",
       "      <td>0.422000</td>\n",
       "      <td>0.392127</td>\n",
       "      <td>0.475000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene          meth_5S_rRNA  meth_7SK  meth_A1BG  meth_A1BG_AS1  meth_A1CF\n",
       "TCGA-2J-AAB1       0.58890  0.025075   0.616812       0.615167   0.595000\n",
       "TCGA-2J-AAB4       0.60546  0.026175   0.379675       0.356613   0.516583\n",
       "TCGA-2J-AAB6       0.64782  0.023825   0.422000       0.392127   0.475000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 34013)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hybridization REF</th>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_followup</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>value</td>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>value</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>value</td>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>293</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Hybridization REF Composite Element REF years_to_birth vital_status  \\\n",
       "Patient_ID                                                            \n",
       "TCGA-2J-AAB1                      value             65            1   \n",
       "TCGA-2J-AAB4                      value             48            0   \n",
       "TCGA-2J-AAB6                      value             75            1   \n",
       "\n",
       "Hybridization REF days_to_death days_to_last_followup  \n",
       "Patient_ID                                             \n",
       "TCGA-2J-AAB1                 66                   NaN  \n",
       "TCGA-2J-AAB4                NaN                   729  \n",
       "TCGA-2J-AAB6                293                   NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 20)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "OS\n",
       "1    93\n",
       "0    84\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(cnv_gistic_processed.iloc[:3,:5])\n",
    "display(cnv_gistic_processed.shape)\n",
    "\n",
    "display(rna_processed.iloc[:3,:5])\n",
    "display(rna_processed.shape)\n",
    "\n",
    "display(meth_processed.iloc[:3,:5])\n",
    "display(meth_processed.shape)\n",
    "\n",
    "display(clinical_processed.iloc[:3,:5])\n",
    "display(clinical_processed.shape)\n",
    "\n",
    "display(targets.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "da7f715e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 15:29:28,190 - bioneuralnet.utils.data - INFO - Starting Beta-to-M value conversion (shape: (177, 34013)). Epsilon: 1e-06\n",
      "2025-11-09 15:29:29,690 - bioneuralnet.utils.data - INFO - Beta-to-M conversion complete.\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "cnv_amp_processed = cnv_gistic_processed.isin([1, 2]).astype(int)\n",
    "cnv_del_processed = cnv_gistic_processed.isin([-1, -2]).astype(int)\n",
    "\n",
    "processed_dfs = {\n",
    "    \"amp_\": cnv_amp_processed,\n",
    "    \"del_\": cnv_del_processed\n",
    "}\n",
    "\n",
    "for prefix, df in processed_dfs.items():\n",
    "    df.columns = df.columns.str.replace(r\"\\?\", \"unknown_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(r\"\\|\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(\"-\", \"_\", regex=False)\n",
    "    df.columns = df.columns.str.replace(r\"_+\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.strip(\"_\")\n",
    "    processed_dfs[prefix] = df.add_prefix(prefix)\n",
    "\n",
    "cnv_amp_processed = processed_dfs[\"amp_\"]\n",
    "cnv_del_processed = processed_dfs[\"del_\"]\n",
    "\n",
    "clinical_processed.drop(columns=[\"Composite Element REF\"], errors=\"ignore\", inplace=True)\n",
    "\n",
    "meth_m_values = bnn.utils.beta_to_m(meth_processed, eps=1e-6) \n",
    "Y_labels = outcomes_processed['OS'].to_frame(name=\"target\")\n",
    "\n",
    "X_meth = meth_m_values.loc[common_patients]\n",
    "X_rna = rna_processed.loc[common_patients]\n",
    "clinical = clinical_processed.loc[common_patients]\n",
    "Y_labels = Y_labels.loc[common_patients]\n",
    "\n",
    "X_cnv_amp = cnv_amp_processed.loc[common_patients]\n",
    "X_cnv_del = cnv_del_processed.loc[common_patients]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e49ab618",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>amp_ACAP3</th>\n",
       "      <th>amp_ACTRT2</th>\n",
       "      <th>amp_AGRN</th>\n",
       "      <th>amp_ANKRD65</th>\n",
       "      <th>amp_ATAD3A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol   amp_ACAP3  amp_ACTRT2  amp_AGRN  amp_ANKRD65  amp_ATAD3A\n",
       "TCGA-2J-AAB1          0           0         0            0           0\n",
       "TCGA-2J-AAB4          0           0         0            0           0\n",
       "TCGA-2J-AAB6          1           1         1            1           1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 24776)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>del_ACAP3</th>\n",
       "      <th>del_ACTRT2</th>\n",
       "      <th>del_AGRN</th>\n",
       "      <th>del_ANKRD65</th>\n",
       "      <th>del_ATAD3A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Symbol   del_ACAP3  del_ACTRT2  del_AGRN  del_ANKRD65  del_ATAD3A\n",
       "TCGA-2J-AAB1          0           0         0            0           0\n",
       "TCGA-2J-AAB4          0           0         0            0           0\n",
       "TCGA-2J-AAB6          0           0         0            0           0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 24776)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>rna_ARHGEF10L</th>\n",
       "      <th>rna_HIF3A</th>\n",
       "      <th>rna_RNF17</th>\n",
       "      <th>rna_RNF10</th>\n",
       "      <th>rna_RNF11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>9.7910</td>\n",
       "      <td>8.7310</td>\n",
       "      <td>0.5732</td>\n",
       "      <td>11.9111</td>\n",
       "      <td>10.2312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>10.5186</td>\n",
       "      <td>5.9351</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.7311</td>\n",
       "      <td>10.7823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>10.2412</td>\n",
       "      <td>6.6414</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>12.2415</td>\n",
       "      <td>11.4475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "sample        rna_ARHGEF10L  rna_HIF3A  rna_RNF17  rna_RNF10  rna_RNF11\n",
       "TCGA-2J-AAB1         9.7910     8.7310     0.5732    11.9111    10.2312\n",
       "TCGA-2J-AAB4        10.5186     5.9351     0.0000    11.7311    10.7823\n",
       "TCGA-2J-AAB6        10.2412     6.6414     0.0000    12.2415    11.4475"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 20530)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>meth_5S_rRNA</th>\n",
       "      <th>meth_7SK</th>\n",
       "      <th>meth_A1BG</th>\n",
       "      <th>meth_A1BG_AS1</th>\n",
       "      <th>meth_A1CF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>0.518533</td>\n",
       "      <td>-5.280970</td>\n",
       "      <td>0.686782</td>\n",
       "      <td>0.676744</td>\n",
       "      <td>0.554968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>0.617860</td>\n",
       "      <td>-5.217401</td>\n",
       "      <td>-0.708259</td>\n",
       "      <td>-0.851325</td>\n",
       "      <td>0.095734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>0.879280</td>\n",
       "      <td>-5.356592</td>\n",
       "      <td>-0.453826</td>\n",
       "      <td>-0.632451</td>\n",
       "      <td>-0.144390</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene          meth_5S_rRNA  meth_7SK  meth_A1BG  meth_A1BG_AS1  meth_A1CF\n",
       "TCGA-2J-AAB1      0.518533 -5.280970   0.686782       0.676744   0.554968\n",
       "TCGA-2J-AAB4      0.617860 -5.217401  -0.708259      -0.851325   0.095734\n",
       "TCGA-2J-AAB6      0.879280 -5.356592  -0.453826      -0.632451  -0.144390"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 34013)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hybridization REF</th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_followup</th>\n",
       "      <th>tumor_tissue_site</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pancreas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>729</td>\n",
       "      <td>pancreas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>293</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pancreas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Hybridization REF years_to_birth vital_status days_to_death  \\\n",
       "Patient_ID                                                    \n",
       "TCGA-2J-AAB1                  65            1            66   \n",
       "TCGA-2J-AAB4                  48            0           NaN   \n",
       "TCGA-2J-AAB6                  75            1           293   \n",
       "\n",
       "Hybridization REF days_to_last_followup tumor_tissue_site  \n",
       "Patient_ID                                                 \n",
       "TCGA-2J-AAB1                        NaN          pancreas  \n",
       "TCGA-2J-AAB4                        729          pancreas  \n",
       "TCGA-2J-AAB6                        NaN          pancreas  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(177, 19)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "target\n",
       "1         93\n",
       "0         84\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(X_cnv_amp.iloc[:3,:5])\n",
    "display(X_cnv_amp.shape)\n",
    "\n",
    "display(X_cnv_del.iloc[:3,:5])\n",
    "display(X_cnv_del.shape)\n",
    "\n",
    "display(X_rna.iloc[:3,:5])\n",
    "display(X_rna.shape)\n",
    "\n",
    "display(X_meth.iloc[:3,:5])\n",
    "display(X_meth.shape)\n",
    "\n",
    "display(clinical.iloc[:3,:5])\n",
    "display(clinical.shape)\n",
    "\n",
    "display(Y_labels.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7767b027",
   "metadata": {},
   "source": [
    "## Feature Selection Methodology\n",
    "\n",
    "### Supported Methods and Interpretation\n",
    "\n",
    "**BioNeuralNet** provides three techniques for feature selection, allowing for different views of the data's statistical profile:\n",
    "\n",
    "- **Variance Thresholding:** Identifies features with the **highest overall variance** across all samples.\n",
    "\n",
    "- **ANOVA F-test:** Pinpoints features that best **distinguish between the target classes** (e.g., Alive vs. Deceased).\n",
    "\n",
    "- **Random Forest Importance:** Assesses **feature utility** based on its contribution to a predictive non-linear model.\n",
    "\n",
    "### PAAD Cohort Selection Strategy\n",
    "\n",
    "A dimensionality reduction step was essential for managing the high-feature-count omics data:\n",
    "\n",
    "- **High-Feature Datasets:** DNA Methylation (34,013), RNA (20,530), CNV Amplification (24,776), and CNV Deletion (24,776) all required significant feature reduction.\n",
    "\n",
    "- **Filtering Process:** As an example strategy, the **top 6,000 features** could be extracted from each high-feature omics dataset using all three methods.\n",
    "\n",
    "- **Final Set:** A consensus set could be built for each omics type by finding the intersection of features selected by the ANOVA F-test and Random Forest Importance, ensuring both statistical relevance and model-based utility.\n",
    "\n",
    "- **Low-Feature Datasets:** The **Clinical** data (19 features) was passed through **without selection**, as its feature count was already manageable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a96f822",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 15:29:33,506 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:33,507 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 6049 NaNs after median imputation\n",
      "2025-11-09 15:29:33,507 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:33,552 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-09 15:29:37,474 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:37,475 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 6049 NaNs after median imputation\n",
      "2025-11-09 15:29:37,475 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:37,580 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 0 significant, 6000 padded\n",
      "2025-11-09 15:29:41,338 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:41,338 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 6049 NaNs after median imputation\n",
      "2025-11-09 15:29:41,339 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:44,369 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:44,369 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:44,369 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 459 columns dropped due to zero variance\n",
      "2025-11-09 15:29:44,389 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-09 15:29:46,607 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:46,607 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:46,608 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 459 columns dropped due to zero variance\n",
      "2025-11-09 15:29:46,663 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 2643 significant, 3357 padded\n",
      "2025-11-09 15:29:48,901 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:48,902 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:48,902 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 459 columns dropped due to zero variance\n",
      "2025-11-09 15:29:51,874 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:51,875 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:51,875 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:51,904 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-09 15:29:54,325 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:54,325 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:54,325 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:54,394 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 0 significant, 6000 padded\n",
      "2025-11-09 15:29:56,830 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:56,831 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:56,831 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:59,642 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:29:59,642 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:29:59,643 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:29:59,666 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-09 15:30:02,074 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:30:02,074 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:30:02,075 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 15:30:02,143 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 0 significant, 6000 padded\n",
      "2025-11-09 15:30:04,566 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 15:30:04,566 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-09 15:30:04,566 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "meth_highvar = bnn.utils.select_top_k_variance(X_meth, k=6000)\n",
    "meth_af = bnn.utils.top_anova_f_features(X_meth, Y_labels, max_features=6000)\n",
    "meth_rf = bnn.utils.select_top_randomforest(X_meth, Y_labels, top_k=6000)\n",
    "\n",
    "rna_highvar = bnn.utils.select_top_k_variance(X_rna, k=6000)\n",
    "rna_af = bnn.utils.top_anova_f_features(X_rna, Y_labels, max_features=6000)\n",
    "rna_rf = bnn.utils.select_top_randomforest(X_rna, Y_labels, top_k=6000)\n",
    "\n",
    "cnv_amp_highvar = bnn.utils.select_top_k_variance(X_cnv_amp, k=6000)\n",
    "cnv_amp_af = bnn.utils.top_anova_f_features(X_cnv_amp, Y_labels, max_features=6000)\n",
    "cnv_amp_rf = bnn.utils.select_top_randomforest(X_cnv_amp, Y_labels, top_k=6000)\n",
    "\n",
    "cnv_del_highvar = bnn.utils.select_top_k_variance(X_cnv_del, k=6000)\n",
    "cnv_del_af = bnn.utils.top_anova_f_features(X_cnv_del, Y_labels, max_features=6000)\n",
    "cnv_del_rf = bnn.utils.select_top_randomforest(X_cnv_del, Y_labels, top_k=6000)\n",
    "\n",
    "meth_var_set = set(meth_highvar.columns)\n",
    "meth_anova_set = set(meth_af.columns)\n",
    "meth_rf_set = set(meth_rf.columns)\n",
    "\n",
    "rna_var_set = set(rna_highvar.columns)\n",
    "rna_anova_set = set(rna_af.columns)\n",
    "rna_rf_set = set(rna_rf.columns)\n",
    "\n",
    "cnv_amp_var_set = set(cnv_amp_highvar.columns)\n",
    "cnv_amp_anova_set = set(cnv_amp_af.columns)\n",
    "cnv_amp_rf_set = set(cnv_amp_rf.columns)\n",
    "\n",
    "cnv_del_var_set = set(cnv_del_highvar.columns)\n",
    "cnv_del_anova_set = set(cnv_del_af.columns)\n",
    "cnv_del_rf_set = set(cnv_del_rf.columns)\n",
    "\n",
    "meth_inter1 = list(meth_anova_set & meth_var_set)\n",
    "meth_inter2 = list(meth_rf_set & meth_var_set)\n",
    "meth_inter3 = list(meth_anova_set & meth_rf_set)\n",
    "meth_all_three = list(meth_anova_set & meth_var_set & meth_rf_set)\n",
    "\n",
    "rna_inter1 = list(rna_anova_set & rna_var_set)\n",
    "rna_inter2 = list(rna_rf_set & rna_var_set)\n",
    "rna_inter3 = list(rna_anova_set & rna_rf_set)\n",
    "rna_all_three = list(rna_anova_set & rna_var_set & rna_rf_set)\n",
    "\n",
    "cnv_amp_inter1 = list(cnv_amp_anova_set & cnv_amp_var_set)\n",
    "cnv_amp_inter2 = list(cnv_amp_rf_set & cnv_amp_var_set)\n",
    "cnv_amp_inter3 = list(cnv_amp_anova_set & cnv_amp_rf_set)\n",
    "cnv_amp_all_three = list(cnv_amp_anova_set & cnv_amp_var_set & cnv_amp_rf_set)\n",
    "\n",
    "cnv_del_inter1 = list(cnv_del_anova_set & cnv_del_var_set)\n",
    "cnv_del_inter2 = list(cnv_del_rf_set & cnv_del_var_set)\n",
    "cnv_del_inter3 = list(cnv_del_anova_set & cnv_del_rf_set)\n",
    "cnv_del_all_three = list(cnv_del_anova_set & cnv_del_var_set & cnv_del_rf_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6d3d1c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FROM THE 6000 Methylation feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 1416 features\n",
      "Random Forest & variance selection share: 1069 features\n",
      "Anova-F & Random Forest share: 1150 features\n",
      "All three methods agree on: 279 features\n",
      "\n",
      "FROM THE 6000 RNA feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 1910 features\n",
      "Random Forest & variance selection share: 1815 features\n",
      "Anova-F & Random Forest share: 1910 features\n",
      "All three methods agree on: 589 features\n",
      "\n",
      "FROM THE 6000 CNV Amplification feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 741 features\n",
      "Random Forest & variance selection share: 2310 features\n",
      "Anova-F & Random Forest share: 967 features\n",
      "All three methods agree on: 216 features\n",
      "\n",
      "FROM THE 6000 CNV Deletion feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 1361 features\n",
      "Random Forest & variance selection share: 2470 features\n",
      "Anova-F & Random Forest share: 1600 features\n",
      "All three methods agree on: 518 features\n"
     ]
    }
   ],
   "source": [
    "print(\"FROM THE 6000 Methylation feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(meth_inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(meth_inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(meth_inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(meth_all_three)} features\")\n",
    "\n",
    "print(\"\\nFROM THE 6000 RNA feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(rna_inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(rna_inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(rna_inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(rna_all_three)} features\")\n",
    "\n",
    "print(\"\\nFROM THE 6000 CNV Amplification feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(cnv_amp_inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(cnv_amp_inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(cnv_amp_inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(cnv_amp_all_three)} features\")\n",
    "\n",
    "print(\"\\nFROM THE 6000 CNV Deletion feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(cnv_del_inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(cnv_del_inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(cnv_del_inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(cnv_del_all_three)} features\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d61f83a",
   "metadata": {},
   "source": [
    "## Feature Selection Summary: ANOVA-RF Intersection\n",
    "\n",
    "The final set of features was determined by the **intersection** of those highlighted by the **ANOVA F-test** and **Random Forest Importance**. This methodology provides a balanced filter, capturing features with both high class-separability (ANOVA) and significant predictive value in a non-linear model (Random Forest). The resulting feature pool is considered highly relevant for the subsequent modeling tasks.\n",
    "\n",
    "### Feature Overlap Results\n",
    "\n",
    "The table below quantifies the shared features identified by the different selection techniques for each omics type.\n",
    "\n",
    "| Omics Data Type | ANOVA-F & Variance | RF & Variance | ANOVA-F & Random Forest (Selected) | All Three Agree |\n",
    "| :--- | :--- | :--- | :--- | :--- |\n",
    "| **Methylation** | 1,416 features | 1,069 features | **1,150 features** | 279 features |\n",
    "| **RNA** | 1,910 features | 1,815 features | **1,910 features** | 589 features |\n",
    "| **CNV Amplification** | 741 features | 2,310 features | **967 features** | 216 features |\n",
    "| **CNV Deletion** | 1,361 features | 2,470 features | **1,600 features** | 518 features |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "11a83aea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Final Shapes for Modeling\n",
      "Methylation (X_meth_selected): (177, 1150)\n",
      "RNA-Seq (X_rna_selected): (177, 1910)\n",
      "CNV Amplification (X_cnv_amp_selected): (177, 967)\n",
      "CNV Deletion (X_cnv_del_selected): (177, 1600)\n",
      "Clinical (clinical_selected): (177, 19)\n",
      "Labels (Y_labels): (177, 1)\n"
     ]
    }
   ],
   "source": [
    "# Subset each omics dataframe using the selected feature lists\n",
    "X_meth_selected = X_meth[meth_inter3]\n",
    "X_rna_selected = X_rna[rna_inter3]\n",
    "X_cnv_amp_selected = X_cnv_amp[cnv_amp_inter3]\n",
    "X_cnv_del_selected = X_cnv_del[cnv_del_inter3]\n",
    "\n",
    "# Clinical data is low-feature and kept as is\n",
    "clinical_selected = clinical\n",
    "\n",
    "print(\"\\nFinal Shapes for Modeling\")\n",
    "print(f\"Methylation (X_meth_selected): {X_meth_selected.shape}\")\n",
    "print(f\"RNA-Seq (X_rna_selected): {X_rna_selected.shape}\")\n",
    "print(f\"CNV Amplification (X_cnv_amp_selected): {X_cnv_amp_selected.shape}\")\n",
    "print(f\"CNV Deletion (X_cnv_del_selected): {X_cnv_del_selected.shape}\")\n",
    "print(f\"Clinical (clinical_selected): {clinical_selected.shape}\")\n",
    "print(f\"Labels (Y_labels): {Y_labels.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2e0b29d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 15:30:04,993 - bioneuralnet.utils.data - INFO - Starting KNN imputation (k=5) on DataFrame (shape: (177, 1150)).\n",
      "2025-11-09 15:30:05,004 - bioneuralnet.utils.data - INFO - KNN imputation complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaNs in dna_meth: 25\n",
      "NaNs in rna: 0\n",
      "NaNs in cnv_amp: 0\n",
      "NaNs in cnv_del: 0\n",
      "\n",
      "NaNs in dna_meth after: 0\n"
     ]
    }
   ],
   "source": [
    "# before we can save this data lets make sure we have no missing values\n",
    "\n",
    "# check NaNs in each omics dataset\n",
    "print(f\"NaNs in dna_meth: {X_meth_selected.isna().sum().sum()}\")\n",
    "print(f\"NaNs in rna: {X_rna_selected.isna().sum().sum()}\")\n",
    "print(f\"NaNs in cnv_amp: {X_cnv_amp_selected.isna().sum().sum()}\")\n",
    "print(f\"NaNs in cnv_del: {X_cnv_del_selected.isna().sum().sum()}\")\n",
    "\n",
    "# Impute missing values using BioNeuralNet KNN imputation\n",
    "X_meth_selected = bnn.utils.impute_omics_knn(X_meth_selected, n_neighbors=5)\n",
    "print(f\"\\nNaNs in dna_meth after: {X_meth_selected.isna().sum().sum()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9706f2e3",
   "metadata": {},
   "source": [
    "## Data Availability\n",
    "\n",
    "To facilitate rapid experimentation and reproduction of our results, the fully processed and feature-selected dataset used in this analysis has been made available directly within the package.\n",
    "\n",
    "Users can load this dataset, bypassing all preceding data acquisition, preprocessing, and feature selection steps. This allows users to proceed immediately from this step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "06f6244c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cnv_amp': (177, 967),\n",
       " 'cnv_del': (177, 1600),\n",
       " 'target': (177, 1),\n",
       " 'clinical': (177, 19),\n",
       " 'rna': (177, 1910),\n",
       " 'meth': (177, 1150)}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "tgca_paad = bnn.datasets.DatasetLoader(\"paad\")\n",
    "display(tgca_paad.shape)\n",
    "\n",
    "# The dataset is returned as a dictionary. We extract each file independetly based on the name (Key).\n",
    "cnv_amp = tgca_paad.data[\"cnv_amp\"]\n",
    "cnv_del = tgca_paad.data[\"cnv_del\"]\n",
    "clinical = tgca_paad.data[\"clinical\"]\n",
    "target = tgca_paad.data[\"target\"]\n",
    "dna_meth = tgca_paad.data[\"meth\"]\n",
    "rna = tgca_paad.data[\"rna\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00eca6b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['years_to_birth', 'vital_status', 'days_to_death',\n",
      "       'days_to_last_followup', 'tumor_tissue_site', 'pathologic_stage',\n",
      "       'pathology_T_stage', 'pathology_N_stage', 'pathology_M_stage', 'gender',\n",
      "       'date_of_initial_pathologic_diagnosis', 'radiation_therapy',\n",
      "       'histological_type', 'number_pack_years_smoked',\n",
      "       'year_of_tobacco_smoking_onset', 'residual_tumor',\n",
      "       'number_of_lymph_nodes', 'race', 'ethnicity'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(clinical.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4cbd23b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 19:19:59,488 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-09 19:19:59,488 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 125 NaNs after median imputation\n",
      "2025-11-09 19:19:59,488 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-09 19:19:59,566 - bioneuralnet.utils.preprocess - INFO - Selected top 7 features by RandomForest importance\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>number_of_lymph_nodes</th>\n",
       "      <th>number_pack_years_smoked</th>\n",
       "      <th>gender_male</th>\n",
       "      <th>pathology_M_stage_mx</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB1</th>\n",
       "      <td>65</td>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB4</th>\n",
       "      <td>48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2J-AAB6</th>\n",
       "      <td>75</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              years_to_birth  number_of_lymph_nodes  number_pack_years_smoked  \\\n",
       "Patient_ID                                                                      \n",
       "TCGA-2J-AAB1              65                    7.0                      25.0   \n",
       "TCGA-2J-AAB4              48                    0.0                      25.0   \n",
       "TCGA-2J-AAB6              75                    0.0                      25.0   \n",
       "\n",
       "              gender_male  pathology_M_stage_mx  \n",
       "Patient_ID                                       \n",
       "TCGA-2J-AAB1         True                 False  \n",
       "TCGA-2J-AAB4         True                 False  \n",
       "TCGA-2J-AAB6         True                 False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# BioNeuralNet provides a preprocessing function to handle clinical data\n",
    "clinical = tgca_paad.data[\"clinical\"]\n",
    "\n",
    "# for more details on the preprocessing functions, see `bioneuralnet.utils.preprocess`\n",
    "clinical_preprocessed = bnn.utils.preprocess_clinical(\n",
    "    clinical, \n",
    "    target, \n",
    "    top_k=7, \n",
    "    scale=False, \n",
    "    ignore_columns=[\n",
    "        # we ignore the following to avoid data leakage\n",
    "        \"vital_status\",\n",
    "        \"days_to_death\",\n",
    "        \"days_to_last_followup\",\n",
    "        \"tumor_tissue_site\",\n",
    "        \"histological_type\",\n",
    "        \"date_of_initial_pathologic_diagnosis\",\n",
    "        \"year_of_tobacco_smoking_onset\"\n",
    "    ])\n",
    "\n",
    "display(clinical_preprocessed.iloc[:3,:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c70a3fea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nan values in X_train_full: 0\n",
      "X_train_full shape: (177, 5627)\n",
      "\n",
      "Network shape: (5627, 5627)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import bioneuralnet as bnn\n",
    "\n",
    "X_train_full = pd.concat([dna_meth, rna, cnv_amp, cnv_del], axis=1)\n",
    "\n",
    "print(f\"Nan values in X_train_full: {X_train_full.isna().sum().sum()}\")\n",
    "print(f\"X_train_full shape: {X_train_full.shape}\")\n",
    "\n",
    "A_train = bnn.utils.gen_similarity_graph(X_train_full, k=15)\n",
    "print(f\"\\nNetwork shape: {A_train.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857dd03f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import os\n",
    "# import random\n",
    "# import logging \n",
    "# import warnings \n",
    "# import numpy as np\n",
    "# import torch\n",
    "# import ray\n",
    "# import bioneuralnet as bnn\n",
    "# from bioneuralnet.utils import logger \n",
    "\n",
    "# os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\"\n",
    "\n",
    "\n",
    "# if not ray.is_initialized():\n",
    "\n",
    "#     # 3. Set the ray init logging level to INFO\n",
    "#     ray.init(logging_level=logging.INFO) \n",
    "    \n",
    "#     # Ignore common warnings\n",
    "#     warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "#     warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03ecd4a7",
   "metadata": {},
   "source": [
    "## Reproducibility and Seeding\n",
    "\n",
    "To ensure our experimental results are fully reproducible, a single global seed is set at the beginning of the analysis.\n",
    "\n",
    "This utility function propagates the seed to all sources of randomness, including `random`, `numpy`, and `torch` (for both CPU and GPU). Critically, it also configures the PyTorch cuDNN backend to use deterministic algorithms.\n",
    "\n",
    "**for each DPMON outer iteration, the seed is incremented to generate a differnt internal test/train split.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b8428d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 19:20:00,656 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 118\n",
      "2025-11-09 19:20:00,657 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:20:00,657 - bioneuralnet.utils.data - INFO - Seed setting complete\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "SEED = 118\n",
    "bnn.utils.set_seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26be878e",
   "metadata": {},
   "source": [
    "## Classification using DPMON: Training and Evaluation\n",
    "\n",
    "* Run 3 outer iterations, each with a different seed.\n",
    "* Each iteration performs hyperparameter tuning.\n",
    "* After tuning, train `repeat_num = 3` models with the best hyperparameters.\n",
    "* Collect predictions from the best model of each iteration.\n",
    "* Compute **Accuracy**, **F1 Weighted**, and **F1 Macro +/- standard deviation** across iterations.\n",
    "\n",
    "This demonstrates the **end-to-end BioNeuralNet pipeline** in action.\n",
    "\n",
    "### Analysis of Hyperparameter Optimization\n",
    "\n",
    "The hyperparameter tuning results below showcase the best configuration found across three distinct GNN model runs.\n",
    "\n",
    "| Parameter | SAGE (GraphSAGE) | GCN (Graph Convolutional) | GAT (Graph Attention) |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| **gnn_layer_num** | 2 | 8 | 2 |\n",
    "| **gnn_hidden_dim** | 128 | 32 | 128 |\n",
    "| **lr (Learning Rate)** | 0.005197 | 0.000401 | 0.005197 |\n",
    "| **weight_decay** | 0.046079 | 0.007823 | 0.046079 |\n",
    "| **nn_hidden_dim1** | 16 | 32 | 16 |\n",
    "| **nn_hidden_dim2** | 32 | 32 | 32 |\n",
    "| **num_epochs** | 4096 | 2048 | 4096 |\n",
    "\n",
    "### Results\n",
    "\n",
    "| Model | Accuracy | F1 Weighted | F1 Macro |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| SAGE | 0.8512  0.1985 | 0.8129  0.2528 | 0.8083  0.2592 |\n",
    "| GCN  | 0.9962  0.0053 | 0.9962  0.0053 | 0.9962  0.0053 |\n",
    "| GAT  | 0.9171  0.1172 | 0.9129  0.1232 | 0.9138  0.1220 |\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403d3092",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 19:20:01,858 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 118\n",
      "2025-11-09 19:20:01,859 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:20:01,859 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:20:01,860 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad\n",
      "2025-11-09 19:20:01,860 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:20:01,860 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:20:01,872 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:20:01,872 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:20:02,095 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:20:05,835 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:20:05,836\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:20:37,481\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0109s.\n",
      "2025-11-09 19:20:37,502 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 2, 'gnn_hidden_dim': 128, 'lr': 0.005197038211089667, 'weight_decay': 0.04607928341516355, 'nn_hidden_dim1': 16, 'nn_hidden_dim2': 32, 'num_epochs': 4096}\n",
      "2025-11-09 19:20:37,503 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.338040828704834\n",
      "2025-11-09 19:20:37,503 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:20:37,505 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              2             128  0.005197      0.046079              16   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0              32        4096  \n",
      "2025-11-09 19:20:37,507 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 128, 'lr': 0.005197038211089667, 'weight_decay': 0.04607928341516355, 'nn_hidden_dim1': 16, 'nn_hidden_dim2': 32, 'num_epochs': 4096}\n",
      "2025-11-09 19:20:37,507 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 128, 'lr': 0.005197038211089667, 'weight_decay': 0.04607928341516355, 'nn_hidden_dim1': 16, 'nn_hidden_dim2': 32, 'num_epochs': 4096}\n",
      "2025-11-09 19:20:37,507 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:20:37,507 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:20:37,730 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:20:41,338 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:21:06,432 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.5706\n",
      "2025-11-09 19:21:06,435 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:21:06,439 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:21:31,442 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.5254\n",
      "2025-11-09 19:21:31,444 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:21:31,448 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:21:56,454 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.5254\n",
      "2025-11-09 19:21:56,457 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:21:56,461 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 0.5706\n",
      "2025-11-09 19:21:56,461 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.5405\n",
      "2025-11-09 19:21:56,461 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.0261\n",
      "2025-11-09 19:21:56,461 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n",
      "2025-11-09 19:21:56,464 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 119\n",
      "2025-11-09 19:21:56,464 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:21:56,464 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:21:56,465 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad\n",
      "2025-11-09 19:21:56,465 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:21:56,465 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:21:56,469 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:21:56,470 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:21:56,831 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:22:00,400 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:22:00,401\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:22:32,435\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0022s.\n",
      "2025-11-09 19:22:32,451 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:22:32,451 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.313837468624115\n",
      "2025-11-09 19:22:32,452 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:22:32,454 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              8              16  0.003115      0.000409              64   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0              64        4096  \n",
      "2025-11-09 19:22:32,456 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:22:32,456 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:22:32,456 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:22:32,456 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:22:32,678 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:22:36,425 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:22:43,418 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:22:43,421 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:22:43,423 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:22:50,329 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:22:50,341 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:22:50,342 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:22:57,424 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9944\n",
      "2025-11-09 19:22:57,437 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:22:57,438 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 1.0000\n",
      "2025-11-09 19:22:57,438 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.9981\n",
      "2025-11-09 19:22:57,438 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.0033\n",
      "2025-11-09 19:22:57,439 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n",
      "2025-11-09 19:22:57,441 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 120\n",
      "2025-11-09 19:22:57,442 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:22:57,442 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:22:57,442 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad\n",
      "2025-11-09 19:22:57,442 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:22:57,442 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:22:57,448 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:22:57,448 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:22:57,665 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:23:01,267 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:23:01,268\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:23:33,214\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0111s.\n",
      "2025-11-09 19:23:33,234 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:23:33,235 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.33702439069747925\n",
      "2025-11-09 19:23:33,235 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 0.9943502824858758\n",
      "2025-11-09 19:23:33,237 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim      lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              2              16  0.0027      0.002403              64   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0               8         512  \n",
      "2025-11-09 19:23:33,239 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:23:33,239 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:23:33,239 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:23:33,240 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:23:33,458 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:23:37,191 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:23:38,111 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.8475\n",
      "2025-11-09 19:23:38,123 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:23:38,126 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:23:39,031 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.5763\n",
      "2025-11-09 19:23:39,034 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:23:39,035 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:23:39,911 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9831\n",
      "2025-11-09 19:23:39,914 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:23:39,915 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 0.9831\n",
      "2025-11-09 19:23:39,915 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.8023\n",
      "2025-11-09 19:23:39,916 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.2071\n",
      "2025-11-09 19:23:39,916 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.8512 +/- 0.1985\n",
      "F1 Weighted: 0.8129 +/- 0.2528\n",
      "F1 Macro: 0.8083 +/- 0.2592\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import f1_score, accuracy_score\n",
    "from bioneuralnet.downstream_task import DPMON\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "\n",
    "output_dir_base_sage = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_results_SAGE_FINAL/paad\")\n",
    "target = target.rename(columns={\"target\": \"phenotype\"})\n",
    "\n",
    "n_repeats = 5\n",
    "all_preds = []\n",
    "\n",
    "for r in range(n_repeats):\n",
    "    bnn.utils.set_seed(SEED+r)\n",
    "    dpmon_repeat = DPMON(\n",
    "        adjacency_matrix=A_train,\n",
    "        omics_list=[dna_meth, rna, cnv_amp, cnv_del],\n",
    "        phenotype_data=target,\n",
    "        clinical_data=clinical_preprocessed,\n",
    "        repeat_num=3,\n",
    "        model='SAGE',\n",
    "        tune=True,\n",
    "        gpu=True,\n",
    "        cuda=0,\n",
    "        output_dir=output_dir_base_sage,\n",
    "    )\n",
    "    \n",
    "    predictions_df, _ = dpmon_repeat.run()\n",
    "    all_preds.append(predictions_df[\"Predicted\"].values)\n",
    "\n",
    "all_preds = np.array(all_preds)\n",
    "\n",
    "\n",
    "f1_macro_list = [f1_score(target, pred, average='macro') for pred in all_preds]\n",
    "f1_weighted_list = [f1_score(target, pred, average='weighted') for pred in all_preds]\n",
    "accuracy_list = [accuracy_score(target, pred) for pred in all_preds]\n",
    "\n",
    "avg_f1_macro = np.mean(f1_macro_list)\n",
    "std_f1_macro = np.std(f1_macro_list)\n",
    "\n",
    "avg_f1_weighted = np.mean(f1_weighted_list)\n",
    "std_f1_weighted = np.std(f1_weighted_list)\n",
    "\n",
    "avg_acc = np.mean(accuracy_list)\n",
    "std_acc = np.std(accuracy_list)\n",
    "\n",
    "print(f\"Accuracy: {avg_acc:.4f} +/- {std_acc:.4f}\")\n",
    "print(f\"F1 Weighted: {avg_f1_weighted:.4f} +/- {std_f1_weighted:.4f}\")\n",
    "print(f\"F1 Macro: {avg_f1_macro:.4f} +/- {std_f1_macro:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc2f416f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 19:23:39,936 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 118\n",
      "2025-11-09 19:23:39,937 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:23:39,937 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:23:39,937 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad\n",
      "2025-11-09 19:23:39,937 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:23:39,938 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:23:39,945 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:23:39,945 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:23:40,164 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:23:43,898 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:23:43,899\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:24:15,545\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0108s.\n",
      "2025-11-09 19:24:15,567 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 8, 'gnn_hidden_dim': 32, 'lr': 0.00040074470948823875, 'weight_decay': 0.007822838314187376, 'nn_hidden_dim1': 32, 'nn_hidden_dim2': 32, 'num_epochs': 2048}\n",
      "2025-11-09 19:24:15,567 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.363297700881958\n",
      "2025-11-09 19:24:15,568 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:24:15,571 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              8              32  0.000401      0.007823              32   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0              32        2048  \n",
      "2025-11-09 19:24:15,572 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 32, 'lr': 0.00040074470948823875, 'weight_decay': 0.007822838314187376, 'nn_hidden_dim1': 32, 'nn_hidden_dim2': 32, 'num_epochs': 2048}\n",
      "2025-11-09 19:24:15,573 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 32, 'lr': 0.00040074470948823875, 'weight_decay': 0.007822838314187376, 'nn_hidden_dim1': 32, 'nn_hidden_dim2': 32, 'num_epochs': 2048}\n",
      "2025-11-09 19:24:15,573 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:24:15,573 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:24:15,795 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:24:19,383 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:24:25,977 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.6215\n",
      "2025-11-09 19:24:25,979 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:24:25,982 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:24:32,528 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:24:32,530 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:24:32,533 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:24:39,105 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:24:39,107 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:24:39,109 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 1.0000\n",
      "2025-11-09 19:24:39,109 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.8738\n",
      "2025-11-09 19:24:39,109 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.2185\n",
      "2025-11-09 19:24:39,109 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n",
      "2025-11-09 19:24:39,112 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 119\n",
      "2025-11-09 19:24:39,112 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:24:39,113 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:24:39,113 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad\n",
      "2025-11-09 19:24:39,113 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:24:39,113 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:24:39,118 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:24:39,118 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:24:39,340 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:24:43,082 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:24:43,082\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:25:14,634\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0113s.\n",
      "2025-11-09 19:25:14,656 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:25:14,656 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.313874751329422\n",
      "2025-11-09 19:25:14,656 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:25:14,660 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              8              16  0.003115      0.000409              64   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0              64        4096  \n",
      "2025-11-09 19:25:14,661 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:25:14,661 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:25:14,662 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:25:14,662 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:25:14,883 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:25:18,622 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:25:27,703 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:25:27,715 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:25:27,717 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:25:36,791 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.7684\n",
      "2025-11-09 19:25:36,794 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:25:36,796 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:25:45,816 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9944\n",
      "2025-11-09 19:25:45,829 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:25:45,831 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 1.0000\n",
      "2025-11-09 19:25:45,831 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.9209\n",
      "2025-11-09 19:25:45,831 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.1321\n",
      "2025-11-09 19:25:45,831 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n",
      "2025-11-09 19:25:45,835 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 120\n",
      "2025-11-09 19:25:45,835 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:25:45,836 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:25:45,836 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad\n",
      "2025-11-09 19:25:45,836 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:25:45,836 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:25:45,844 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:25:45,844 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:25:46,072 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:25:49,661 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:25:49,661\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:26:24,981\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0115s.\n",
      "2025-11-09 19:26:25,000 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:26:25,001 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.34891048073768616\n",
      "2025-11-09 19:26:25,001 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:26:25,005 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim      lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              2              16  0.0027      0.002403              64   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0               8         512  \n",
      "2025-11-09 19:26:25,007 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:26:25,007 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:26:25,007 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:26:25,007 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:26:25,227 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:26:29,017 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:26:30,175 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.5480\n",
      "2025-11-09 19:26:30,177 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:26:30,179 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:26:31,319 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9887\n",
      "2025-11-09 19:26:31,322 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:26:31,324 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:26:32,490 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9492\n",
      "2025-11-09 19:26:32,493 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:26:32,495 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 0.9887\n",
      "2025-11-09 19:26:32,495 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.8286\n",
      "2025-11-09 19:26:32,495 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.2438\n",
      "2025-11-09 19:26:32,495 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9962 +/- 0.0053\n",
      "F1 Weighted: 0.9962 +/- 0.0053\n",
      "F1 Macro: 0.9962 +/- 0.0053\n"
     ]
    }
   ],
   "source": [
    "output_dir_base_gcn = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_results_GCN_FINAL/paad\")\n",
    "n_repeats = 5\n",
    "all_preds = []\n",
    "\n",
    "for r in range(n_repeats):\n",
    "    bnn.utils.set_seed(SEED+r)\n",
    "    dpmon_repeat = DPMON(\n",
    "        adjacency_matrix=A_train,\n",
    "        omics_list=[dna_meth, rna, cnv_amp, cnv_del],\n",
    "        phenotype_data=target,\n",
    "        clinical_data=clinical_preprocessed,\n",
    "        repeat_num=3,\n",
    "        model='GCN',\n",
    "        tune=True,\n",
    "        gpu=True,\n",
    "        cuda=0,\n",
    "        output_dir=output_dir_base_gcn,\n",
    "    )\n",
    "    \n",
    "    predictions_df, _ = dpmon_repeat.run()\n",
    "    all_preds.append(predictions_df[\"Predicted\"].values)\n",
    "\n",
    "all_preds = np.array(all_preds)\n",
    "\n",
    "\n",
    "f1_macro_list = [f1_score(target, pred, average='macro') for pred in all_preds]\n",
    "f1_weighted_list = [f1_score(target, pred, average='weighted') for pred in all_preds]\n",
    "accuracy_list = [accuracy_score(target, pred) for pred in all_preds]\n",
    "\n",
    "avg_f1_macro = np.mean(f1_macro_list)\n",
    "std_f1_macro = np.std(f1_macro_list)\n",
    "\n",
    "avg_f1_weighted = np.mean(f1_weighted_list)\n",
    "std_f1_weighted = np.std(f1_weighted_list)\n",
    "\n",
    "avg_acc = np.mean(accuracy_list)\n",
    "std_acc = np.std(accuracy_list)\n",
    "\n",
    "print(f\"Accuracy: {avg_acc:.4f} +/- {std_acc:.4f}\")\n",
    "print(f\"F1 Weighted: {avg_f1_weighted:.4f} +/- {std_f1_weighted:.4f}\")\n",
    "print(f\"F1 Macro: {avg_f1_macro:.4f} +/- {std_f1_macro:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d76bbf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 19:26:32,516 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 118\n",
      "2025-11-09 19:26:32,517 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:26:32,517 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:26:32,518 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad\n",
      "2025-11-09 19:26:32,518 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:26:32,518 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:26:32,524 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:26:32,524 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:26:32,748 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:26:36,495 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:26:36,495\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:27:10,528\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0120s.\n",
      "2025-11-09 19:27:10,565 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 2, 'gnn_hidden_dim': 128, 'lr': 0.005197038211089667, 'weight_decay': 0.04607928341516355, 'nn_hidden_dim1': 16, 'nn_hidden_dim2': 32, 'num_epochs': 4096}\n",
      "2025-11-09 19:27:10,566 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.33870771527290344\n",
      "2025-11-09 19:27:10,566 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:27:10,569 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              2             128  0.005197      0.046079              16   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0              32        4096  \n",
      "2025-11-09 19:27:10,571 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 128, 'lr': 0.005197038211089667, 'weight_decay': 0.04607928341516355, 'nn_hidden_dim1': 16, 'nn_hidden_dim2': 32, 'num_epochs': 4096}\n",
      "2025-11-09 19:27:10,571 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 128, 'lr': 0.005197038211089667, 'weight_decay': 0.04607928341516355, 'nn_hidden_dim1': 16, 'nn_hidden_dim2': 32, 'num_epochs': 4096}\n",
      "2025-11-09 19:27:10,572 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:27:10,572 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:27:10,813 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:27:14,433 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:28:30,362 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.7514\n",
      "2025-11-09 19:28:30,365 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:28:30,372 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:29:46,303 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.4746\n",
      "2025-11-09 19:29:46,306 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:29:46,313 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:31:02,325 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.4746\n",
      "2025-11-09 19:31:02,328 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:31:02,335 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 0.7514\n",
      "2025-11-09 19:31:02,336 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.5669\n",
      "2025-11-09 19:31:02,336 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.1598\n",
      "2025-11-09 19:31:02,336 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n",
      "2025-11-09 19:31:02,339 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 119\n",
      "2025-11-09 19:31:02,340 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:31:02,340 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:31:02,341 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad\n",
      "2025-11-09 19:31:02,341 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:31:02,341 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:31:02,346 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:31:02,347 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:31:02,571 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:31:06,350 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:31:06,351\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:31:38,520\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0117s.\n",
      "2025-11-09 19:31:38,546 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:31:38,547 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.31380516290664673\n",
      "2025-11-09 19:31:38,547 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:31:38,551 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim        lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              8              16  0.003115      0.000409              64   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0              64        4096  \n",
      "2025-11-09 19:31:38,552 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:31:38,553 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 8, 'gnn_hidden_dim': 16, 'lr': 0.0031152551982754385, 'weight_decay': 0.0004093955830993196, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 64, 'num_epochs': 4096}\n",
      "2025-11-09 19:31:38,553 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:31:38,553 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:31:38,776 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:31:42,378 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:31:57,582 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:31:57,594 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:31:57,597 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:32:12,782 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:32:12,787 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:32:12,791 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:32:28,239 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9887\n",
      "2025-11-09 19:32:28,251 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:32:28,254 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 1.0000\n",
      "2025-11-09 19:32:28,254 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.9962\n",
      "2025-11-09 19:32:28,254 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.0065\n",
      "2025-11-09 19:32:28,255 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n",
      "2025-11-09 19:32:28,258 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 120\n",
      "2025-11-09 19:32:28,259 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-09 19:32:28,259 - bioneuralnet.utils.data - INFO - Seed setting complete\n",
      "2025-11-09 19:32:28,259 - bioneuralnet.downstream_task.dpmon - INFO - Output directory set to: /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad\n",
      "2025-11-09 19:32:28,260 - bioneuralnet.downstream_task.dpmon - INFO - Initialized DPMON with the provided parameters.\n",
      "2025-11-09 19:32:28,260 - bioneuralnet.downstream_task.dpmon - INFO - Starting DPMON run.\n",
      "2025-11-09 19:32:28,265 - bioneuralnet.downstream_task.dpmon - INFO - Running hyperparameter tuning for DPMON.\n",
      "2025-11-09 19:32:28,266 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:32:28,487 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:32:32,244 - bioneuralnet.downstream_task.dpmon - INFO - Starting hyperparameter tuning for dataset shape: (177, 5628)\n",
      "2025-11-09 19:32:32,245\tINFO tune.py:616 -- [output] This uses the legacy output and progress reporter, as Jupyter notebooks are not supported by the new engine, yet. For more information, please see https://github.com/ray-project/ray/issues/36949\n",
      "2025-11-09 19:33:04,735\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/vicente/ray_results/tune_dp' in 0.0108s.\n",
      "2025-11-09 19:33:04,758 - bioneuralnet.downstream_task.dpmon - INFO - Best trial config: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:33:04,758 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final loss: 0.34250253438949585\n",
      "2025-11-09 19:33:04,759 - bioneuralnet.downstream_task.dpmon - INFO - Best trial final accuracy: 1.0\n",
      "2025-11-09 19:33:04,762 - bioneuralnet.downstream_task.dpmon - INFO -    gnn_layer_num  gnn_hidden_dim      lr  weight_decay  nn_hidden_dim1  \\\n",
      "0              2              16  0.0027      0.002403              64   \n",
      "\n",
      "   nn_hidden_dim2  num_epochs  \n",
      "0               8         512  \n",
      "2025-11-09 19:33:04,764 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:33:04,765 - bioneuralnet.downstream_task.dpmon - INFO - Best tuned parameters: {'gnn_layer_num': 2, 'gnn_hidden_dim': 16, 'lr': 0.0027003956214294257, 'weight_decay': 0.0024031420469003027, 'nn_hidden_dim1': 64, 'nn_hidden_dim2': 8, 'num_epochs': 512}\n",
      "2025-11-09 19:33:04,765 - bioneuralnet.downstream_task.dpmon - INFO - Running standard training with tuned parameters.\n",
      "2025-11-09 19:33:04,765 - bioneuralnet.downstream_task.dpmon - INFO - Using GPU 0\n",
      "2025-11-09 19:33:04,991 - bioneuralnet.downstream_task.dpmon - INFO - Number of nodes in network: 5627\n",
      "2025-11-09 19:33:08,770 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 1/3\n",
      "2025-11-09 19:33:10,704 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 0.9379\n",
      "2025-11-09 19:33:10,707 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_1.pth\n",
      "2025-11-09 19:33:10,709 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 2/3\n",
      "2025-11-09 19:33:12,602 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:33:12,605 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_2.pth\n",
      "2025-11-09 19:33:12,608 - bioneuralnet.downstream_task.dpmon - INFO - Training iteration 3/3\n",
      "2025-11-09 19:33:14,498 - bioneuralnet.downstream_task.dpmon - INFO - Training Accuracy: 1.0000\n",
      "2025-11-09 19:33:14,501 - bioneuralnet.downstream_task.dpmon - INFO - Model saved to /home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/paad/dpm_model_iter_3.pth\n",
      "2025-11-09 19:33:14,503 - bioneuralnet.downstream_task.dpmon - INFO - Best Accuracy: 1.0000\n",
      "2025-11-09 19:33:14,503 - bioneuralnet.downstream_task.dpmon - INFO - Average Accuracy across 3 models: 0.9793\n",
      "2025-11-09 19:33:14,503 - bioneuralnet.downstream_task.dpmon - INFO - Standard Deviation across all models: 0.0359\n",
      "2025-11-09 19:33:14,504 - bioneuralnet.downstream_task.dpmon - INFO - Returning best model predictions and average accuracy (predictions, avg_accuracy).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9171 +/- 0.1172\n",
      "F1 Weighted: 0.9129 +/- 0.1232\n",
      "F1 Macro: 0.9138 +/- 0.1220\n"
     ]
    }
   ],
   "source": [
    "output_dir_base_gat = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_results_GAT_FINAL/paad\")\n",
    "\n",
    "n_repeats = 5\n",
    "all_preds = []\n",
    "\n",
    "for r in range(n_repeats):\n",
    "    bnn.utils.set_seed(SEED+r)\n",
    "    dpmon_repeat = DPMON(\n",
    "        adjacency_matrix=A_train,\n",
    "        omics_list=[dna_meth, rna, cnv_amp, cnv_del],\n",
    "        phenotype_data=target,\n",
    "        clinical_data=clinical_preprocessed,\n",
    "        repeat_num=3,\n",
    "        model='GAT',\n",
    "        tune=True,\n",
    "        gpu=True,\n",
    "        cuda=0,\n",
    "        output_dir=output_dir_base_gat,\n",
    "    )\n",
    "    \n",
    "    predictions_df, _ = dpmon_repeat.run()\n",
    "    all_preds.append(predictions_df[\"Predicted\"].values)\n",
    "\n",
    "all_preds = np.array(all_preds)\n",
    "\n",
    "f1_macro_list = [f1_score(target, pred, average='macro') for pred in all_preds]\n",
    "f1_weighted_list = [f1_score(target, pred, average='weighted') for pred in all_preds]\n",
    "accuracy_list = [accuracy_score(target, pred) for pred in all_preds]\n",
    "\n",
    "avg_f1_macro = np.mean(f1_macro_list)\n",
    "std_f1_macro = np.std(f1_macro_list)\n",
    "\n",
    "avg_f1_weighted = np.mean(f1_weighted_list)\n",
    "std_f1_weighted = np.std(f1_weighted_list)\n",
    "\n",
    "avg_acc = np.mean(accuracy_list)\n",
    "std_acc = np.std(accuracy_list)\n",
    "\n",
    "print(f\"Accuracy: {avg_acc:.4f} +/- {std_acc:.4f}\")\n",
    "print(f\"F1 Weighted: {avg_f1_weighted:.4f} +/- {std_f1_weighted:.4f}\")\n",
    "print(f\"F1 Macro: {avg_f1_macro:.4f} +/- {std_f1_macro:.4f}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d80237ea",
   "metadata": {},
   "source": [
    "## values below are just placeholder for nuw need to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d8f2d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split, RandomizedSearchCV, StratifiedKFold\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.exceptions import ConvergenceWarning\n",
    "from scipy.stats import loguniform, randint\n",
    "\n",
    "X = pd.concat([dna_meth, rna, cnv_amp, cnv_del,clinical_preprocessed], axis=1)\n",
    "y = target['phenotype']\n",
    "print(f\"Successfully created X matrix with shape: {X.shape}\")\n",
    "print(f\"Successfully created y vector with shape: {y.shape}\")\n",
    "\n",
    "all_results = {\n",
    "    \"LogisticRegression\": {\"acc\": [], \"f1_w\": [], \"f1_m\": []},\n",
    "    \"MLP\": {\"acc\": [], \"f1_w\": [], \"f1_m\": []},\n",
    "    \"XGBoost\": {\"acc\": [], \"f1_w\": [], \"f1_m\": []},\n",
    "}\n",
    "\n",
    "all_results = {\n",
    "    \"LogisticRegression\": {\"acc\": [], \"f1_w\": [], \"f1_m\": []},\n",
    "    \"MLP\": {\"acc\": [], \"f1_w\": [], \"f1_m\": []},\n",
    "    \"XGBoost\": {\"acc\": [], \"f1_w\": [], \"f1_m\": []},\n",
    "}\n",
    "\n",
    "N_REPEATS = 5\n",
    "TEST_SPLIT_SIZE = .7\n",
    "CV_FOLDS = 3\n",
    "N_ITER_SEARCH = 10\n",
    "\n",
    "pipe_lr = Pipeline([('scaler', StandardScaler()),\n",
    "    ('model', LogisticRegression(\n",
    "        solver='lbfgs',\n",
    "        max_iter=1000,\n",
    "        penalty=None \n",
    "    ))\n",
    "])\n",
    "\n",
    "pipe_mlp = Pipeline([('scaler', StandardScaler()),\n",
    "    ('model', MLPClassifier(\n",
    "        max_iter=500,\n",
    "        early_stopping=True,\n",
    "        n_iter_no_change=10\n",
    "    ))\n",
    "])\n",
    "\n",
    "pipe_xgb = Pipeline([('scaler', StandardScaler()),\n",
    "    ('model', XGBClassifier(\n",
    "        eval_metric='logloss'\n",
    "    ))\n",
    "])\n",
    "\n",
    "params_lr = {\n",
    "    'model__penalty': ['l2'], \n",
    "    'model__C': loguniform(1e-4, 1e2)\n",
    "}\n",
    "\n",
    "params_mlp = {\n",
    "    'model__hidden_layer_sizes': [(100,), (100, 50), (50, 50)],\n",
    "    'model__activation': ['relu', 'tanh'],\n",
    "    'model__alpha': loguniform(1e-5, 1e-1),\n",
    "    'model__learning_rate_init': loguniform(1e-4, 1e-2)\n",
    "}\n",
    "\n",
    "params_xgb = {\n",
    "    'model__n_estimators': randint(100, 500),\n",
    "    'model__learning_rate': loguniform(0.01, 0.3),\n",
    "    'model__max_depth': randint(3, 10),\n",
    "    'model__subsample': [0.7, 0.8, 0.9, 1.0],\n",
    "    'model__colsample_bytree': [0.7, 0.8, 0.9, 1.0]\n",
    "}\n",
    "\n",
    "models_to_tune = {\n",
    "    \"LogisticRegression\": (pipe_lr, params_lr),\n",
    "    \"MLP\": (pipe_mlp, params_mlp),\n",
    "    \"XGBoost\": (pipe_xgb, params_xgb)\n",
    "}\n",
    "\n",
    "for r in range(N_REPEATS):\n",
    "    seed = SEED + r\n",
    "    print(f\"\\nRunning Repeat {r+1}/{N_REPEATS} (Seed: {seed})\")\n",
    "\n",
    "    X_train, X_test, y_train, y_test = train_test_split(\n",
    "        X, y, \n",
    "        test_size=TEST_SPLIT_SIZE, \n",
    "        random_state=seed, \n",
    "        stratify=y)\n",
    "    \n",
    "    cv_splitter = StratifiedKFold(n_splits=CV_FOLDS, shuffle=True, random_state=seed)\n",
    "    for name, (pipeline, params) in models_to_tune.items():\n",
    "        print(f\"Tuning {name}\")\n",
    "        search = RandomizedSearchCV(\n",
    "            estimator=pipeline,\n",
    "            param_distributions=params,\n",
    "            n_iter=N_ITER_SEARCH,\n",
    "            cv=cv_splitter,\n",
    "            scoring='f1_weighted',\n",
    "            n_jobs=-1,\n",
    "            random_state=seed\n",
    "        )\n",
    "        \n",
    "        with warnings.catch_warnings():\n",
    "            warnings.filterwarnings(\"ignore\", category=ConvergenceWarning)\n",
    "            warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "            search.fit(X_train, y_train)\n",
    "\n",
    "        print(f\"Best params for {name}: {search.best_params_}\")\n",
    "        \n",
    "        best_model = search.best_estimator_\n",
    "        preds = best_model.predict(X_test)\n",
    "\n",
    "        acc = accuracy_score(y_test, preds)\n",
    "        f1_w = f1_score(y_test, preds, average='weighted', zero_division=0)\n",
    "        f1_m = f1_score(y_test, preds, average='macro', zero_division=0)\n",
    "\n",
    "        all_results[name][\"acc\"].append(acc)\n",
    "        all_results[name][\"f1_w\"].append(f1_w)\n",
    "        all_results[name][\"f1_m\"].append(f1_m)\n",
    "\n",
    "print(f\"Tuned Model Results (Averaged over {N_REPEATS} runs)\")\n",
    "print(f\"(Tuning was {N_ITER_SEARCH} iterations with {CV_FOLDS}fold CV)\")\n",
    "\n",
    "for model_name, metrics in all_results.items():\n",
    "    avg_acc = np.mean(metrics[\"acc\"])\n",
    "    std_acc = np.std(metrics[\"acc\"])\n",
    "    \n",
    "    avg_f1_w = np.mean(metrics[\"f1_w\"])\n",
    "    std_f1_w = np.std(metrics[\"f1_w\"])\n",
    "    \n",
    "    avg_f1_m = np.mean(metrics[\"f1_m\"])\n",
    "    std_f1_m = np.std(metrics[\"f1_m\"])\n",
    "\n",
    "    print(f\"Results for {model_name}:\")\n",
    "    print(f\"Accuracy: {avg_acc:.4f} +/- {std_acc:.4f}\")\n",
    "    print(f\"F1 Weighted: {avg_f1_w:.4f} +/- {std_f1_w:.4f}\")\n",
    "    print(f\"F1 Macro: {avg_f1_m:.4f} +/- {std_f1_m:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c265a190",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "gnn_plot_data = {\n",
    "    \"Accuracy\": {\n",
    "        \"SAGE\": (0.9903, 0.0173), \n",
    "        \"GCN\": (0.9720, 0.0379), \n",
    "        \"GAT\": (0.9626, 0.0575)\n",
    "    },\n",
    "    \"F1 Weighted\": {\n",
    "        \"SAGE\": (0.9903, 0.0173), \n",
    "        \"GCN\": (0.9631, 0.0557), \n",
    "        \"GAT\": (0.9546, 0.0732)\n",
    "    },\n",
    "    \"F1 Macro\": {\n",
    "        \"SAGE\": (0.9876, 0.0203), \n",
    "        \"GCN\": (0.9150, 0.1410), \n",
    "        \"GAT\": (0.9107, 0.1569)\n",
    "    }\n",
    "}\n",
    "\n",
    "baseline_plot_data = {\n",
    "    \"Accuracy\": {\n",
    "        \"SAGE\": (0.9903, 0.0173), \n",
    "        \"LogReg\": (0.9553, 0.0090), \n",
    "        \"XGBoost\": (0.9527, 0.0059), \n",
    "        \"MLP\": (0.9362, 0.0133)\n",
    "    },\n",
    "    \"F1 Weighted\": {\n",
    "        \"SAGE\": (0.9903, 0.0173), \n",
    "        \"LogReg\": (0.9557, 0.0088), \n",
    "        \"XGBoost\": (0.9529, 0.0058), \n",
    "        \"MLP\": (0.9379, 0.0125)\n",
    "    },\n",
    "    \"F1 Macro\": {\n",
    "        \"SAGE\": (0.9876, 0.0203), \n",
    "        \"LogReg\": (0.9413, 0.0124), \n",
    "        \"XGBoost\": (0.9451, 0.0106), \n",
    "        \"MLP\": (0.9138, 0.0170)\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "bnn.metrics.plot_multiple_metrics(\n",
    "    gnn_plot_data,\n",
    "    title_map={\n",
    "        \"Accuracy\": \"GNNs Comparison: Accuracy\",\n",
    "        \"F1 Weighted\": \"GNNs Comparison: F1 Weighted\",\n",
    "        \"F1 Macro\": \"GNNs Comparison: F1 Macro\"\n",
    "    }\n",
    ")\n",
    "\n",
    "bnn.metrics.plot_multiple_metrics(\n",
    "    baseline_plot_data,\n",
    "    title_map={\n",
    "        \"Accuracy\": \"SAGE vs. Baselines: Accuracy\",\n",
    "        \"F1 Weighted\": \"SAGE vs. Baselines: F1 Weighted\",\n",
    "        \"F1 Macro\": \"SAGE vs. Baselines: F1 Macro\"\n",
    "    }\n",
    ")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".enviroment",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
