{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d408841e",
   "metadata": {},
   "source": [
    "# TCGA-KIPAN Analysis Demo\n",
    "\n",
    "- **Cohort**: Focuses on the TCGA-KIPAN dataset, a vital resource merging three major kidney cancer subtypes:\n",
    "    - `Kidney Renal Clear Cell Carcinoma` (KIRC)\n",
    "    - `Kidney Renal Papillary Cell Carcinoma` (KIRP)\n",
    "    - `Kidney Chromophobe `(KICH)\n",
    "\n",
    "- **Goal**: Perform histological subtype classification.\n",
    "- **Prediction Target**: Predict the specific kidney cancer subtype (`KIRC`, `KIRP`, or `KICH`) from its multi-omics profile.\n",
    "\n",
    "**Data Source:** Broad Institute FireHose (`http://firebrowse.org/?cohort=KIPAN`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c56af27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KL-8324-11</th>\n",
       "      <th>TCGA-KL-8324-01</th>\n",
       "      <th>TCGA-KL-8325-01</th>\n",
       "      <th>TCGA-KL-8326-11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <td>13.350564</td>\n",
       "      <td>12.669290</td>\n",
       "      <td>12.138842</td>\n",
       "      <td>12.186424</td>\n",
       "      <td>12.251624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <td>14.345422</td>\n",
       "      <td>13.671356</td>\n",
       "      <td>13.139199</td>\n",
       "      <td>13.182016</td>\n",
       "      <td>13.244735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>13.354717</td>\n",
       "      <td>12.696054</td>\n",
       "      <td>12.157156</td>\n",
       "      <td>12.178790</td>\n",
       "      <td>12.260827</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              TCGA-KL-8323-01  TCGA-KL-8324-11  TCGA-KL-8324-01  \\\n",
       "gene                                                              \n",
       "hsa-let-7a-1        13.350564        12.669290        12.138842   \n",
       "hsa-let-7a-2        14.345422        13.671356        13.139199   \n",
       "hsa-let-7a-3        13.354717        12.696054        12.157156   \n",
       "\n",
       "              TCGA-KL-8325-01  TCGA-KL-8326-11  \n",
       "gene                                            \n",
       "hsa-let-7a-1        12.186424        12.251624  \n",
       "hsa-let-7a-2        13.182016        13.244735  \n",
       "hsa-let-7a-3        12.178790        12.260827  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(472, 1005)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KL-8324-11</th>\n",
       "      <th>TCGA-KL-8324-01</th>\n",
       "      <th>TCGA-KL-8325-01</th>\n",
       "      <th>TCGA-KL-8326-11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>?|100133144</th>\n",
       "      <td>3.145221</td>\n",
       "      <td>2.847776</td>\n",
       "      <td>3.356102</td>\n",
       "      <td>2.591009</td>\n",
       "      <td>2.097307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|100134869</th>\n",
       "      <td>2.462733</td>\n",
       "      <td>2.555203</td>\n",
       "      <td>4.644127</td>\n",
       "      <td>3.272561</td>\n",
       "      <td>3.044499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|10357</th>\n",
       "      <td>6.624215</td>\n",
       "      <td>6.822777</td>\n",
       "      <td>5.916350</td>\n",
       "      <td>6.813293</td>\n",
       "      <td>6.846008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TCGA-KL-8323-01  TCGA-KL-8324-11  TCGA-KL-8324-01  \\\n",
       "gene                                                             \n",
       "?|100133144         3.145221         2.847776         3.356102   \n",
       "?|100134869         2.462733         2.555203         4.644127   \n",
       "?|10357             6.624215         6.822777         5.916350   \n",
       "\n",
       "             TCGA-KL-8325-01  TCGA-KL-8326-11  \n",
       "gene                                           \n",
       "?|100133144         2.591009         2.097307  \n",
       "?|100134869         3.272561         3.044499  \n",
       "?|10357             6.813293         6.846008  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20117, 867)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KL-8324-01</th>\n",
       "      <th>TCGA-KL-8325-01</th>\n",
       "      <th>TCGA-KL-8326-01</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hybridization REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.586272204702</td>\n",
       "      <td>0.258615822462</td>\n",
       "      <td>0.346998160003</td>\n",
       "      <td>0.303477029822</td>\n",
       "      <td>0.31424190458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.612681857814</td>\n",
       "      <td>0.585391864334</td>\n",
       "      <td>0.469213665938</td>\n",
       "      <td>0.572474360793</td>\n",
       "      <td>0.595785276504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      TCGA-KL-8323-01 TCGA-KL-8324-01 TCGA-KL-8325-01  \\\n",
       "Hybridization REF                                                       \n",
       "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
       "A1BG                   0.586272204702  0.258615822462  0.346998160003   \n",
       "A1CF                   0.612681857814  0.585391864334  0.469213665938   \n",
       "\n",
       "                      TCGA-KL-8326-01 TCGA-KL-8327-01  \n",
       "Hybridization REF                                      \n",
       "Composite Element REF      Beta_Value      Beta_Value  \n",
       "A1BG                   0.303477029822   0.31424190458  \n",
       "A1CF                   0.572474360793  0.595785276504  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20117, 867)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcga-kl-8328</th>\n",
       "      <th>tcga-kl-8339</th>\n",
       "      <th>tcga-km-8439</th>\n",
       "      <th>tcga-km-8441</th>\n",
       "      <th>tcga-km-8442</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hybridization REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>years_to_birth</th>\n",
       "      <td>60</td>\n",
       "      <td>67</td>\n",
       "      <td>30</td>\n",
       "      <td>61</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vital_status</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tcga-kl-8328 tcga-kl-8339 tcga-km-8439 tcga-km-8441  \\\n",
       "Hybridization REF                                                           \n",
       "Composite Element REF        value        value        value        value   \n",
       "years_to_birth                  60           67           30           61   \n",
       "vital_status                     0            1            0            0   \n",
       "\n",
       "                      tcga-km-8442  \n",
       "Hybridization REF                   \n",
       "Composite Element REF        value  \n",
       "years_to_birth                  38  \n",
       "vital_status                     0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20, 941)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "root = Path(\"/home/vicente/Github/BioNeuralNet/KIPAN\")\n",
    "\n",
    "mirna_raw = pd.read_csv(root/\"KIPAN.miRseq_RPKM_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)                            \n",
    "rna_raw = pd.read_csv(root / \"KIPAN.uncv2.mRNAseq_RSEM_normalized_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)\n",
    "meth_raw = pd.read_csv(root/\"KIPAN.meth.by_mean.data.txt\", sep='\\t',index_col=0,low_memory=False)\n",
    "clinical_raw = pd.read_csv(root / \"KIPAN.clin.merged.picked.txt\",sep=\"\\t\", index_col=0, low_memory=False)\n",
    "\n",
    "# display all shapes and first few rows of each dataset\n",
    "display(mirna_raw.iloc[:3,:5])\n",
    "display(mirna_raw.shape)\n",
    "\n",
    "display(rna_raw.iloc[:3,:5])\n",
    "display(meth_raw.shape)\n",
    "\n",
    "display(meth_raw.iloc[:3,:5])\n",
    "display(meth_raw.shape)\n",
    "\n",
    "display(clinical_raw.iloc[:3,:5])\n",
    "display(clinical_raw.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13c6054",
   "metadata": {},
   "source": [
    "## Data Processing Summary\n",
    "\n",
    "1. **Transpose Data:** All raw data (miRNA, RNA, etc.) is flipped so rows represent patients and columns represent features.\n",
    "2. **Standardize Patient IDs:** Patient IDs in all tables are cleaned to the 12-character TCGA format (e.g., `TCGA-AB-1234`) for matching.\n",
    "3. **Handle Duplicates:** Duplicate patient rows are averaged in the omics data. The first entry is kept for duplicate patients in the clinical data.\n",
    "4. **Find Common Patients:** The script identifies the list of patients that exist in *all* datasets.\n",
    "5. **Subset Data:** All data tables are filtered down to *only* this common list of patients, ensuring alignment.\n",
    "6. **Extract Target:** The `histological_type` column is pulled from the processed clinical data to be used as the prediction target (y-variable)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33acffac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miRNA (samples, features): (1005, 472)\n",
      "RNA (samples, features): (1020, 18272)\n",
      "Methylation (samples, features): (867, 20117)\n",
      "Clinical (samples, features): (941, 20)\n",
      "\n",
      "Methylation shape: (660, 20117)\n",
      "RNA shape: (889, 18272)\n",
      "miRNA shape: (873, 472)\n",
      "Clinical shape: (941, 20)\n",
      "\n",
      "Found: 658 patients across all data types.\n"
     ]
    }
   ],
   "source": [
    "mirna = mirna_raw.T\n",
    "rna = rna_raw.T\n",
    "meth = meth_raw.T\n",
    "clinical = clinical_raw.T\n",
    "\n",
    "print(f\"miRNA (samples, features): {mirna.shape}\")\n",
    "print(f\"RNA (samples, features): {rna.shape}\")\n",
    "print(f\"Methylation (samples, features): {meth.shape}\")\n",
    "print(f\"Clinical (samples, features): {clinical.shape}\")\n",
    "\n",
    "def trim_barcode(idx):\n",
    "    return idx.to_series().str.slice(0, 12)\n",
    "\n",
    "# standarized patient IDs across all files\n",
    "meth.index = trim_barcode(meth.index)\n",
    "rna.index = trim_barcode(rna.index)\n",
    "mirna.index = trim_barcode(mirna.index)\n",
    "clinical.index = clinical.index.str.upper()\n",
    "clinical.index.name = \"Patient_ID\"\n",
    "\n",
    "# convert all data to numeric, coercing errors to NaN\n",
    "meth = meth.apply(pd.to_numeric, errors='coerce')\n",
    "rna = rna.apply(pd.to_numeric, errors='coerce')\n",
    "mirna = mirna.apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "# for any duplicate columns in the omics data, we average their values\n",
    "meth = meth.groupby(meth.index).mean()\n",
    "rna = rna.groupby(rna.index).mean()\n",
    "mirna = mirna.groupby(mirna.index).mean()\n",
    "\n",
    "# for any duplicate rows in the clinical data, we keep the first occurrence\n",
    "clinical = clinical[~clinical.index.duplicated(keep='first')]\n",
    "\n",
    "print(f\"\\nMethylation shape: {meth.shape}\")\n",
    "print(f\"RNA shape: {rna.shape}\")\n",
    "print(f\"miRNA shape: {mirna.shape}\")\n",
    "print(f\"Clinical shape: {clinical.shape}\")\n",
    "\n",
    "for df in [meth, rna, mirna]:\n",
    "    df.columns = df.columns.str.replace(r\"\\?\", \"unknown_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(r\"\\|\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(\"-\", \"_\", regex=False)\n",
    "    df.columns = df.columns.str.replace(r\"_+\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.strip(\"_\")\n",
    "    \n",
    "    df.fillna(df.mean(), inplace=True)\n",
    "\n",
    "# to see which pateints are common across all data files\n",
    "common_patients = sorted(list(set(meth.index)&set(rna.index)&set(mirna.index)&set(clinical.index)))\n",
    "\n",
    "print(f\"\\nFound: {len(common_patients)} patients across all data types.\")\n",
    "\n",
    "# subset to only common patients\n",
    "meth_processed = meth.loc[common_patients]\n",
    "rna_processed= rna.loc[common_patients]\n",
    "mirna_processed = mirna.loc[common_patients]\n",
    "clinical_processed = clinical.loc[common_patients]\n",
    "\n",
    "# extract target labels from clinical data\n",
    "targets = clinical_processed['histological_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4ca526c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>hsa_let_7a_1</th>\n",
       "      <th>hsa_let_7a_2</th>\n",
       "      <th>hsa_let_7a_3</th>\n",
       "      <th>hsa_let_7b</th>\n",
       "      <th>hsa_let_7c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2K-A9WE</th>\n",
       "      <td>12.933499</td>\n",
       "      <td>13.933025</td>\n",
       "      <td>12.938528</td>\n",
       "      <td>12.861969</td>\n",
       "      <td>11.474055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J1</th>\n",
       "      <td>12.535658</td>\n",
       "      <td>13.536437</td>\n",
       "      <td>12.531655</td>\n",
       "      <td>12.710724</td>\n",
       "      <td>10.355773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J2</th>\n",
       "      <td>11.832278</td>\n",
       "      <td>12.838388</td>\n",
       "      <td>11.840725</td>\n",
       "      <td>11.038718</td>\n",
       "      <td>8.360210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene          hsa_let_7a_1  hsa_let_7a_2  hsa_let_7a_3  hsa_let_7b  hsa_let_7c\n",
       "TCGA-2K-A9WE     12.933499     13.933025     12.938528   12.861969   11.474055\n",
       "TCGA-2Z-A9J1     12.535658     13.536437     12.531655   12.710724   10.355773\n",
       "TCGA-2Z-A9J2     11.832278     12.838388     11.840725   11.038718    8.360210"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(658, 472)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>unknown_100133144</th>\n",
       "      <th>unknown_100134869</th>\n",
       "      <th>unknown_10357</th>\n",
       "      <th>unknown_10431</th>\n",
       "      <th>unknown_155060</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2K-A9WE</th>\n",
       "      <td>2.336112</td>\n",
       "      <td>2.520498</td>\n",
       "      <td>5.772965</td>\n",
       "      <td>9.610685</td>\n",
       "      <td>8.198804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J1</th>\n",
       "      <td>3.006962</td>\n",
       "      <td>3.558929</td>\n",
       "      <td>6.177374</td>\n",
       "      <td>10.177077</td>\n",
       "      <td>7.656137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J2</th>\n",
       "      <td>1.516973</td>\n",
       "      <td>1.736691</td>\n",
       "      <td>4.853207</td>\n",
       "      <td>10.345265</td>\n",
       "      <td>6.263288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene          unknown_100133144  unknown_100134869  unknown_10357  \\\n",
       "TCGA-2K-A9WE           2.336112           2.520498       5.772965   \n",
       "TCGA-2Z-A9J1           3.006962           3.558929       6.177374   \n",
       "TCGA-2Z-A9J2           1.516973           1.736691       4.853207   \n",
       "\n",
       "gene          unknown_10431  unknown_155060  \n",
       "TCGA-2K-A9WE       9.610685        8.198804  \n",
       "TCGA-2Z-A9J1      10.177077        7.656137  \n",
       "TCGA-2Z-A9J2      10.345265        6.263288  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(658, 18272)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hybridization REF</th>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2BP1</th>\n",
       "      <th>A2LD1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2K-A9WE</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.498835</td>\n",
       "      <td>0.814418</td>\n",
       "      <td>0.536619</td>\n",
       "      <td>0.771750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.400956</td>\n",
       "      <td>0.554575</td>\n",
       "      <td>0.517050</td>\n",
       "      <td>0.505130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.438116</td>\n",
       "      <td>0.656936</td>\n",
       "      <td>0.535795</td>\n",
       "      <td>0.678655</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Hybridization REF  Composite Element REF      A1BG      A1CF     A2BP1  \\\n",
       "TCGA-2K-A9WE                         NaN  0.498835  0.814418  0.536619   \n",
       "TCGA-2Z-A9J1                         NaN  0.400956  0.554575  0.517050   \n",
       "TCGA-2Z-A9J2                         NaN  0.438116  0.656936  0.535795   \n",
       "\n",
       "Hybridization REF     A2LD1  \n",
       "TCGA-2K-A9WE       0.771750  \n",
       "TCGA-2Z-A9J1       0.505130  \n",
       "TCGA-2Z-A9J2       0.678655  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(658, 20117)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hybridization REF</th>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_followup</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2K-A9WE</th>\n",
       "      <td>value</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J1</th>\n",
       "      <td>value</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J2</th>\n",
       "      <td>value</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Hybridization REF Composite Element REF years_to_birth vital_status  \\\n",
       "Patient_ID                                                            \n",
       "TCGA-2K-A9WE                      value             53            0   \n",
       "TCGA-2Z-A9J1                      value             71            0   \n",
       "TCGA-2Z-A9J2                      value             71            0   \n",
       "\n",
       "Hybridization REF days_to_death days_to_last_followup  \n",
       "Patient_ID                                             \n",
       "TCGA-2K-A9WE                NaN                   214  \n",
       "TCGA-2Z-A9J1                NaN                  2298  \n",
       "TCGA-2Z-A9J2                NaN                  1795  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(658, 20)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "histological_type\n",
       "kidney clear cell renal carcinoma        318\n",
       "kidney papillary renal cell carcinoma    274\n",
       "kidney chromophobe                        66\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(mirna_processed.iloc[:3,:5])\n",
    "display(mirna_processed.shape)\n",
    "\n",
    "display(rna_processed.iloc[:3,:5])\n",
    "display(rna_processed.shape)\n",
    "\n",
    "display(meth_processed.iloc[:3,:5])\n",
    "display(meth_processed.shape)\n",
    "\n",
    "display(clinical_processed.iloc[:3,:5])\n",
    "display(clinical_processed.shape)\n",
    "\n",
    "display(targets.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e924114f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 15:12:24,843 - bioneuralnet.utils.data - INFO - Starting Beta-to-M value conversion (shape: (658, 20116)). Epsilon: 1e-06\n",
      "2025-11-08 15:12:25,944 - bioneuralnet.utils.data - INFO - Beta-to-M conversion complete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DNA_Methylation shape: (658, 20116)\n",
      "RNA shape: (658, 18272)\n",
      "miRNA shape: (658, 472)\n",
      "Clinical shape: (658, 19)\n",
      "target\n",
      "0         318\n",
      "1         274\n",
      "2          66\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "# drop unwanted columns from clinical data\n",
    "clinical_processed.drop(columns=[\"Composite Element REF\"], errors=\"ignore\", inplace=True)\n",
    "\n",
    "# we transform the methylation beta values to M-values and drop unwanted columns\n",
    "meth_m = meth_processed.drop(columns=[\"Composite Element REF\"], errors=\"ignore\")\n",
    "\n",
    "# convert beta values to M-values using bioneuralnet utility with small epsilon to avoid log(0)\n",
    "meth_m = bnn.utils.beta_to_m(meth_m, eps=1e-6) \n",
    "\n",
    "# lastly we turn the target labels into numerical classes\n",
    "mapping = {\"kidney clear cell renal carcinoma\": 0, \"kidney papillary renal cell carcinoma\": 1, \"kidney chromophobe\": 2}\n",
    "target_labels = targets.map(mapping).to_frame(name=\"target\")\n",
    "\n",
    "# as a safety check we align the indices once more\n",
    "X_meth = meth_m.loc[common_patients]\n",
    "X_rna = rna_processed.loc[common_patients]\n",
    "X_mirna = mirna_processed.loc[common_patients]\n",
    "Y_labels = target_labels.loc[common_patients]\n",
    "clinical_final = clinical_processed.loc[common_patients]\n",
    "\n",
    "print(f\"\\nDNA_Methylation shape: {X_meth.shape}\")\n",
    "print(f\"RNA shape: {X_rna.shape}\")\n",
    "print(f\"miRNA shape: {X_mirna.shape}\")\n",
    "print(f\"Clinical shape: {clinical_final.shape}\")\n",
    "print(Y_labels.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "173f6145",
   "metadata": {},
   "source": [
    "## Feature Selection Methodology\n",
    "\n",
    "### Supported Methods and Interpretation\n",
    "\n",
    "**BioNeuralNet** provides three techniques for feature selection, allowing for different views of the data's statistical profile:\n",
    "\n",
    "- **Variance Thresholding:** Identifies features with the **highest overall variance** across all samples.\n",
    "\n",
    "- **ANOVA F-test:** Pinpoints features that best **distinguish between the target classes** (KIRC, KIRP, and KICH).\n",
    "\n",
    "- **Random Forest Importance:** Assesses **feature utility** based on its contribution to a predictive non-linear model.\n",
    "\n",
    "### KIPAN Cohort Selection Strategy\n",
    "\n",
    "A dimensionality reduction step was essential for managing the high-feature-count omics data:\n",
    "\n",
    "- **High-Feature Datasets:** Both DNA Methylation (20,116) and RNA (18,272) required significant feature reduction.\n",
    "\n",
    "- **Filtering Process:** The **top 6,000 features** were initially extracted from the Methylation and RNA datasets using all three methods.\n",
    "\n",
    "- **Final Set:** A consensus set was built by finding the intersection of features selected by the ANOVA F-test and Random Forest Importance, ensuring both statistical relevance and model-based utility.\n",
    "\n",
    "- **Low-Feature Datasets:** The miRNA data (472 features) was passed through **without selection**, as its feature count was already manageable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "241a4bb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 15:12:53,133 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:12:53,133 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 15:12:53,134 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 15:12:53,224 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-08 15:12:55,954 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:12:55,955 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 15:12:55,955 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 15:12:56,036 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-08 15:12:58,995 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:12:58,995 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 15:12:58,995 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 15:12:59,135 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 18558 significant, 0 padded\n",
      "2025-11-08 15:13:01,868 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:13:01,868 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 15:13:01,869 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 15:13:01,995 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 17137 significant, 0 padded\n",
      "2025-11-08 15:13:05,056 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:13:05,056 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 15:13:05,056 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 15:13:10,605 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:13:10,605 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 15:13:10,606 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "# feature selection\n",
    "meth_highvar = bnn.utils.select_top_k_variance(X_meth, k=6000)\n",
    "rna_highvar = bnn.utils.select_top_k_variance(X_rna, k=6000)\n",
    "\n",
    "meth_af = bnn.utils.top_anova_f_features(X_meth, Y_labels, max_features=6000)\n",
    "rna_af = bnn.utils.top_anova_f_features(X_rna, Y_labels, max_features=6000)\n",
    "\n",
    "meth_rf = bnn.utils.select_top_randomforest(X_meth, Y_labels, top_k=6000)\n",
    "rna_rf = bnn.utils.select_top_randomforest(X_rna, Y_labels, top_k=6000)\n",
    "\n",
    "meth_var_set = set(meth_highvar.columns)\n",
    "meth_anova_set = set(meth_af.columns)\n",
    "meth_rf_set = set(meth_rf.columns)\n",
    "\n",
    "rna_var_set = set(rna_highvar.columns)\n",
    "rna_anova_set = set(rna_af.columns)\n",
    "rna_rf_set = set(rna_rf.columns)\n",
    "\n",
    "meth_inter1 = list(meth_anova_set & meth_var_set)\n",
    "meth_inter2 = list(meth_rf_set & meth_var_set)\n",
    "meth_inter3 = list(meth_anova_set & meth_rf_set)\n",
    "meth_all_three = list(meth_anova_set & meth_var_set & meth_rf_set)\n",
    "\n",
    "rna_inter4 = list(rna_anova_set & rna_var_set)\n",
    "rna_inter5 = list(rna_rf_set & rna_var_set)\n",
    "rna_inter6 = list(rna_anova_set & rna_rf_set)\n",
    "rna_all_three = list(rna_anova_set & rna_var_set & rna_rf_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b02d2358",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FROM THE 6000 Methylation feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 1875 features\n",
      "Random Forest & variance selection share: 1656 features\n",
      "Anova-F & Random Forest share: 2102 features\n",
      "All three methods agree on: 666 features\n"
     ]
    }
   ],
   "source": [
    "print(\"FROM THE 6000 Methylation feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(meth_inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(meth_inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(meth_inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(meth_all_three)} features\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3abea9c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "FROM THE 6000 RNA feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 2271 features\n",
      "Random Forest & variance selection share: 2141 features\n",
      "Anova-F & Random Forest share: 2284 features\n",
      "All three methods agree on: 943 features\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nFROM THE 6000 RNA feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(rna_inter4)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(rna_inter5)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(rna_inter6)} features\")\n",
    "print(f\"All three methods agree on: {len(rna_all_three)} features\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee3a3e98",
   "metadata": {},
   "source": [
    "## Feature Selection Summary: ANOVA-RF Intersection\n",
    "\n",
    "The final set of features was determined by the **intersection** of those highlighted by the **ANOVA F-test** and **Random Forest Importance**. This methodology provides a balanced filter, capturing features with both high class-separability (ANOVA) and significant predictive value in a non-linear model (Random Forest). The resulting feature pool is considered highly relevant for the subsequent modeling tasks.\n",
    "\n",
    "### Feature Overlap Results\n",
    "\n",
    "The table below quantifies the shared features identified by the different selection techniques for each omics type.\n",
    "\n",
    "| Omics Data Type | ANOVA-F & Variance | RF & Variance | ANOVA-F & Random Forest (Selected) | All Three Agree |\n",
    "| :--- | :--- | :--- | :--- | :--- |\n",
    "| **Methylation** | 1,875 features | 1,656 features | **2,102 features** | 666 features |\n",
    "| **RNA** | 2,271 features | 2,141 features | **2,284 features** | 943 features |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cba88e86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Final Shapes for Modeling\n",
      "Methylation (X1): (658, 2102)\n",
      "RNA-Seq (X2): (658, 2284)\n",
      "miRNA-Seq (X3): (658, 472)\n",
      "Labels (Y): (658, 1)\n"
     ]
    }
   ],
   "source": [
    "X_meth_selected = X_meth[meth_inter3]\n",
    "X_rna_selected = X_rna[rna_inter6]\n",
    "\n",
    "print(\"\\nFinal Shapes for Modeling\")\n",
    "print(f\"Methylation (X1): {X_meth_selected.shape}\")\n",
    "print(f\"RNA-Seq (X2): {X_rna_selected.shape}\")\n",
    "print(f\"miRNA-Seq (X3): {X_mirna.shape}\")\n",
    "print(f\"Labels (Y): {Y_labels.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0d52489",
   "metadata": {},
   "source": [
    "## Data Availability\n",
    "\n",
    "To facilitate rapid experimentation and reproduction of our results, the fully processed and feature-selected dataset used in this analysis has been made available directly within the package.\n",
    "\n",
    "Users can load this dataset, bypassing all preceding data acquisition, preprocessing, and feature selection steps. This allows users to proceed immediately from this step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e99139",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mirna': (658, 472),\n",
       " 'target': (658, 1),\n",
       " 'clinical': (658, 19),\n",
       " 'rna': (658, 2284),\n",
       " 'meth': (658, 2102)}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "tgca_kipan = bnn.datasets.DatasetLoader(\"kipan\")\n",
    "display(tgca_kipan.shape)\n",
    "\n",
    "# The dataset is returned as a dictionary. We extract each file independetly based on the name (Key).\n",
    "dna_meth = tgca_kipan.data[\"meth\"]\n",
    "rna = tgca_kipan.data[\"rna\"]\n",
    "mirna = tgca_kipan.data[\"mirna\"]\n",
    "clinical = tgca_kipan.data[\"clinical\"]\n",
    "target = tgca_kipan.data[\"target\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39d9b227",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 15:17:58,736 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 15:17:58,737 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 2306 NaNs after median imputation\n",
      "2025-11-08 15:17:58,737 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 15:17:58,826 - bioneuralnet.utils.preprocess - INFO - Selected top 7 features by RandomForest importance\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>histological_type_kidney clear cell renal carcinoma</th>\n",
       "      <th>histological_type_kidney papillary renal cell carcinoma</th>\n",
       "      <th>radiation_therapy_no</th>\n",
       "      <th>pathology_M_stage_mx</th>\n",
       "      <th>ethnicity_not hispanic or latino</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2K-A9WE</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J1</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Z-A9J2</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              histological_type_kidney clear cell renal carcinoma  \\\n",
       "Patient_ID                                                          \n",
       "TCGA-2K-A9WE                                              False     \n",
       "TCGA-2Z-A9J1                                              False     \n",
       "TCGA-2Z-A9J2                                              False     \n",
       "\n",
       "              histological_type_kidney papillary renal cell carcinoma  \\\n",
       "Patient_ID                                                              \n",
       "TCGA-2K-A9WE                                               True         \n",
       "TCGA-2Z-A9J1                                               True         \n",
       "TCGA-2Z-A9J2                                               True         \n",
       "\n",
       "              radiation_therapy_no  pathology_M_stage_mx  \\\n",
       "Patient_ID                                                 \n",
       "TCGA-2K-A9WE                  True                 False   \n",
       "TCGA-2Z-A9J1                  True                  True   \n",
       "TCGA-2Z-A9J2                  True                  True   \n",
       "\n",
       "              ethnicity_not hispanic or latino  \n",
       "Patient_ID                                      \n",
       "TCGA-2K-A9WE                              True  \n",
       "TCGA-2Z-A9J1                              True  \n",
       "TCGA-2Z-A9J2                              True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# BioNeuralNet provides a preprocessing function to handle clinical data\n",
    "clinical = tgca_kipan.data[\"clinical\"]\n",
    "\n",
    "# for more details on the preprocessing functions, see `bioneuralnet.utils.preprocess``\n",
    "clinical_preprocessed = bnn.utils.preprocess_clinical(\n",
    "    clinical, \n",
    "    target, \n",
    "    top_k=7, \n",
    "    scale=False, \n",
    "    ignore_columns=[ \"days_to_last_followup\",  \"years_to_birth\", \"days_to_death\", \"date_of_initial_pathologic_diagnosis\"])\n",
    "\n",
    "display(clinical_preprocessed.iloc[:3,:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d4d4acfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nan values in X_train_full: 0\n",
      "Nan value in X_train_full after dropping: 0\n",
      "X_train_full shape: (658, 4858)\n",
      "\n",
      "Network shape: (4858, 4858)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "X_train_full = pd.concat([dna_meth, rna, mirna], axis=1)\n",
    "\n",
    "print(f\"Nan values in X_train_full: {X_train_full.isna().sum().sum()}\")\n",
    "X_train_full = X_train_full.dropna()\n",
    "print(f\"Nan value in X_train_full after dropping: {X_train_full.isna().sum().sum()}\")\n",
    "\n",
    "print(f\"X_train_full shape: {X_train_full.shape}\")\n",
    "# building the graph using the similarity graph function with k=15\n",
    "A_train = bnn.utils.gen_similarity_graph(X_train_full, k=15)\n",
    "\n",
    "print(f\"\\nNetwork shape: {A_train.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c68ddb7",
   "metadata": {},
   "source": [
    "## Reproducibility and Seeding\n",
    "\n",
    "To ensure our experimental results are fully reproducible, a single global seed is set at the beginning of the analysis.\n",
    "\n",
    "This utility function propagates the seed to all sources of randomness, including `random`, `numpy`, and `torch` (for both CPU and GPU).\n",
    "\n",
    "Critically, it also configures the PyTorch cuDNN backend to use deterministic algorithms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "339c3c5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 16:34:44,524 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 118\n",
      "2025-11-08 16:34:44,525 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-08 16:34:44,525 - bioneuralnet.utils.data - INFO - Seed setting complete\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "SEED = 118\n",
    "bnn.utils.set_seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38315975",
   "metadata": {},
   "source": [
    "## Hyperparameter Tuning\n",
    "\n",
    "- First step on our benchmark analysis is dedicated to **hyperparameter tuning** for:\n",
    "    - **SAGE:** GraphSAGE (Graph Sample and Aggregate)\n",
    "    - **GCN:** Graph Convolutional Network\n",
    "    - **GAT:** Graph Attention Network\n",
    "    \n",
    "- We use `tune=True` and `repeat_num=5` within **DPMON** (our classification framework) to let the framework automatically search for the best architectural and training parameters.\n",
    "\n",
    "- After the tuning phase, we will manually set the best-performing parameters for each model.\n",
    "- We then use these optimal configurations in the final cross-validation to rigorously test each model performance and stability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c05cb625",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "# We rename the target to phenotype so we keep it consistent with DPMON expectations.\n",
    "target = target.rename(columns={\"target\": \"phenotype\"})\n",
    "\n",
    "dpmon_sage_tune = bnn.downstream_task.DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[dna_meth, rna, mirna],\n",
    "    phenotype_data=target,\n",
    "    clinical_data=clinical_preprocessed,\n",
    "    model=\"SAGE\",\n",
    "    tune=True,\n",
    "    repeat_num=5,\n",
    "    gpu=True, cuda=0,\n",
    "    output_dir=Path(\"/home/vicente/Github/BioNeuralNet/dpmon_tuning/KIPAN_SAGE\"),\n",
    ")\n",
    "\n",
    "dpmon_sage_tune.run()\n",
    "\n",
    "dpmon_gcn_tune = bnn.downstream_task.DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[dna_meth, rna, mirna],\n",
    "    phenotype_data=target,\n",
    "    clinical_data=clinical_preprocessed,\n",
    "    model=\"GCN\",\n",
    "    tune=True,\n",
    "    repeat_num=5,\n",
    "    gpu=True, cuda=0,\n",
    "    output_dir=Path(\"/home/vicente/Github/BioNeuralNet/dpmon_tuning/KIPAN_GCN\"),\n",
    ")\n",
    "\n",
    "dpmon_gcn_tune.run()\n",
    "\n",
    "\n",
    "dpmon_gat_tune = bnn.downstream_task.DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[dna_meth, rna, mirna],\n",
    "    phenotype_data=target,\n",
    "    clinical_data=clinical_preprocessed,\n",
    "    model=\"GAT\",\n",
    "    tune=True,\n",
    "    repeat_num=5,\n",
    "    gpu=True, cuda=0,\n",
    "    output_dir=Path(\"/home/vicente/Github/BioNeuralNet/dpmon_tuning/KIPAN_GAT\"),\n",
    ")\n",
    "\n",
    "dpmon_gcn_tune.run()\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-KIPAN.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e9023fe",
   "metadata": {},
   "source": [
    "## Analysis of Hyperparameter Optimization\n",
    "\n",
    "**Hyperparameter tuning using `SEED = 118`**\n",
    "\n",
    "### Tuning Search Space\n",
    "\n",
    "**DPMON** parameter search space is extensive, offering over 12,000 discrete combinations plus a continuous space for learning rates. By sampling **50 trials** from this space, we are effectively probing it to find the distinct, optimal architecture best suited for each GNN (SAGE, GCN, and GAT).\n",
    "\n",
    "\n",
    "| Parameter | SAGE (GraphSAGE) | GCN (Graph Convolutional) | GAT (Graph Attention) |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| **gnn\\_layer\\_num** | 2 | 2 | 16 |\n",
    "| **gnn\\_hidden\\_dim** | 128 | 16 | 4 |\n",
    "| **lr (Learning Rate)** | 0.005197 | 0.005345 | 0.003226 |\n",
    "| **weight\\_decay** | 0.046079 | 0.003940 | 0.046863 |\n",
    "| **nn\\_hidden\\_dim1** | 16 | 64 | 16 |\n",
    "| **nn\\_hidden\\_dim2** | 32 | 16 | 128 |\n",
    "| **num\\_epochs** | 4096 | 512 | 512 |\n",
    "\n",
    "\n",
    "### SAGE Model Configuration\n",
    "The optimal SAGE configuration was **shallow (2 layers) and wide (128 hidden dim)**. This confirms a preference for a high-dimensional feature representation over recursive depth. It also required the longest training time at 4,096 epochs.\n",
    "\n",
    "### GCN Model Configuration\n",
    "In a major shift, the GCN model favored a classic, shallow architecture: **2 layers and 16 hidden dim**. This highly efficient configuration converged in just 512 epochs and achieved a perfect (1.0) tuning accuracy, indicating it found a strong, generalizable pattern quickly.\n",
    "\n",
    "### GAT Model Configuration\n",
    "The GAT configuration favored a **deep (16 layers) and narrow (4 hidden dim)** structure. It was also highly efficient, converging in 512 epoch*, suggesting GATs attention mechanism works well even when stacked deeply with low-dimensional features."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "165b9714",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54cec9f0",
   "metadata": {},
   "source": [
    "## Cross-Validation Results with 5 folds\n",
    "\n",
    "| Model | Avg. Accuracy | Avg. F1 Weighted | Avg. F1 Macro |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| **GCN** | **0.9525** +/- 0.0267 | **0.9563** +/- 0.0241 | 0.9232 +/- 0.0410 |\n",
    "| **GAT** | 0.9506 +/- 0.0759 | 0.9528 +/- 0.0718 | **0.9349** +/- 0.0954 |\n",
    "| **SAGE** | 0.9206 +/- 0.0581 | 0.9119 +/- 0.0695 | 0.8695 +/- 0.1418 |\n",
    "\n",
    "The cross-validation results for the KIPAN dataset show a new performance ranking. GCN emerged as the top model, followed closely by GAT, with SAGE also performing strongly.\n",
    "\n",
    "### 1. GAT Performance\n",
    "The GAT model was a top contender, achieving an average accuracy of 0.9506. Its deep-and-narrow architecture (16 layers, 4 hidden dim) proved highly effective on this dataset. Its stability was significantly improved, though its variance was the highest of the three.\n",
    "\n",
    "- Avg. Accuracy: 0.9506 (+/- 0.0759)\n",
    "- Avg. F1 Weighted: 0.9528 (+/- 0.0718)\n",
    "\n",
    "### 2. GCN Performance\n",
    "GCN was the **top-performing and most stable model**. Its shallow-and-narrow architecture (2 layers, 16 hidden dim) translated perfectly from its 1.0 tuning accuracy. Most importantly, it had an **extremely low standard deviation** (+/- 0.0267), indicating its performance was highly reliable across all folds.\n",
    "\n",
    "- Avg. Accuracy: 0.9525 (+/- 0.0267)\n",
    "- Avg. F1 Weighted: 0.9563 (+/- 0.0241)\n",
    "\n",
    "### 3. SAGE Performance\n",
    "SAGE delivered a strong and reliable performance, with its average accuracy exceeding 0.92. Its shallow-and-wide architecture (2 layers, 128 hidden dim) was effective and showed good stability with a low standard deviation (+/- 0.0581), second only to GCN.\n",
    "\n",
    "- Avg. Accuracy: 0.9206 (+/- 0.0581)\n",
    "- Avg. F1 Weighted: 0.9119 (+/- 0.0695)\n",
    "\n",
    "### Summary\n",
    "In summary, all three models performed exceptionally well on the KIPAN dataset, with GCN taking the top spot. The most significant finding was the **dramatic increase in stability** across all models in comparison to the **GBMLGG** dataset.\n",
    "\n",
    "**These results are shown below**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4906eda5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import f1_score, accuracy_score\n",
    "from bioneuralnet.downstream_task import DPMON\n",
    "\n",
    "X_meth_full = dna_meth\n",
    "X_rna_full = rna\n",
    "X_mirna_full = mirna\n",
    "C_full = clinical_preprocessed\n",
    "Y_full = target\n",
    "A_full = A_train\n",
    "\n",
    "patient_indices = np.arange(len(Y_full))\n",
    "y_target_classes = Y_full.squeeze()\n",
    "\n",
    "sage_params = {\n",
    "    'layer_num': 2,\n",
    "    'gnn_hidden_dim': 128,\n",
    "    'lr': 0.005197038211089667,\n",
    "    'weight_decay': 0.04607928341516355,\n",
    "    'nn_hidden_dim1': 16,\n",
    "    'nn_hidden_dim2': 32,\n",
    "    'num_epochs': 4096\n",
    "}\n",
    "\n",
    "N_FOLDS = 5\n",
    "skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=SEED) \n",
    "output_dir_base_sage = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_cv_results_SAGE_FINAL/kipan\")\n",
    "\n",
    "accuracy_scores_sage = []\n",
    "f1_macro_scores_sage = []\n",
    "f1_weighted_scores_sage = []\n",
    "\n",
    "for fold_num, (train_index, test_index) in enumerate(skf.split(patient_indices, y_target_classes)):\n",
    "    print(f\"\\nRunning DPMON SAGE Fold {fold_num + 1}/{N_FOLDS}\")\n",
    "\n",
    "    X_train_omics = [X_meth_full.iloc[train_index], X_rna_full.iloc[train_index], X_mirna_full.iloc[train_index]]\n",
    "    Y_train = Y_full.iloc[train_index]\n",
    "    C_train = C_full.iloc[train_index]\n",
    "    A_train_fold = A_full.iloc[train_index, train_index] \n",
    "    \n",
    "    dpmon_fold = DPMON(\n",
    "        adjacency_matrix=A_train_fold, omics_list=X_train_omics, phenotype_data=Y_train,\n",
    "        clinical_data=C_train, repeat_num=1, tune=False, gpu=True, cuda=0,\n",
    "        output_dir=output_dir_base_sage / f\"fold_{fold_num + 1}\",\n",
    "        **sage_params\n",
    "    )\n",
    "\n",
    "    predictions_df, _ = dpmon_fold.run() \n",
    "    actual = predictions_df[\"Actual\"]\n",
    "    pred = predictions_df[\"Predicted\"]\n",
    "\n",
    "    test_f1_macro = f1_score(actual, pred, average='macro')\n",
    "    test_f1_weighted = f1_score(actual, pred, average='weighted')\n",
    "    test_acc = accuracy_score(actual, pred)\n",
    "\n",
    "    accuracy_scores_sage.append(test_acc)\n",
    "    f1_macro_scores_sage.append(test_f1_macro)\n",
    "    f1_weighted_scores_sage.append(test_f1_weighted)\n",
    "    \n",
    "    print(f\"SAGE Fold {fold_num + 1} | Accuracy: {test_acc:.4f} | F1-Macro: {test_f1_macro:.4f}\")\n",
    "\n",
    "mean_acc_sage = np.mean(accuracy_scores_sage)\n",
    "std_acc_sage = np.std(accuracy_scores_sage)\n",
    "\n",
    "mean_f1_macro_sage = np.mean(f1_macro_scores_sage)\n",
    "std_f1_macro_sage = np.std(f1_macro_scores_sage)\n",
    "\n",
    "mean_f1_weighted_sage = np.mean(f1_weighted_scores_sage)\n",
    "std_f1_weighted_sage = np.std(f1_weighted_scores_sage)\n",
    "\n",
    "print(\"\\nClassification with SAGE and 5-FOLD Cross-Validation\")\n",
    "print(f\"Avg. Accuracy: {mean_acc_sage:.4f} +/- {std_acc_sage:.4f}\")\n",
    "print(f\"Avg. F1 Weighted: {mean_f1_weighted_sage:.4f} +/- {std_f1_weighted_sage:.4f}\")\n",
    "print(f\"Avg. F1 Macro: {mean_f1_macro_sage:.4f} +/- {std_f1_macro_sage:.4f}\")\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-KIPAN.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81315c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir_base_gcn = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL/kipan\")\n",
    "\n",
    "gcn_params = {\n",
    "    'layer_num': 2,\n",
    "    'gnn_hidden_dim': 16,\n",
    "    'lr': 0.005345517248048608,\n",
    "    'weight_decay': 0.003940908881257014,\n",
    "    'nn_hidden_dim1': 64,\n",
    "    'nn_hidden_dim2': 16,\n",
    "    'num_epochs': 512\n",
    "}\n",
    "\n",
    "skf_gcn = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=SEED) \n",
    "\n",
    "accuracy_scores_gcn = []\n",
    "f1_macro_scores_gcn = []\n",
    "f1_weighted_scores_gcn = []\n",
    "\n",
    "for fold_num, (train_index, test_index) in enumerate(skf_gcn.split(patient_indices, y_target_classes)):\n",
    "    print(f\"\\nDPMON GCN Fold {fold_num + 1}/{N_FOLDS}\")\n",
    "\n",
    "    X_train_omics = [X_meth_full.iloc[train_index], X_rna_full.iloc[train_index], X_mirna_full.iloc[train_index]]\n",
    "    Y_train = Y_full.iloc[train_index]\n",
    "    C_train = C_full.iloc[train_index]\n",
    "    A_train_fold = A_full.iloc[train_index, train_index] \n",
    "    \n",
    "    dpmon_fold = DPMON(\n",
    "        adjacency_matrix=A_train_fold, omics_list=X_train_omics, phenotype_data=Y_train,\n",
    "        clinical_data=C_train, repeat_num=1, tune=False, gpu=True, cuda=0,\n",
    "        output_dir=output_dir_base_gcn / f\"fold_{fold_num + 1}\",\n",
    "        **gcn_params\n",
    "    )\n",
    "\n",
    "    predictions_df, _ = dpmon_fold.run() \n",
    "    actual = predictions_df[\"Actual\"]\n",
    "    pred = predictions_df[\"Predicted\"]\n",
    "\n",
    "    test_acc = accuracy_score(actual, pred)\n",
    "    test_f1 = f1_score(actual, pred, average='macro')\n",
    "    test_f1w = f1_score(actual, pred, average='weighted')\n",
    "\n",
    "    accuracy_scores_gcn.append(test_acc)\n",
    "    f1_macro_scores_gcn.append(test_f1)\n",
    "    f1_weighted_scores_gcn.append(test_f1w)\n",
    "    \n",
    "    print(f\"GCN Fold {fold_num + 1} | Accuracy: {test_acc:.4f} | F1-Macro: {test_f1:.4f}\")\n",
    "\n",
    "mean_acc_gcn = np.mean(accuracy_scores_gcn)\n",
    "std_acc_gcn = np.std(accuracy_scores_gcn)\n",
    "\n",
    "mean_f1_macro_gcn = np.mean(f1_macro_scores_gcn)\n",
    "std_f1_macro_gcn = np.std(f1_macro_scores_gcn)\n",
    "\n",
    "mean_f1_weighted_gcn = np.mean(f1_weighted_scores_gcn)\n",
    "std_f1_weighted_gcn = np.std(f1_weighted_scores_gcn)\n",
    "\n",
    "print(\"Classification with GCN and 5-FOLD Cross-Validation\")\n",
    "print(f\"Avg. Accuracy: {mean_acc_gcn:.4f} +/- {std_acc_gcn:.4f}\")\n",
    "print(f\"Avg. F1 Weighted: {mean_f1_weighted_gcn:.4f} +/- {std_f1_weighted_gcn:.4f}\")\n",
    "print(f\"Avg. F1 Macro: {mean_f1_macro_gcn:.4f} +/- {std_f1_macro_gcn:.4f}\")\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-KIPAN.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efbff52f",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir_base_gat = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL/kipan\")\n",
    "\n",
    "gat_params = {\n",
    "    'layer_num': 16,\n",
    "    'gnn_hidden_dim': 4,\n",
    "    'lr': 0.003226646486974539,\n",
    "    'weight_decay': 0.04686394640851084,\n",
    "    'nn_hidden_dim1': 16,\n",
    "    'nn_hidden_dim2': 128,\n",
    "    'num_epochs': 512\n",
    "}\n",
    "\n",
    "skf_gat = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=SEED) \n",
    "\n",
    "accuracy_scores_gat = []\n",
    "f1_macro_scores_gat = []\n",
    "f1_weighted_scores_gat = []\n",
    "\n",
    "for fold_num, (train_index, test_index) in enumerate(skf_gat.split(patient_indices, y_target_classes)):\n",
    "    print(f\"\\nDPMON GAT Fold {fold_num + 1}/{N_FOLDS}\")\n",
    "\n",
    "    X_train_omics = [X_meth_full.iloc[train_index], X_rna_full.iloc[train_index], X_mirna_full.iloc[train_index]]\n",
    "    Y_train = Y_full.iloc[train_index]\n",
    "    C_train = C_full.iloc[train_index]\n",
    "    A_train_fold = A_full.iloc[train_index, train_index] \n",
    "    \n",
    "    dpmon_fold = DPMON(\n",
    "        adjacency_matrix=A_train_fold, omics_list=X_train_omics, phenotype_data=Y_train,\n",
    "        clinical_data=C_train, repeat_num=1, tune=False, gpu=True, cuda=0,\n",
    "        output_dir=output_dir_base_gat / f\"fold_{fold_num + 1}\",\n",
    "        **gat_params\n",
    "    )\n",
    "\n",
    "    predictions_df, _ = dpmon_fold.run() \n",
    "    actual = predictions_df[\"Actual\"]\n",
    "    pred = predictions_df[\"Predicted\"]\n",
    "\n",
    "    test_acc = accuracy_score(actual, pred)\n",
    "    test_f1 = f1_score(actual, pred, average='macro')\n",
    "    test_f1w = f1_score(actual, pred, average='weighted')\n",
    "\n",
    "    accuracy_scores_gat.append(test_acc)\n",
    "    f1_macro_scores_gat.append(test_f1)\n",
    "    f1_weighted_scores_gat.append(test_f1w)\n",
    "    \n",
    "    print(f\"GAT Fold {fold_num + 1} | Accuracy: {test_acc:.4f} | F1-Macro: {test_f1:.4f}\")\n",
    "\n",
    "mean_acc_gat = np.mean(accuracy_scores_gat)\n",
    "std_acc_gat = np.std(accuracy_scores_gat)\n",
    "\n",
    "mean_f1_macro_gat = np.mean(f1_macro_scores_gat)\n",
    "std_f1_macro_gat = np.std(f1_macro_scores_gat)\n",
    "\n",
    "mean_f1_weighted_gat = np.mean(f1_weighted_scores_gat)\n",
    "std_f1_weighted_gat = np.std(f1_weighted_scores_gat)\n",
    "\n",
    "print(\"Classification with GAT and 5-FOLD Cross-Validation\")\n",
    "print(f\"Avg. Accuracy: {mean_acc_gat:.4f} +/- {std_acc_gat:.4f}\")\n",
    "print(f\"Avg. F1 Weighted: {mean_f1_weighted_gat:.4f} +/- {std_f1_weighted_gat:.4f}\")\n",
    "print(f\"Avg. F1 Macro: {mean_f1_macro_gat:.4f} +/- {std_f1_macro_gat:.4f}\")\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-KIPAN.ipynb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".enviroment",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
