{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b42de981",
   "metadata": {},
   "source": [
    "# TCGA-GBMLGG Analysis Demo\n",
    "\n",
    "- **Cohort**: Focuses on the TCGA-GBMLGG dataset, a vital resource merging Glioblastoma Multiforme (GBM) and Lower-Grade Glioma (LGG).\n",
    "- **Goal**: Perform histological subtype classification.\n",
    "- **Prediction Target**: Predict whether a tumor is an `astrocytoma`, `oligodendroglioma`, or `oligoastrocytoma` based on its multi-omics profile.\n",
    "\n",
    "**Data Source:** Broad Institute FireHose (http://firebrowse.org/?cohort=GBMLGG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "027dc20c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-06-0675-11</th>\n",
       "      <th>TCGA-06-0678-11</th>\n",
       "      <th>TCGA-06-0680-11</th>\n",
       "      <th>TCGA-06-0681-11</th>\n",
       "      <th>TCGA-06-AABW-11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <td>12.847399</td>\n",
       "      <td>13.789578</td>\n",
       "      <td>13.603454</td>\n",
       "      <td>13.346797</td>\n",
       "      <td>13.545128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <td>13.850719</td>\n",
       "      <td>14.792970</td>\n",
       "      <td>14.597877</td>\n",
       "      <td>14.344260</td>\n",
       "      <td>14.554888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>12.873946</td>\n",
       "      <td>13.810832</td>\n",
       "      <td>13.611074</td>\n",
       "      <td>13.364372</td>\n",
       "      <td>13.583039</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              TCGA-06-0675-11  TCGA-06-0678-11  TCGA-06-0680-11  \\\n",
       "gene                                                              \n",
       "hsa-let-7a-1        12.847399        13.789578        13.603454   \n",
       "hsa-let-7a-2        13.850719        14.792970        14.597877   \n",
       "hsa-let-7a-3        12.873946        13.810832        13.611074   \n",
       "\n",
       "              TCGA-06-0681-11  TCGA-06-AABW-11  \n",
       "gene                                            \n",
       "hsa-let-7a-1        13.346797        13.545128  \n",
       "hsa-let-7a-2        14.344260        14.554888  \n",
       "hsa-let-7a-3        13.364372        13.583039  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(548, 531)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <th>TCGA-02-2486-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>?|100133144</th>\n",
       "      <td>1.619742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.559100</td>\n",
       "      <td>3.999567</td>\n",
       "      <td>2.475344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|100134869</th>\n",
       "      <td>2.757258</td>\n",
       "      <td>3.972445</td>\n",
       "      <td>3.801138</td>\n",
       "      <td>3.902759</td>\n",
       "      <td>2.264506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|10357</th>\n",
       "      <td>5.773564</td>\n",
       "      <td>4.972440</td>\n",
       "      <td>5.915141</td>\n",
       "      <td>6.520796</td>\n",
       "      <td>5.966629</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TCGA-02-0047-01  TCGA-02-0055-01  TCGA-02-2483-01  \\\n",
       "gene                                                             \n",
       "?|100133144         1.619742              NaN         1.559100   \n",
       "?|100134869         2.757258         3.972445         3.801138   \n",
       "?|10357             5.773564         4.972440         5.915141   \n",
       "\n",
       "             TCGA-02-2485-01  TCGA-02-2486-01  \n",
       "gene                                           \n",
       "?|100133144         3.999567         2.475344  \n",
       "?|100134869         3.902759         2.264506  \n",
       "?|10357             6.520796         5.966629  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20115, 685)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-06-0125-01</th>\n",
       "      <th>TCGA-06-0125-02</th>\n",
       "      <th>TCGA-06-0152-01</th>\n",
       "      <th>TCGA-06-0152-02</th>\n",
       "      <th>TCGA-06-0171-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hybridization REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "      <td>Beta_Value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.438986043005</td>\n",
       "      <td>0.565094788162</td>\n",
       "      <td>0.461699906718</td>\n",
       "      <td>0.534127262606</td>\n",
       "      <td>0.455267108058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.681141812896</td>\n",
       "      <td>0.724487443757</td>\n",
       "      <td>0.601439733092</td>\n",
       "      <td>0.632221318323</td>\n",
       "      <td>0.691054589549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      TCGA-06-0125-01 TCGA-06-0125-02 TCGA-06-0152-01  \\\n",
       "Hybridization REF                                                       \n",
       "Composite Element REF      Beta_Value      Beta_Value      Beta_Value   \n",
       "A1BG                   0.438986043005  0.565094788162  0.461699906718   \n",
       "A1CF                   0.681141812896  0.724487443757  0.601439733092   \n",
       "\n",
       "                      TCGA-06-0152-02 TCGA-06-0171-01  \n",
       "Hybridization REF                                      \n",
       "Composite Element REF      Beta_Value      Beta_Value  \n",
       "A1BG                   0.534127262606  0.455267108058  \n",
       "A1CF                   0.632221318323  0.691054589549  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(20115, 685)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcga-06-6391</th>\n",
       "      <th>tcga-19-a6j4</th>\n",
       "      <th>tcga-cs-6665</th>\n",
       "      <th>tcga-cs-6670</th>\n",
       "      <th>tcga-db-a4xc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hybridization REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "      <td>value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>years_to_birth</th>\n",
       "      <td>44</td>\n",
       "      <td>68</td>\n",
       "      <td>51</td>\n",
       "      <td>43</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vital_status</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tcga-06-6391 tcga-19-a6j4 tcga-cs-6665 tcga-cs-6670  \\\n",
       "Hybridization REF                                                           \n",
       "Composite Element REF        value        value        value        value   \n",
       "years_to_birth                  44           68           51           43   \n",
       "vital_status                     1            1            0            0   \n",
       "\n",
       "                      tcga-db-a4xc  \n",
       "Hybridization REF                   \n",
       "Composite Element REF        value  \n",
       "years_to_birth                  26  \n",
       "vital_status                     0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(14, 1110)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "root = Path(\"/home/vicente/Github/BioNeuralNet/GBMLGG\")\n",
    "\n",
    "mirna_raw = pd.read_csv(root/\"GBMLGG.miRseq_RPKM_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)                            \n",
    "rna_raw = pd.read_csv(root / \"GBMLGG.uncv2.mRNAseq_RSEM_normalized_log2.txt\", sep=\"\\t\",index_col=0,low_memory=False)\n",
    "meth_raw = pd.read_csv(root/\"GBMLGG.meth.by_mean.data.txt\", sep='\\t',index_col=0,low_memory=False)\n",
    "clinical_raw = pd.read_csv(root / \"GBMLGG.clin.merged.picked.txt\",sep=\"\\t\", index_col=0, low_memory=False)\n",
    "\n",
    "# display shapes and first few rows-columns of each file\n",
    "display(mirna_raw.iloc[:3,:5])\n",
    "display(mirna_raw.shape)\n",
    "\n",
    "display(rna_raw.iloc[:3,:5])\n",
    "display(meth_raw.shape)\n",
    "\n",
    "display(meth_raw.iloc[:3,:5])\n",
    "display(meth_raw.shape)\n",
    "\n",
    "display(clinical_raw.iloc[:3,:5])\n",
    "display(clinical_raw.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbacccaf",
   "metadata": {},
   "source": [
    "## Data Processing Summary\n",
    "\n",
    "1. **Transpose Data:** All raw data (miRNA, RNA, etc.) is flipped so rows represent patients and columns represent features.\n",
    "2. **Standardize Patient IDs:** Patient IDs in all tables are cleaned to the 12-character TCGA format (e.g., `TCGA-AB-1234`) for matching.\n",
    "3. **Handle Duplicates:** Duplicate patient rows are averaged in the omics data. The first entry is kept for duplicate patients in the clinical data.\n",
    "4. **Find Common Patients:** The script identifies the list of patients that exist in *all* datasets.\n",
    "5. **Subset Data:** All data tables are filtered down to *only* this common list of patients, ensuring alignment.\n",
    "6. **Extract Target:** The `histological_type` column is pulled from the processed clinical data to be used as the prediction target (y-variable)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "820583f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miRNA (samples, features): (531, 548)\n",
      "RNA (samples, features): (701, 18328)\n",
      "Methylation (samples, features): (685, 20115)\n",
      "Clinical (samples, features): (1110, 14)\n",
      "\n",
      "Methylation shape: (658, 20115)\n",
      "RNA shape: (681, 18328)\n",
      "miRNA shape: (517, 548)\n",
      "Clinical shape: (1110, 14)\n",
      "\n",
      "Found: 511 patients across all data types.\n"
     ]
    }
   ],
   "source": [
    "mirna = mirna_raw.T\n",
    "rna = rna_raw.T\n",
    "meth = meth_raw.T\n",
    "clinical = clinical_raw.T\n",
    "\n",
    "print(f\"miRNA (samples, features): {mirna.shape}\")\n",
    "print(f\"RNA (samples, features): {rna.shape}\")\n",
    "print(f\"Methylation (samples, features): {meth.shape}\")\n",
    "print(f\"Clinical (samples, features): {clinical.shape}\")\n",
    "\n",
    "def trim_barcode(idx):\n",
    "    return idx.to_series().str.slice(0, 12)\n",
    "\n",
    "# standarized patient IDs across all files\n",
    "meth.index = trim_barcode(meth.index)\n",
    "rna.index = trim_barcode(rna.index)\n",
    "mirna.index = trim_barcode(mirna.index)\n",
    "clinical.index = clinical.index.str.upper()\n",
    "clinical.index.name = \"Patient_ID\"\n",
    "\n",
    "# convert all data to numeric, coercing errors to NaN\n",
    "meth = meth.apply(pd.to_numeric, errors='coerce')\n",
    "rna = rna.apply(pd.to_numeric, errors='coerce')\n",
    "mirna = mirna.apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "# for any duplicate columns in the omics data, we average their values\n",
    "meth = meth.groupby(meth.index).mean()\n",
    "rna = rna.groupby(rna.index).mean()\n",
    "mirna = mirna.groupby(mirna.index).mean()\n",
    "\n",
    "# for any duplicate rows in the clinical data, we keep the first occurrence\n",
    "clinical = clinical[~clinical.index.duplicated(keep='first')]\n",
    "\n",
    "print(f\"\\nMethylation shape: {meth.shape}\")\n",
    "print(f\"RNA shape: {rna.shape}\")\n",
    "print(f\"miRNA shape: {mirna.shape}\")\n",
    "print(f\"Clinical shape: {clinical.shape}\")\n",
    "\n",
    "for df in [meth, rna, mirna]:\n",
    "    df.columns = df.columns.str.replace(r\"\\?\", \"unknown_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(r\"\\|\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.replace(\"-\", \"_\", regex=False)\n",
    "    df.columns = df.columns.str.replace(r\"_+\", \"_\", regex=True)\n",
    "    df.columns = df.columns.str.strip(\"_\")\n",
    "    \n",
    "    df.fillna(df.mean(), inplace=True)\n",
    "\n",
    "# to see which pateints are common across all data files\n",
    "common_patients = sorted(list(set(meth.index)&set(rna.index)&set(mirna.index)&set(clinical.index)))\n",
    "\n",
    "print(f\"\\nFound: {len(common_patients)} patients across all data types.\")\n",
    "\n",
    "# subset to only common patients\n",
    "meth_processed = meth.loc[common_patients]\n",
    "rna_processed= rna.loc[common_patients]\n",
    "mirna_processed = mirna.loc[common_patients]\n",
    "clinical_processed = clinical.loc[common_patients]\n",
    "\n",
    "# extract target labels from clinical data\n",
    "targets = clinical_processed['histological_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d9c562ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>hsa_let_7a_1</th>\n",
       "      <th>hsa_let_7a_2</th>\n",
       "      <th>hsa_let_7a_3</th>\n",
       "      <th>hsa_let_7b</th>\n",
       "      <th>hsa_let_7c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4938</th>\n",
       "      <td>12.622353</td>\n",
       "      <td>13.632728</td>\n",
       "      <td>12.651613</td>\n",
       "      <td>14.208930</td>\n",
       "      <td>14.376942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4941</th>\n",
       "      <td>11.809808</td>\n",
       "      <td>12.815815</td>\n",
       "      <td>11.820061</td>\n",
       "      <td>13.047853</td>\n",
       "      <td>11.955006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4942</th>\n",
       "      <td>11.113995</td>\n",
       "      <td>12.128618</td>\n",
       "      <td>11.165523</td>\n",
       "      <td>12.481790</td>\n",
       "      <td>11.858545</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene          hsa_let_7a_1  hsa_let_7a_2  hsa_let_7a_3  hsa_let_7b  hsa_let_7c\n",
       "TCGA-CS-4938     12.622353     13.632728     12.651613   14.208930   14.376942\n",
       "TCGA-CS-4941     11.809808     12.815815     11.820061   13.047853   11.955006\n",
       "TCGA-CS-4942     11.113995     12.128618     11.165523   12.481790   11.858545"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(511, 548)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>unknown_100133144</th>\n",
       "      <th>unknown_100134869</th>\n",
       "      <th>unknown_10357</th>\n",
       "      <th>unknown_10431</th>\n",
       "      <th>unknown_155060</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4938</th>\n",
       "      <td>3.123352</td>\n",
       "      <td>4.507940</td>\n",
       "      <td>8.069184</td>\n",
       "      <td>9.724198</td>\n",
       "      <td>7.511790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4941</th>\n",
       "      <td>5.187819</td>\n",
       "      <td>4.404406</td>\n",
       "      <td>7.291745</td>\n",
       "      <td>8.608326</td>\n",
       "      <td>8.344526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4942</th>\n",
       "      <td>3.562316</td>\n",
       "      <td>3.462602</td>\n",
       "      <td>7.532460</td>\n",
       "      <td>9.279502</td>\n",
       "      <td>7.034985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene          unknown_100133144  unknown_100134869  unknown_10357  \\\n",
       "TCGA-CS-4938           3.123352           4.507940       8.069184   \n",
       "TCGA-CS-4941           5.187819           4.404406       7.291745   \n",
       "TCGA-CS-4942           3.562316           3.462602       7.532460   \n",
       "\n",
       "gene          unknown_10431  unknown_155060  \n",
       "TCGA-CS-4938       9.724198        7.511790  \n",
       "TCGA-CS-4941       8.608326        8.344526  \n",
       "TCGA-CS-4942       9.279502        7.034985  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(511, 18328)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hybridization REF</th>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2BP1</th>\n",
       "      <th>A2LD1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4938</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.683179</td>\n",
       "      <td>0.776869</td>\n",
       "      <td>0.652055</td>\n",
       "      <td>0.919739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4941</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.521934</td>\n",
       "      <td>0.784401</td>\n",
       "      <td>0.563447</td>\n",
       "      <td>0.865717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4942</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.610067</td>\n",
       "      <td>0.828194</td>\n",
       "      <td>0.607771</td>\n",
       "      <td>0.875369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Hybridization REF  Composite Element REF      A1BG      A1CF     A2BP1  \\\n",
       "TCGA-CS-4938                         NaN  0.683179  0.776869  0.652055   \n",
       "TCGA-CS-4941                         NaN  0.521934  0.784401  0.563447   \n",
       "TCGA-CS-4942                         NaN  0.610067  0.828194  0.607771   \n",
       "\n",
       "Hybridization REF     A2LD1  \n",
       "TCGA-CS-4938       0.919739  \n",
       "TCGA-CS-4941       0.865717  \n",
       "TCGA-CS-4942       0.875369  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(511, 20115)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hybridization REF</th>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_followup</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4938</th>\n",
       "      <td>value</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4941</th>\n",
       "      <td>value</td>\n",
       "      <td>67</td>\n",
       "      <td>1</td>\n",
       "      <td>234</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4942</th>\n",
       "      <td>value</td>\n",
       "      <td>44</td>\n",
       "      <td>1</td>\n",
       "      <td>1335</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Hybridization REF Composite Element REF years_to_birth vital_status  \\\n",
       "Patient_ID                                                            \n",
       "TCGA-CS-4938                      value             31            0   \n",
       "TCGA-CS-4941                      value             67            1   \n",
       "TCGA-CS-4942                      value             44            1   \n",
       "\n",
       "Hybridization REF days_to_death days_to_last_followup  \n",
       "Patient_ID                                             \n",
       "TCGA-CS-4938                NaN                  3574  \n",
       "TCGA-CS-4941                234                   NaN  \n",
       "TCGA-CS-4942               1335                   NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(511, 14)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "histological_type\n",
       "astrocytoma          193\n",
       "oligodendroglioma    191\n",
       "oligoastrocytoma     127\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(mirna_processed.iloc[:3,:5])\n",
    "display(mirna_processed.shape)\n",
    "\n",
    "display(rna_processed.iloc[:3,:5])\n",
    "display(rna_processed.shape)\n",
    "\n",
    "display(meth_processed.iloc[:3,:5])\n",
    "display(meth_processed.shape)\n",
    "\n",
    "display(clinical_processed.iloc[:3,:5])\n",
    "display(clinical_processed.shape)\n",
    "\n",
    "display(targets.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edc9e7f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 13:06:12,202 - bioneuralnet.utils.data - INFO - Starting Beta-to-M value conversion (shape: (511, 20114)). Epsilon: 1e-06\n",
      "2025-11-08 13:06:13,301 - bioneuralnet.utils.data - INFO - Beta-to-M conversion complete.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Methylation shape: (511, 20114)\n",
      "RNA shape: (511, 18328)\n",
      "miRNA shape: (511, 548)\n",
      "Clinical shape: (511, 13)\n",
      "target\n",
      "0         193\n",
      "1         191\n",
      "2         127\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "# drop unwanted columns from clinical data\n",
    "clinical_processed.drop(columns=[\"Composite Element REF\"], errors=\"ignore\", inplace=True)\n",
    "\n",
    "# we transform the methylation beta values to M-values and drop unwanted columns\n",
    "meth_m = meth_processed.drop(columns=[\"Composite Element REF\"], errors=\"ignore\")\n",
    "\n",
    "# convert beta values to M-values using bioneuralnet utility with small epsilon to avoid log(0)\n",
    "meth_m = bnn.utils.beta_to_m(meth_m, eps=1e-6) \n",
    "\n",
    "# lastly we turn the target labels into numerical classes\n",
    "mapping = {\"astrocytoma\": 0, \"oligodendroglioma\": 1, \"oligoastrocytoma\": 2}\n",
    "target_labels = targets.map(mapping).to_frame(name=\"target\")\n",
    "\n",
    "# as a safety check we align the indices once more\n",
    "X_meth = meth_m.loc[common_patients]\n",
    "X_rna = rna_processed.loc[common_patients]\n",
    "X_mirna = mirna_processed.loc[common_patients]\n",
    "Y_labels = target_labels.loc[common_patients]\n",
    "clinical_final = clinical_processed.loc[common_patients]\n",
    "\n",
    "print(f\"\\nDNA_Methylation shape: {X_meth.shape}\")\n",
    "print(f\"RNA shape: {X_rna.shape}\")\n",
    "print(f\"miRNA shape: {X_mirna.shape}\")\n",
    "print(f\"Clinical shape: {clinical_final.shape}\")\n",
    "print(Y_labels.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0643ec13",
   "metadata": {},
   "source": [
    "## Feature Selection Methodology\n",
    "\n",
    "### Supported Methods and Interpretation\n",
    "\n",
    "**BioNeuralNet** provides three techniques for feature selection, allowing for different views of the data's statistical profile:\n",
    "\n",
    "- **Variance Thresholding:** Identifies features with the **highest overall variance** across all samples.\n",
    "\n",
    "- **ANOVA F-test:** Pinpoints features that best **distinguish between the target classes** (KIRC, KIRP, and KICH).\n",
    "\n",
    "- **Random Forest Importance:** Assesses **feature utility** based on its contribution to a predictive non-linear model.\n",
    "\n",
    "### GBMLGG Cohort Selection Strategy\n",
    "\n",
    "A dimensionality reduction step was essential for managing the high-feature-count omics data:\n",
    "\n",
    "- **High-Feature Datasets:** Both DNA Methylation (20,114) and RNA (18,328) required significant feature reduction.\n",
    "\n",
    "- **Filtering Process:** The **top 6,000 features** were initially extracted from the Methylation and RNA datasets using all three methods.\n",
    "\n",
    "- **Final Set:** A consensus set was built by finding the intersection of features selected by the ANOVA F-test and Random Forest Importance, ensuring both statistical relevance and model-based utility.\n",
    "\n",
    "- **Low-Feature Datasets:** The miRNA data (548 features) was passed through **without selection**, as its feature count was already manageable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "978e9281",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 13:15:33,012 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:15:33,012 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 13:15:33,012 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 13:15:33,085 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-08 13:15:35,778 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:15:35,779 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 13:15:35,779 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 13:15:35,843 - bioneuralnet.utils.preprocess - INFO - Selected top 6000 features by variance\n",
      "2025-11-08 13:15:38,824 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:15:38,824 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 13:15:38,824 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 13:15:38,962 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 12015 significant, 0 padded\n",
      "2025-11-08 13:15:41,666 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:15:41,666 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 13:15:41,666 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 13:15:41,773 - bioneuralnet.utils.preprocess - INFO - Selected 6000 features by ANOVA (task=classification), 10820 significant, 0 padded\n",
      "2025-11-08 13:15:44,759 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:15:44,760 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 13:15:44,760 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n",
      "2025-11-08 13:15:49,982 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:15:49,982 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 0 NaNs after median imputation\n",
      "2025-11-08 13:15:49,983 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 0 columns dropped due to zero variance\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "# feature selection\n",
    "meth_highvar = bnn.utils.select_top_k_variance(X_meth, k=6000)\n",
    "rna_highvar = bnn.utils.select_top_k_variance(X_rna, k=6000)\n",
    "\n",
    "meth_af = bnn.utils.top_anova_f_features(X_meth, Y_labels, max_features=6000)\n",
    "rna_af = bnn.utils.top_anova_f_features(X_rna, Y_labels, max_features=6000)\n",
    "\n",
    "meth_rf = bnn.utils.select_top_randomforest(X_meth, Y_labels, top_k=6000)\n",
    "rna_rf = bnn.utils.select_top_randomforest(X_rna, Y_labels, top_k=6000)\n",
    "\n",
    "meth_var_set = set(meth_highvar.columns)\n",
    "meth_anova_set = set(meth_af.columns)\n",
    "meth_rf_set = set(meth_rf.columns)\n",
    "\n",
    "rna_var_set = set(rna_highvar.columns)\n",
    "rna_anova_set = set(rna_af.columns)\n",
    "rna_rf_set = set(rna_rf.columns)\n",
    "\n",
    "meth_inter1 = list(meth_anova_set & meth_var_set)\n",
    "meth_inter2 = list(meth_rf_set & meth_var_set)\n",
    "meth_inter3 = list(meth_anova_set & meth_rf_set)\n",
    "meth_all_three = list(meth_anova_set & meth_var_set & meth_rf_set)\n",
    "\n",
    "rna_inter4 = list(rna_anova_set & rna_var_set)\n",
    "rna_inter5 = list(rna_rf_set & rna_var_set)\n",
    "rna_inter6 = list(rna_anova_set & rna_rf_set)\n",
    "rna_all_three = list(rna_anova_set & rna_var_set & rna_rf_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "467d9ff1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FROM THE 6000 Methylation feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 2704 features\n",
      "Random Forest & variance selection share: 1768 features\n",
      "Anova-F & Random Forest share: 1823 features\n",
      "All three methods agree on: 809 features\n"
     ]
    }
   ],
   "source": [
    "print(\"FROM THE 6000 Methylation feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(meth_inter1)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(meth_inter2)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(meth_inter3)} features\")\n",
    "print(f\"All three methods agree on: {len(meth_all_three)} features\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6fb1e10b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "FROM THE 6000 RNA feature selection:\n",
      "\n",
      "Anova-F & variance selection share: 2183 features\n",
      "Random Forest & variance selection share: 1977 features\n",
      "Anova-F & Random Forest share: 2127 features\n",
      "All three methods agree on: 763 features\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nFROM THE 6000 RNA feature selection:\\n\")\n",
    "print(f\"Anova-F & variance selection share: {len(rna_inter4)} features\")\n",
    "print(f\"Random Forest & variance selection share: {len(rna_inter5)} features\")\n",
    "print(f\"Anova-F & Random Forest share: {len(rna_inter6)} features\")\n",
    "print(f\"All three methods agree on: {len(rna_all_three)} features\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be2ce6f8",
   "metadata": {},
   "source": [
    "## Feature Selection Summary: ANOVA-RF Intersection\n",
    "\n",
    "The chosen strategy for feature selection is based on the **overlap** between features identified by the **ANOVA F-test** and **Random Forest Importance**. This approach offers comprehensive filtering by balancing class-based relevance (ANOVA) with non-linear model importance (Random Forest). The resulting feature sets are considered the most robust for downstream analysis.\n",
    "\n",
    "### Feature Overlap Results\n",
    "\n",
    "The following table details the number of features resulting from the intersection of different selection methods for each omics data type.\n",
    "\n",
    "| Omics Data Type | ANOVA-F & Variance | RF & Variance | ANOVA-F & Random Forest (Selected) | All Three Agree |\n",
    "| :--- | :--- | :--- | :--- | :--- |\n",
    "| **Methylation** | 2,704 features | 1,768 features | **1,823 features** | 809 features |\n",
    "| **RNA** | 2,183 features | 1,977 features | **2,127 features** | 763 features |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a1934672",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Final Shapes for Modeling\n",
      "Methylation (X1): (511, 1823)\n",
      "RNA-Seq (X2): (511, 2127)\n",
      "miRNA-Seq (X3): (511, 548)\n",
      "Labels (Y): (511, 1)\n"
     ]
    }
   ],
   "source": [
    "X_meth_selected = X_meth[meth_inter3]\n",
    "X_rna_selected = X_rna[rna_inter6]\n",
    "\n",
    "print(\"\\nFinal Shapes for Modeling\")\n",
    "print(f\"Methylation (X1): {X_meth_selected.shape}\")\n",
    "print(f\"RNA-Seq (X2): {X_rna_selected.shape}\")\n",
    "print(f\"miRNA-Seq (X3): {X_mirna.shape}\")\n",
    "print(f\"Labels (Y): {Y_labels.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9348ca2a",
   "metadata": {},
   "source": [
    "## Data Availability\n",
    "\n",
    "To facilitate rapid experimentation and reproduction of our results, the fully processed and feature-selected dataset used in this analysis has been made available directly within the package.\n",
    "\n",
    "Users can load this dataset, bypassing all preceding data acquisition, preprocessing, and feature selection steps. This allows users to proceed immediately from this step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ec1120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mirna': (511, 548),\n",
       " 'target': (511, 1),\n",
       " 'clinical': (511, 13),\n",
       " 'rna': (511, 2127),\n",
       " 'meth': (511, 1823)}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "tgca_lgg = bnn.datasets.DatasetLoader(\"lgg\")\n",
    "display(tgca_lgg.shape)\n",
    "\n",
    "# The dataset is returned as a dictionary. We extract each file independetly based on the name( Key).\n",
    "dna_meth = tgca_lgg.data[\"meth\"]\n",
    "rna = tgca_lgg.data[\"rna\"]\n",
    "mirna = tgca_lgg.data[\"mirna\"]\n",
    "clinical = tgca_lgg.data[\"clinical\"]\n",
    "target = tgca_lgg.data[\"target\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c76e4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 13:33:26,861 - bioneuralnet.utils.preprocess - INFO - [Inf]: Replaced 0 infinite values\n",
      "2025-11-08 13:33:26,862 - bioneuralnet.utils.preprocess - INFO - [NaN]: Replaced 717 NaNs after median imputation\n",
      "2025-11-08 13:33:26,862 - bioneuralnet.utils.preprocess - INFO - [Zero-Var]: 1 columns dropped due to zero variance\n",
      "2025-11-08 13:33:26,944 - bioneuralnet.utils.preprocess - INFO - Selected top 7 features by RandomForest importance\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>histological_type_oligodendroglioma</th>\n",
       "      <th>histological_type_oligoastrocytoma</th>\n",
       "      <th>karnofsky_performance_score</th>\n",
       "      <th>radiation_therapy_no</th>\n",
       "      <th>radiation_therapy_yes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4938</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>90.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4941</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>90.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CS-4942</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>70.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              histological_type_oligodendroglioma  \\\n",
       "Patient_ID                                          \n",
       "TCGA-CS-4938                                False   \n",
       "TCGA-CS-4941                                False   \n",
       "TCGA-CS-4942                                False   \n",
       "\n",
       "              histological_type_oligoastrocytoma  karnofsky_performance_score  \\\n",
       "Patient_ID                                                                      \n",
       "TCGA-CS-4938                               False                         90.0   \n",
       "TCGA-CS-4941                               False                         90.0   \n",
       "TCGA-CS-4942                               False                         70.0   \n",
       "\n",
       "              radiation_therapy_no  radiation_therapy_yes  \n",
       "Patient_ID                                                 \n",
       "TCGA-CS-4938                  True                  False  \n",
       "TCGA-CS-4941                 False                   True  \n",
       "TCGA-CS-4942                 False                   True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# BioNeuralNet provides a preprocessing function to handle clinical data\n",
    "clinical = tgca_lgg.data[\"clinical\"]\n",
    "\n",
    "# For more details on the preprocessing functions, see `bioneuralnet.utils.preprocess``\n",
    "clinical_preprocessed = bnn.utils.preprocess_clinical(\n",
    "    clinical, \n",
    "    target, \n",
    "    top_k=7, \n",
    "    scale=False, \n",
    "    ignore_columns=[ \"days_to_last_followup\",  \"years_to_birth\", \"days_to_death\", \"date_of_initial_pathologic_diagnosis\"])\n",
    "\n",
    "display(clinical_preprocessed.iloc[:3,:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fb76a376",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nan values in X_train_full: 0\n",
      "Nan value in X_train_full after dropping: 0\n",
      "X_train_full shape: (511, 4498)\n",
      "\n",
      "Network shape: (4498, 4498)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "X_train_full = pd.concat([dna_meth, rna, mirna], axis=1)\n",
    "\n",
    "print(f\"Nan values in X_train_full: {X_train_full.isna().sum().sum()}\")\n",
    "X_train_full = X_train_full.dropna()\n",
    "print(f\"Nan value in X_train_full after dropping: {X_train_full.isna().sum().sum()}\")\n",
    "\n",
    "print(f\"X_train_full shape: {X_train_full.shape}\")\n",
    "# building the graph using the similarity graph function with k=15\n",
    "A_train = bnn.utils.gen_similarity_graph(X_train_full, k=15)\n",
    "\n",
    "print(f\"\\nNetwork shape: {A_train.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a2b14e",
   "metadata": {},
   "source": [
    "## Reproducibility and Seeding\n",
    "\n",
    "To ensure our experimental results are fully reproducible, a single global seed is set at the beginning of the analysis.\n",
    "\n",
    "This utility function propagates the seed to all sources of randomness, including `random`, `numpy`, and `torch` (for both CPU and GPU).\n",
    "\n",
    "Critically, it also configures the PyTorch cuDNN backend to use deterministic algorithms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f28d80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 16:33:32,623 - bioneuralnet.utils.data - INFO - Setting global seed for reproducibility to: 118\n",
      "2025-11-08 16:33:32,633 - bioneuralnet.utils.data - INFO - CUDA available. Applying seed to all GPU operations\n",
      "2025-11-08 16:33:32,634 - bioneuralnet.utils.data - INFO - Seed setting complete\n"
     ]
    }
   ],
   "source": [
    "import bioneuralnet as bnn\n",
    "\n",
    "SEED = 118\n",
    "bnn.utils.set_seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb5e0121",
   "metadata": {},
   "source": [
    "## Hyperparameter Tuning\n",
    "\n",
    "- First step on our benchmark analysis is dedicated to **hyperparameter tuning** for:\n",
    "    - **SAGE:** GraphSAGE (Graph Sample and Aggregate)\n",
    "    - **GCN:** Graph Convolutional Network\n",
    "    - **GAT:** Graph Attention Network\n",
    "    \n",
    "- We use `tune=True` and `repeat_num=5` within **DPMON** (our classification framework) to let the framework automatically search for the best architectural and training parameters.\n",
    "\n",
    "- After the tuning phase, we will manually set the best-performing parameters for each model.\n",
    "- We then use these optimal configurations in the final cross-validation to rigorously test each model performance and stability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a6d4721",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "# We rename the target to phenotype so we keep it consistent with DPMON expectations.\n",
    "target = target.rename(columns={\"target\": \"phenotype\"})\n",
    "\n",
    "dpmon_sage_tune = bnn.downstream_task.DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[dna_meth, rna, mirna],\n",
    "    phenotype_data=target,\n",
    "    clinical_data=clinical_preprocessed,\n",
    "    model=\"SAGE\",\n",
    "    tune=True,\n",
    "    repeat_num=5,\n",
    "    gpu=True, cuda=0,\n",
    "    output_dir=Path(\"/home/vicente/Github/BioNeuralNet/dpmon_tuning/GBM_SAGE\"),\n",
    ")\n",
    "\n",
    "dpmon_sage_tune.run()\n",
    "\n",
    "dpmon_gcn_tune = bnn.downstream_task.DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[dna_meth, rna, mirna],\n",
    "    phenotype_data=target,\n",
    "    clinical_data=clinical_preprocessed,\n",
    "    model=\"GCN\",\n",
    "    tune=True,\n",
    "    repeat_num=5,\n",
    "    gpu=True, cuda=0,\n",
    "    output_dir=Path(\"/home/vicente/Github/BioNeuralNet/dpmon_tuning/GBM_GCN\"),\n",
    ")\n",
    "\n",
    "dpmon_gcn_tune.run()\n",
    "\n",
    "\n",
    "dpmon_gat_tune = bnn.downstream_task.DPMON(\n",
    "    adjacency_matrix=A_train,\n",
    "    omics_list=[dna_meth, rna, mirna],\n",
    "    phenotype_data=target,\n",
    "    clinical_data=clinical_preprocessed,\n",
    "    model=\"GAT\",\n",
    "    tune=True,\n",
    "    repeat_num=5,\n",
    "    gpu=True, cuda=0,\n",
    "    output_dir=Path(\"/home/vicente/Github/BioNeuralNet/dpmon_tuning/GBM_GAT\"),\n",
    ")\n",
    "\n",
    "dpmon_gcn_tune.run()\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-GBMLGG.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "383740ac",
   "metadata": {},
   "source": [
    "## Analysis of Hyperparameter Optimization\n",
    "\n",
    "**Hyperparameter tuning using `SEED = 118`**\n",
    "\n",
    "### Tuning Search Space\n",
    "\n",
    "**DPMON** parameter search space is extensive, offering over 12,000 discrete combinations plus a continuous space for learning rates. By sampling **50 trials** from this space, we are effectively probing it to find the distinct, optimal architecture best suited for each GNN (SAGE, GCN, and GAT).\n",
    "\n",
    "\n",
    "| Parameter | SAGE (GraphSAGE) | GCN (Graph Convolutional) | GAT (Graph Attention) |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| **gnn\\_layer\\_num** | 2 | 64 | 16 |\n",
    "| **gnn\\_hidden\\_dim** | 128 | 16 | 4 |\n",
    "| **lr (Learning Rate)** | 0.005197 | 0.004397 | 0.003226 |\n",
    "| **weight\\_decay** | 0.046079 | 0.000405 | 0.046863 |\n",
    "| **nn\\_hidden\\_dim1** | 1 | 64 | 16 |\n",
    "| **nn\\_hidden\\_dim2** | 32 | 8 | 128 |\n",
    "| **num\\_epochs** | 4096 | 1024 | 512 |\n",
    "\n",
    "### SAGE Model Configuration\n",
    "The optimal SAGE configuration was **shallow but wide**, requiring an unusually long training time. This completely reverses the previous \"deep and narrow\" result, indicating that a high-dimensional feature representation was prioritized over recursive aggregation in this specific run.\n",
    "\n",
    "### GCN Model Configuration\n",
    "The GCN model favored a highly atypical architecture: **deep yet narrow**. The success of this extremely deep GCN suggests the model learned highly specialized, low-dimensional features over many layers, despite the typical risk of over-smoothing.\n",
    "\n",
    "### GAT Model Configuration\n",
    "The GAT model found a balanced, efficient architecture: **moderate depth and narrow**. It was the fastest to converge, suggesting GAT's attention mechanism effectively selected features, allowing it to perform well with a minimal, low-dimensional representation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b44593d0",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88084af0",
   "metadata": {},
   "source": [
    "## Cross-Validation Results with 5 folds\n",
    "\n",
    "| Model | Avg. Accuracy | Avg. F1 Weighted | Avg. F1 Macro |\n",
    "| :--- | :--- | :--- | :--- |\n",
    "| **GAT** | **0.9213** +/- 0.1321 | **0.9045** +/- 0.1656 | **0.8906** +/- 0.1910 |\n",
    "| **GCN** | 0.8821 +/- 0.1353 | 0.8639 +/- 0.1666 | 0.8644 +/- 0.1620 |\n",
    "| **SAGE** | 0.7567 +/- 0.1846 | 0.7311 +/- 0.2160 | 0.7135 +/- 0.2361 |\n",
    "\n",
    "The cross-validation results demonstrate that the optimal architectures found with `SEED = 118` performed best with GAT, followed by GCN, and then SAGE.\n",
    "\n",
    "### 1. GAT Performance\n",
    "\n",
    "The GAT model achieved the highest average accuracy and F1 scores, suggesting its attention mechanism effectively utilized the narrow-but-moderately-deep architecture (16 layers, 4 hidden dim) for classification.\n",
    "\n",
    "- Avg. Accuracy: 0.9213 (+/- 0.1321)\n",
    "- Avg. F1 Weighted: 0.9045 (+/- 0.1656)\n",
    "\n",
    "### 2. GCN Performance\n",
    "\n",
    "GCN performed very strongly, showing that its highly unusual deep-and-narrow architecture (64 layers, 16 hidden dim) was successful. However, like the other models, it exhibited high standard deviations, indicating its performance was also unstable across the folds.\n",
    "\n",
    "- Avg. Accuracy: 0.8821 (+/- 0.1353)\n",
    "- Avg. F1 Weighted: 0.8639 (+/- 0.1666)\n",
    "\n",
    "### 3. SAGE Performance\n",
    "\n",
    "SAGE yielded the lowest average performance and the highest variability (largest +/- values). This suggests that the shallow-but-wide architecture (2 layers, 128 hidden dim) was less stable and less effective for this specific classification task compared to the other two models.\n",
    "\n",
    "- Avg. Accuracy: 0.7567 (+/- 0.1846)\n",
    "- Avg. F1 Weighted: 0.7311 (+/- 0.2160)\n",
    "\n",
    "\n",
    "### Summary\n",
    "\n",
    "While the GAT and GCN models achieved high average performance, all three models suffered from high standard deviations. This indicates that while the architectures are capable of high accuracy, their performance is highly sensitive.\n",
    "\n",
    "**These results are shown below**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a77d244",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import f1_score, accuracy_score\n",
    "from bioneuralnet.downstream_task import DPMON\n",
    "\n",
    "X_meth_full = dna_meth\n",
    "X_rna_full = rna\n",
    "X_mirna_full = mirna\n",
    "C_full = clinical_preprocessed\n",
    "Y_full = target\n",
    "A_full = A_train\n",
    "\n",
    "patient_indices = np.arange(len(Y_full))\n",
    "y_target_classes = Y_full.squeeze()\n",
    "\n",
    "sage_params = {\n",
    "    'layer_num': 2,\n",
    "    'gnn_hidden_dim': 128,\n",
    "    'lr': 0.005197038211089667,\n",
    "    'weight_decay': 0.04607928341516355,\n",
    "    'nn_hidden_dim1': 16,\n",
    "    'nn_hidden_dim2': 32,\n",
    "    'num_epochs': 4096\n",
    "}\n",
    "\n",
    "N_FOLDS = 5\n",
    "skf = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=SEED) \n",
    "output_dir_base_sage = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_cv_results_SAGE_FINAL\")\n",
    "\n",
    "accuracy_scores_sage = []\n",
    "f1_macro_scores_sage = []\n",
    "f1_weighted_scores_sage = []\n",
    "\n",
    "for fold_num, (train_index, test_index) in enumerate(skf.split(patient_indices, y_target_classes)):\n",
    "    print(f\"\\nRunning DPMON SAGE Fold {fold_num + 1}/{N_FOLDS}\")\n",
    "\n",
    "    X_train_omics = [X_meth_full.iloc[train_index], X_rna_full.iloc[train_index], X_mirna_full.iloc[train_index]]\n",
    "    Y_train = Y_full.iloc[train_index]\n",
    "    C_train = C_full.iloc[train_index]\n",
    "    A_train_fold = A_full.iloc[train_index, train_index] \n",
    "    \n",
    "    dpmon_fold = DPMON(\n",
    "        adjacency_matrix=A_train_fold, omics_list=X_train_omics, phenotype_data=Y_train,\n",
    "        clinical_data=C_train, repeat_num=1, tune=False, gpu=True, cuda=0,\n",
    "        output_dir=output_dir_base_sage / f\"fold_{fold_num + 1}\",\n",
    "        **sage_params\n",
    "    )\n",
    "\n",
    "    predictions_df, _ = dpmon_fold.run() \n",
    "    actual = predictions_df[\"Actual\"]\n",
    "    pred = predictions_df[\"Predicted\"]\n",
    "\n",
    "    test_f1_macro = f1_score(actual, pred, average='macro')\n",
    "    test_f1_weighted = f1_score(actual, pred, average='weighted')\n",
    "    test_acc = accuracy_score(actual, pred)\n",
    "\n",
    "    accuracy_scores_sage.append(test_acc)\n",
    "    f1_macro_scores_sage.append(test_f1_macro)\n",
    "    f1_weighted_scores_sage.append(test_f1_weighted)\n",
    "    \n",
    "    print(f\"SAGE Fold {fold_num + 1} | Accuracy: {test_acc:.4f} | F1-Macro: {test_f1_macro:.4f}\")\n",
    "\n",
    "mean_acc_sage = np.mean(accuracy_scores_sage)\n",
    "std_acc_sage = np.std(accuracy_scores_sage)\n",
    "\n",
    "mean_f1_macro_sage = np.mean(f1_macro_scores_sage)\n",
    "std_f1_macro_sage = np.std(f1_macro_scores_sage)\n",
    "\n",
    "mean_f1_weighted_sage = np.mean(f1_weighted_scores_sage)\n",
    "std_f1_weighted_sage = np.std(f1_weighted_scores_sage)\n",
    "\n",
    "print(\"\\nClassification with SAGE and 5-FOLD Cross-Validation\")\n",
    "print(f\"Avg. Accuracy: {mean_acc_sage:.4f} +/- {std_acc_sage:.4f}\")\n",
    "print(f\"Avg. F1 Weighted: {mean_f1_weighted_sage:.4f} +/- {std_f1_weighted_sage:.4f}\")\n",
    "print(f\"Avg. F1 Macro: {mean_f1_macro_sage:.4f} +/- {std_f1_macro_sage:.4f}\")\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-GBMLGG.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8b344bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir_base_gcn = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_cv_results_GCN_FINAL\")\n",
    "\n",
    "gcn_params = {\n",
    "    'layer_num': 64,\n",
    "    'gnn_hidden_dim': 16,\n",
    "    'lr': 0.004397636847528492,\n",
    "    'weight_decay': 0.00040569459559797747,\n",
    "    'nn_hidden_dim1': 4,\n",
    "    'nn_hidden_dim2': 8,\n",
    "    'num_epochs': 1024\n",
    "}\n",
    "\n",
    "skf_gcn = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=SEED) \n",
    "\n",
    "accuracy_scores_gcn = []\n",
    "f1_macro_scores_gcn = []\n",
    "f1_weighted_scores_gcn = []\n",
    "\n",
    "for fold_num, (train_index, test_index) in enumerate(skf_gcn.split(patient_indices, y_target_classes)):\n",
    "    print(f\"\\nDPMON GCN Fold {fold_num + 1}/{N_FOLDS}\")\n",
    "\n",
    "    X_train_omics = [X_meth_full.iloc[train_index], X_rna_full.iloc[train_index], X_mirna_full.iloc[train_index]]\n",
    "    Y_train = Y_full.iloc[train_index]\n",
    "    C_train = C_full.iloc[train_index]\n",
    "    A_train_fold = A_full.iloc[train_index, train_index] \n",
    "    \n",
    "    dpmon_fold = DPMON(\n",
    "        adjacency_matrix=A_train_fold, omics_list=X_train_omics, phenotype_data=Y_train,\n",
    "        clinical_data=C_train, repeat_num=1, tune=False, gpu=True, cuda=0,\n",
    "        output_dir=output_dir_base_gcn / f\"fold_{fold_num + 1}\",\n",
    "        **gcn_params\n",
    "    )\n",
    "\n",
    "    predictions_df, _ = dpmon_fold.run() \n",
    "    actual = predictions_df[\"Actual\"]\n",
    "    pred = predictions_df[\"Predicted\"]\n",
    "\n",
    "    test_acc = accuracy_score(actual, pred)\n",
    "    test_f1 = f1_score(actual, pred, average='macro')\n",
    "    test_f1w = f1_score(actual, pred, average='weighted')\n",
    "\n",
    "    accuracy_scores_gcn.append(test_acc)\n",
    "    f1_macro_scores_gcn.append(test_f1)\n",
    "    f1_weighted_scores_gcn.append(test_f1w)\n",
    "    \n",
    "    print(f\"GCN Fold {fold_num + 1} | Accuracy: {test_acc:.4f} | F1-Macro: {test_f1:.4f}\")\n",
    "\n",
    "mean_acc_gcn = np.mean(accuracy_scores_gcn)\n",
    "std_acc_gcn = np.std(accuracy_scores_gcn)\n",
    "\n",
    "mean_f1_macro_gcn = np.mean(f1_macro_scores_gcn)\n",
    "std_f1_macro_gcn = np.std(f1_macro_scores_gcn)\n",
    "\n",
    "mean_f1_weighted_gcn = np.mean(f1_weighted_scores_gcn)\n",
    "std_f1_weighted_gcn = np.std(f1_weighted_scores_gcn)\n",
    "\n",
    "print(\"Classification with GCN and 5-FOLD Cross-Validation\")\n",
    "print(f\"Avg. Accuracy: {mean_acc_gcn:.4f} +/- {std_acc_gcn:.4f}\")\n",
    "print(f\"Avg. F1 Weighted: {mean_f1_weighted_gcn:.4f} +/- {std_f1_weighted_gcn:.4f}\")\n",
    "print(f\"Avg. F1 Macro: {mean_f1_macro_gcn:.4f} +/- {std_f1_macro_gcn:.4f}\")\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-GBMLGG.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a40cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir_base_gat = Path(\"/home/vicente/Github/BioNeuralNet/dpmon_cv_results_GAT_FINAL\")\n",
    "\n",
    "gat_params = {\n",
    "    'layer_num': 16,\n",
    "    'gnn_hidden_dim': 4,\n",
    "    'lr': 0.003226646486974539,\n",
    "    'weight_decay': 0.04686394640851084,\n",
    "    'nn_hidden_dim1': 16,\n",
    "    'nn_hidden_dim2': 128,\n",
    "    'num_epochs': 512\n",
    "}\n",
    "\n",
    "skf_gat = StratifiedKFold(n_splits=N_FOLDS, shuffle=True, random_state=SEED) \n",
    "\n",
    "accuracy_scores_gat = []\n",
    "f1_macro_scores_gat = []\n",
    "f1_weighted_scores_gat = []\n",
    "\n",
    "for fold_num, (train_index, test_index) in enumerate(skf_gat.split(patient_indices, y_target_classes)):\n",
    "    print(f\"\\nDPMON GAT Fold {fold_num + 1}/{N_FOLDS}\")\n",
    "\n",
    "    X_train_omics = [X_meth_full.iloc[train_index], X_rna_full.iloc[train_index], X_mirna_full.iloc[train_index]]\n",
    "    Y_train = Y_full.iloc[train_index]\n",
    "    C_train = C_full.iloc[train_index]\n",
    "    A_train_fold = A_full.iloc[train_index, train_index] \n",
    "    \n",
    "    dpmon_fold = DPMON(\n",
    "        adjacency_matrix=A_train_fold, omics_list=X_train_omics, phenotype_data=Y_train,\n",
    "        clinical_data=C_train, repeat_num=1, tune=False, gpu=True, cuda=0,\n",
    "        output_dir=output_dir_base_gat / f\"fold_{fold_num + 1}\",\n",
    "        **gat_params\n",
    "    )\n",
    "\n",
    "    predictions_df, _ = dpmon_fold.run() \n",
    "    actual = predictions_df[\"Actual\"]\n",
    "    pred = predictions_df[\"Predicted\"]\n",
    "\n",
    "    test_acc = accuracy_score(actual, pred)\n",
    "    test_f1 = f1_score(actual, pred, average='macro')\n",
    "    test_f1w = f1_score(actual, pred, average='weighted')\n",
    "\n",
    "    accuracy_scores_gat.append(test_acc)\n",
    "    f1_macro_scores_gat.append(test_f1)\n",
    "    f1_weighted_scores_gat.append(test_f1w)\n",
    "    \n",
    "    print(f\"GAT Fold {fold_num + 1} | Accuracy: {test_acc:.4f} | F1-Macro: {test_f1:.4f}\")\n",
    "\n",
    "mean_acc_gat = np.mean(accuracy_scores_gat)\n",
    "std_acc_gat = np.std(accuracy_scores_gat)\n",
    "\n",
    "mean_f1_macro_gat = np.mean(f1_macro_scores_gat)\n",
    "std_f1_macro_gat = np.std(f1_macro_scores_gat)\n",
    "\n",
    "mean_f1_weighted_gat = np.mean(f1_weighted_scores_gat)\n",
    "std_f1_weighted_gat = np.std(f1_weighted_scores_gat)\n",
    "\n",
    "print(\"Classification with GAT and 5-FOLD Cross-Validation\")\n",
    "print(f\"Avg. Accuracy: {mean_acc_gat:.4f} +/- {std_acc_gat:.4f}\")\n",
    "print(f\"Avg. F1 Weighted: {mean_f1_weighted_gat:.4f} +/- {std_f1_weighted_gat:.4f}\")\n",
    "print(f\"Avg. F1 Macro: {mean_f1_macro_gat:.4f} +/- {std_f1_macro_gat:.4f}\")\n",
    "\n",
    "# Due to the extensive length, the terminal output has been omitted. To see the full terminal output for the results below please visit: https://github.com/UCD-BDLab/BioNeuralNet/TCGA-Notebooks/TCGA-GBMLGG.ipynb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".enviroment",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
